// @dcl/quests-client/dist-cjs/protocol/google/protobuf/empty.gen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Empty = exports.protobufPackage = void 0;
const minimal_1 = require("protobufjs/minimal");
exports.protobufPackage = "google.protobuf";
function createBaseEmpty() {
    return {};
}
var Empty;
(function (Empty) {
    function encode(_, writer = minimal_1.default.Writer.create()) {
        return writer;
    }
    Empty.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseEmpty();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    Empty.decode = decode;
    function fromJSON(_) {
        return {};
    }
    Empty.fromJSON = fromJSON;
    function toJSON(_) {
        const obj = {};
        return obj;
    }
    Empty.toJSON = toJSON;
    function create(base) {
        return Empty.fromPartial(base ?? {});
    }
    Empty.create = create;
    function fromPartial(_) {
        const message = createBaseEmpty();
        return message;
    }
    Empty.fromPartial = fromPartial;
})(Empty = exports.Empty || (exports.Empty = {}));
