// @dcl/quests-client/dist/hud.js
import { engine, Schemas } from '@dcl/sdk/ecs';
import { Color4 } from '@dcl/sdk/math';
import ReactEcs, { Button, Label, ReactEcsRenderer, UiEntity } from '@dcl/sdk/react-ecs';
const QuestComponent = engine.defineComponent('dcl:quests:QuestComponent', {
    name: Schemas.String,
    steps: Schemas.Array(Schemas.Map({
        id: Schemas.String,
        description: Schemas.String,
        tasks: Schemas.Array(Schemas.Map({
            done: Schemas.Boolean,
            description: Schemas.String
        }))
    })),
    nextSteps: Schemas.Array(Schemas.String),
    done: Schemas.Boolean
});
export function createQuestHUD(opts) {
    let state = {
        hideHUD: false,
        selectedStepId: null,
        rawEntity: null,
        quest: null
    };
    function getState() {
        return { ...state };
    }
    function setState(newState) {
        state = {
            ...state,
            ...newState
        };
    }
    if (opts?.autoRender) {
        ReactEcsRenderer.setUiRenderer(() => HUDPanel([getState, setState], opts));
    }
    function getQuest() {
        return [...engine.getEntitiesWith(QuestComponent)][0][1];
    }
    return {
        upsert: (instance) => {
            const questUIData = generateQuestUI(instance);
            if (state.rawEntity) {
                QuestComponent.createOrReplace(state.rawEntity, {
                    ...questUIData
                });
                state.quest = getQuest();
            }
            else {
                const questEntity = engine.addEntity();
                QuestComponent.create(questEntity, { ...questUIData });
                state.rawEntity = questEntity;
                state.quest = getQuest();
            }
        },
        getHUDComponent: () => () => HUDPanel([getState, setState], opts),
        render: () => ReactEcsRenderer.setUiRenderer(() => HUDPanel([getState, setState], opts)),
        updateOptions: (newOptions) => ReactEcsRenderer.setUiRenderer(() => HUDPanel([getState, setState], { ...opts, ...newOptions })),
        getHUDComponentWithUpdatedOptions: (newOptions) => () => HUDPanel([getState, setState], { ...opts, ...newOptions })
    };
}
const defaultHudBox = {
    flexDirection: 'column',
    justifyContent: 'flex-start',
    margin: { left: 10 },
    positionType: 'absolute',
    position: { top: '28%' }
};
export const HUDPanel = (handlers, props) => {
    const [getState, setState] = handlers;
    const { selectedStepId: showTask, hideHUD, quest } = getState();
    return (ReactEcs.createElement(UiEntity, { uiTransform: {
            ...defaultHudBox,
            ...(props?.leftSidePanel || {})
        } },
        !showTask && quest ? QuestProgress(quest, handlers, props) : StepTasks(handlers, props),
        quest ? (ReactEcs.createElement(UiEntity, { uiTransform: { display: 'flex' } }, !showTask ? (ReactEcs.createElement(Button, { value: hideHUD ? 'Show Quest Progress' : 'Hide', variant: "primary", uiTransform: {
                width: !hideHUD ? 80 : 120,
                height: !hideHUD ? 20 : 30,
                margin: {
                    top: 5
                }
            }, onMouseDown: () => setState({ ...getState(), hideHUD: !hideHUD }), ...props?.showHideToggleButton })) : (ReactEcs.createElement(Button, { value: "Close", variant: "primary", uiTransform: {
                width: 80,
                height: 20,
                margin: {
                    top: 5
                }
            }, onMouseDown: () => setState({ ...getState(), selectedStepId: null }), ...props?.closeTasksBoxButton })))) : null));
};
const QuestProgress = (quest, handlers, stylingOpts) => {
    const [getState] = handlers;
    return (ReactEcs.createElement(UiEntity, { uiTransform: {
            display: getState().hideHUD ? 'none' : 'flex',
            flexDirection: 'column',
            padding: 10,
            minWidth: 180,
            maxWidth: 400,
            width: 180,
            ...stylingOpts?.questBox?.uiTransform
        }, uiBackground: {
            color: Color4.create(0, 0, 0, 0.9),
            ...stylingOpts?.questBox?.uiBackground
        } },
        ReactEcs.createElement(UiEntity, { uiTransform: { margin: { bottom: 10 }, ...stylingOpts?.questNameContainer?.uiTransform } },
            ReactEcs.createElement(Label, { textAlign: "middle-left", fontSize: 13, value: `Quest: ${quest.name}`, ...stylingOpts?.questNameContainer?.label })),
        !quest.done ? (ReactEcs.createElement(UiEntity, { uiTransform: {
                flexDirection: 'column',
                padding: { top: 10, bottom: 10 },
                ...stylingOpts?.stepsContainer?.uiTransform
            } }, [...quest.steps].map((s) => CurrentStep(s, handlers, stylingOpts)))) : (ReactEcs.createElement(UiEntity, { uiTransform: {
                flexDirection: 'column',
                padding: { top: 10, bottom: 10 },
                ...stylingOpts?.questCompletionLabel?.uiTransform
            } },
            ReactEcs.createElement(Label, { uiTransform: {
                    margin: { bottom: 10 }
                }, textAlign: "middle-center", fontSize: 12, value: "Quest is completed!", ...stylingOpts?.questCompletionLabel?.label }))),
        quest.nextSteps.length > 0 && !quest.done ? (ReactEcs.createElement(UiEntity, { uiTransform: {
                display: 'flex',
                flexDirection: 'column',
                margin: { bottom: 10 },
                ...stylingOpts?.nextSteps?.nextTitleUiEntity
            } },
            ReactEcs.createElement(Label, { uiTransform: {
                    padding: { bottom: 10 },
                    margin: { bottom: 10 },
                    ...stylingOpts?.nextSteps?.nextTitleProps?.uiTransform
                }, textAlign: "middle-left", fontSize: 12, value: "Next: ", ...stylingOpts?.nextSteps?.nextTitleProps }),
            [...quest.nextSteps].map((next) => (ReactEcs.createElement(UiEntity, { uiTransform: { margin: { bottom: 5 }, ...stylingOpts?.nextSteps?.labels?.labelUiEntity } },
                ReactEcs.createElement(Label, { textAlign: "middle-left", fontSize: 11, value: next, ...stylingOpts?.nextSteps?.labels?.props })))))) : null));
};
const CurrentStep = (step, handlers, stylingOpts) => {
    return (ReactEcs.createElement(UiEntity, { uiTransform: { flexDirection: 'column', ...stylingOpts?.stepsContainer?.labels?.labelUiEntity } },
        ReactEcs.createElement(Label, { uiTransform: {
                overflow: 'hidden',
                padding: { top: 10, bottom: 10 },
                ...stylingOpts?.stepsContainer?.labels?.props?.uiTransform
            }, textAlign: "middle-left", fontSize: 12, value: `Step: ${step.description}`, ...stylingOpts?.stepsContainer?.labels?.props }),
        ReactEcs.createElement(ShowTasksButton, { stepId: step.id, handlers: handlers, stylingOpts: stylingOpts })));
};
const ShowTasksButton = ({ stepId, handlers, stylingOpts }) => {
    const [getState, setState] = handlers;
    return (ReactEcs.createElement(UiEntity, { uiTransform: { display: 'flex', ...stylingOpts?.stepsContainer?.showTasksButton?.buttonUiEntity } },
        ReactEcs.createElement(Button, { value: "Show Tasks", variant: "primary", uiTransform: { width: 80, height: 20, margin: 4 }, onMouseDown: () => {
                setState({ ...getState(), selectedStepId: stepId });
            }, ...stylingOpts?.stepsContainer?.showTasksButton?.buttonProps })));
};
const StepTasks = ([getState], stylingOpts) => {
    const { selectedStepId, quest } = getState();
    return (ReactEcs.createElement(UiEntity, { uiTransform: {
            flexDirection: 'column',
            padding: 10,
            minWidth: 180,
            maxWidth: 400,
            width: 180,
            ...stylingOpts?.tasksBox?.uiTransform
        }, uiBackground: {
            color: Color4.create(0, 0, 0, 0.9),
            ...stylingOpts?.tasksBox?.uiBackground
        } }, selectedStepId != null && quest != null
        ? quest.steps
            .find((s) => s.id === selectedStepId)
            ?.tasks.map((t) => Task(t, stylingOpts))
        : null));
};
const Task = (task, stylingOpts) => (ReactEcs.createElement(UiEntity, { uiTransform: {
        flexDirection: 'column',
        margin: { bottom: 10 },
        ...stylingOpts?.tasksBox?.labels?.labelUiEntity
    } },
    ReactEcs.createElement(Label, { textAlign: "middle-left", fontSize: 12, value: `${task.done ? 'DONE: ' : 'TODO: '} ${task.description}`, ...stylingOpts?.tasksBox?.labels?.props })));
function generateQuestUI(questInstance) {
    const steps = [];
    const nextSteps = [];
    if (questInstance.quest.definition?.steps) {
        for (const step of questInstance.quest.definition?.steps) {
            if (questInstance.state.currentSteps[step.id]) {
                const content = questInstance.state.currentSteps[step.id];
                steps.push({
                    id: step.id,
                    description: step.description,
                    tasks: step.tasks.map((task) => {
                        return {
                            description: task.description,
                            done: !!content.tasksCompleted.find((t) => t.id == task.id)
                        };
                    })
                });
                nextSteps.push(...questInstance.quest.definition?.connections
                    .filter((conn) => conn.stepFrom === step.id)
                    .map((conn) => questInstance.quest.definition?.steps.find((step) => step.id === conn.stepTo)?.description ||
                    '(Step without description)'));
            }
        }
    }
    return { name: questInstance.quest.name, steps, nextSteps, done: questInstance.state.stepsLeft == 0 };
}
