// @dcl/sdk/network/binary-message-bus.js
import { ReadWriteByteBuffer } from '@dcl/ecs/dist/serialization/ByteBuffer';
export var CommsMessage;
(function (CommsMessage) {
    CommsMessage[CommsMessage["CRDT"] = 1] = "CRDT";
    CommsMessage[CommsMessage["REQ_CRDT_STATE"] = 2] = "REQ_CRDT_STATE";
    CommsMessage[CommsMessage["RES_CRDT_STATE"] = 3] = "RES_CRDT_STATE";
})(CommsMessage || (CommsMessage = {}));
export function BinaryMessageBus(send) {
    const mapping = new Map();
    return {
        on: (message, callback) => {
            mapping.set(message, callback);
        },
        emit: (message, value) => {
            send(craftCommsMessage(message, value));
        },
        __processMessages: (messages) => {
            for (const message of messages) {
                const commsMsg = decodeCommsMessage(message);
                if (!commsMsg)
                    continue;
                const { sender, messageType, data } = commsMsg;
                const fn = mapping.get(messageType);
                if (fn)
                    fn(data, sender);
            }
        }
    };
}
export function craftCommsMessage(messageType, payload) {
    const msg = new Uint8Array(payload.byteLength + 1);
    msg.set([messageType]);
    msg.set(payload, 1);
    return msg;
}
export function decodeCommsMessage(data) {
    try {
        let offset = 0;
        const r = new Uint8Array(data);
        const view = new DataView(r.buffer);
        const senderLength = view.getUint8(offset);
        offset += 1;
        const sender = decodeString(data.subarray(1, senderLength + 1));
        offset += senderLength;
        const messageType = view.getUint8(offset);
        offset += 1;
        const message = r.subarray(offset);
        return {
            sender,
            messageType,
            data: message
        };
    }
    catch (e) {
        console.error('Invalid Comms message', e);
    }
}
export function decodeString(data) {
    const buffer = new ReadWriteByteBuffer();
    buffer.writeBuffer(data, true);
    return buffer.readUtf8String();
}
export function encodeString(s) {
    const buffer = new ReadWriteByteBuffer();
    buffer.writeUtf8String(s);
    return buffer.readBuffer();
}
