// @dcl/protocol/out-ts/decentraland/realm/about.gen.ts
/* eslint-disable */
import _m0 from "protobufjs/minimal";
import { BorderRect } from "../common/border_rect.gen";
import { Vector2 } from "../common/vectors.gen";

export const protobufPackage = "decentraland.realm";

export interface AboutResponse {
  healthy: boolean;
  configurations: AboutResponse_AboutConfiguration | undefined;
  content: AboutResponse_ContentInfo | undefined;
  comms: AboutResponse_CommsInfo | undefined;
  lambdas: AboutResponse_LambdasInfo | undefined;
  bff?: AboutResponse_BffInfo | undefined;
  acceptingUsers: boolean;
}

/**
 * @deprecated This message was never used but it's still here for compatibility reasons
 * The reserved keyword is not working for all the toolset we have, so in the meanwhile it keeps commented
 */
export interface AboutResponse_MinimapConfiguration {
}

export interface AboutResponse_MapConfiguration {
  /** whether the minimap should be rendered */
  minimapEnabled: boolean;
  /**
   * the union of all rects here represents the places where that MAY have scenes
   *  - all the other parcels are considered as empty, up to the explorer to decide if they're walkable or not
   *  - the contained square is determined by points:
   *    - top-left      with  minX, maxY
   *    - bottom-left   with  minX, minY
   *    - top-right     with  maxX, maxY
   *    - bottom-right  with  maxX, maxY
   *  Note: the coordinate system used is the Cartesian coordinate system, where the y-axis increases upwards,
   *    not the screen coordinate system, where the origin is at the top-left corner and the y-axis increases downwards.
   */
  sizes: BorderRect[];
  satelliteView?: AboutResponse_MapConfiguration_ImageViewWithZoomLevel | undefined;
  parcelView?: AboutResponse_MapConfiguration_ParcelView | undefined;
  thumbnailView?: AboutResponse_MapConfiguration_ImageViewWithZoomLevel | undefined;
}

/**
 * ImageViewWithZoomLevel uses a description to render the minimap
 *  using different images with different zoom levels.
 * When `version='v1'`:
 * - The description allows the explorer to form the URL:
 *  url(x,y,zoom_level) = `{base_url}/{zoom_level}/{x},{y}{suffix_url}`
 * - The given URL is for an image where the pixel `0,0` for the image `0,0` always
 *    points to the top-left contained square.
 * - The `zoom_level=1` is to a ratio of 3.2 pixel per parcel, this means in a 32x32 pixel square
 *    you get 10x10 parcels.
 * - Each increase of zoom level, double the ratio pixel per parcels.
 */
export interface AboutResponse_MapConfiguration_ImageViewWithZoomLevel {
  /** options: ['v1'] */
  version: string;
  baseUrl?: string | undefined;
  suffixUrl?: string | undefined;
  topLeftOffset?: Vector2 | undefined;
}

/**
 * ParcelView uses a description to render the minimap
 *  using a image where each pixel is a parcel and each pixel
 *  has the metadata to make a representation (using a shader or image-generation client-side)
 * When `version='v1'`:
 *  - The pixel `0,0` is the top-left contained square
 *  - The image has to be at least of contained square pixels size
 *  - The metadata inside each pixel follows the generated with the atlas server https://github.com/decentraland/atlas-server/blob/af371f2a59745a1f50b0b0b2382984288c4ae891/src/adapters/mini-map-renderer.ts#L27
 */
export interface AboutResponse_MapConfiguration_ParcelView {
  /** options: `v1` */
  version: string;
  imageUrl?: string | undefined;
}

export interface AboutResponse_SkyboxConfiguration {
  /** only one value at a time */
  fixedHour?: number | undefined;
}

export interface AboutResponse_AboutConfiguration {
  realmName?: string | undefined;
  networkId: number;
  globalScenesUrn: string[];
  scenesUrn: string[];
  skybox?:
    | AboutResponse_SkyboxConfiguration
    | undefined;
  /**
   * A content server to be used to load the parcels around the user. Uses the POST /entities/active endpoint
   * to continously fetch the parcels around the users. if null, then the default content server will be used
   * if == "" then the city_loader will be disabled and the scenes_urn will be used to load the world
   */
  cityLoaderContentServer?: string | undefined;
  map?: AboutResponse_MapConfiguration | undefined;
}

export interface AboutResponse_ContentInfo {
  /** common properties */
  healthy: boolean;
  version?: string | undefined;
  commitHash?: string | undefined;
  publicUrl: string;
}

export interface AboutResponse_LambdasInfo {
  /** common properties */
  healthy: boolean;
  version?: string | undefined;
  commitHash?: string | undefined;
  publicUrl: string;
}

export interface AboutResponse_CommsInfo {
  /** common properties */
  healthy: boolean;
  version?: string | undefined;
  commitHash?: string | undefined;
  publicUrl?:
    | string
    | undefined;
  /** specific properties */
  protocol: string;
  usersCount?: number | undefined;
  fixedAdapter?: string | undefined;
}

export interface AboutResponse_BffInfo {
  /** common properties */
  healthy: boolean;
  version?: string | undefined;
  commitHash?: string | undefined;
  publicUrl: string;
  /** specific properties */
  userCount?: number | undefined;
  protocolVersion?: string | undefined;
}

function createBaseAboutResponse(): AboutResponse {
  return {
    healthy: false,
    configurations: undefined,
    content: undefined,
    comms: undefined,
    lambdas: undefined,
    bff: undefined,
    acceptingUsers: false,
  };
}

export namespace AboutResponse {
  export function encode(message: AboutResponse, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.healthy === true) {
      writer.uint32(8).bool(message.healthy);
    }
    if (message.configurations !== undefined) {
      AboutResponse_AboutConfiguration.encode(message.configurations, writer.uint32(18).fork()).ldelim();
    }
    if (message.content !== undefined) {
      AboutResponse_ContentInfo.encode(message.content, writer.uint32(26).fork()).ldelim();
    }
    if (message.comms !== undefined) {
      AboutResponse_CommsInfo.encode(message.comms, writer.uint32(34).fork()).ldelim();
    }
    if (message.lambdas !== undefined) {
      AboutResponse_LambdasInfo.encode(message.lambdas, writer.uint32(42).fork()).ldelim();
    }
    if (message.bff !== undefined) {
      AboutResponse_BffInfo.encode(message.bff, writer.uint32(50).fork()).ldelim();
    }
    if (message.acceptingUsers === true) {
      writer.uint32(56).bool(message.acceptingUsers);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): AboutResponse {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseAboutResponse();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 8) {
            break;
          }

          message.healthy = reader.bool();
          continue;
        case 2:
          if (tag !== 18) {
            break;
          }

          message.configurations = AboutResponse_AboutConfiguration.decode(reader, reader.uint32());
          continue;
        case 3:
          if (tag !== 26) {
            break;
          }

          message.content = AboutResponse_ContentInfo.decode(reader, reader.uint32());
          continue;
        case 4:
          if (tag !== 34) {
            break;
          }

          message.comms = AboutResponse_CommsInfo.decode(reader, reader.uint32());
          continue;
        case 5:
          if (tag !== 42) {
            break;
          }

          message.lambdas = AboutResponse_LambdasInfo.decode(reader, reader.uint32());
          continue;
        case 6:
          if (tag !== 50) {
            break;
          }

          message.bff = AboutResponse_BffInfo.decode(reader, reader.uint32());
          continue;
        case 7:
          if (tag !== 56) {
            break;
          }

          message.acceptingUsers = reader.bool();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): AboutResponse {
    return {
      healthy: isSet(object.healthy) ? Boolean(object.healthy) : false,
      configurations: isSet(object.configurations)
        ? AboutResponse_AboutConfiguration.fromJSON(object.configurations)
        : undefined,
      content: isSet(object.content) ? AboutResponse_ContentInfo.fromJSON(object.content) : undefined,
      comms: isSet(object.comms) ? AboutResponse_CommsInfo.fromJSON(object.comms) : undefined,
      lambdas: isSet(object.lambdas) ? AboutResponse_LambdasInfo.fromJSON(object.lambdas) : undefined,
      bff: isSet(object.bff) ? AboutResponse_BffInfo.fromJSON(object.bff) : undefined,
      acceptingUsers: isSet(object.acceptingUsers) ? Boolean(object.acceptingUsers) : false,
    };
  }

  export function toJSON(message: AboutResponse): unknown {
    const obj: any = {};
    message.healthy !== undefined && (obj.healthy = message.healthy);
    message.configurations !== undefined &&
      (obj.configurations = message.configurations
        ? AboutResponse_AboutConfiguration.toJSON(message.configurations)
        : undefined);
    message.content !== undefined &&
      (obj.content = message.content ? AboutResponse_ContentInfo.toJSON(message.content) : undefined);
    message.comms !== undefined &&
      (obj.comms = message.comms ? AboutResponse_CommsInfo.toJSON(message.comms) : undefined);
    message.lambdas !== undefined &&
      (obj.lambdas = message.lambdas ? AboutResponse_LambdasInfo.toJSON(message.lambdas) : undefined);
    message.bff !== undefined && (obj.bff = message.bff ? AboutResponse_BffInfo.toJSON(message.bff) : undefined);
    message.acceptingUsers !== undefined && (obj.acceptingUsers = message.acceptingUsers);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<AboutResponse>, I>>(base?: I): AboutResponse {
    return AboutResponse.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<AboutResponse>, I>>(object: I): AboutResponse {
    const message = createBaseAboutResponse();
    message.healthy = object.healthy ?? false;
    message.configurations = (object.configurations !== undefined && object.configurations !== null)
      ? AboutResponse_AboutConfiguration.fromPartial(object.configurations)
      : undefined;
    message.content = (object.content !== undefined && object.content !== null)
      ? AboutResponse_ContentInfo.fromPartial(object.content)
      : undefined;
    message.comms = (object.comms !== undefined && object.comms !== null)
      ? AboutResponse_CommsInfo.fromPartial(object.comms)
      : undefined;
    message.lambdas = (object.lambdas !== undefined && object.lambdas !== null)
      ? AboutResponse_LambdasInfo.fromPartial(object.lambdas)
      : undefined;
    message.bff = (object.bff !== undefined && object.bff !== null)
      ? AboutResponse_BffInfo.fromPartial(object.bff)
      : undefined;
    message.acceptingUsers = object.acceptingUsers ?? false;
    return message;
  }
}

function createBaseAboutResponse_MinimapConfiguration(): AboutResponse_MinimapConfiguration {
  return {};
}

export namespace AboutResponse_MinimapConfiguration {
  export function encode(_: AboutResponse_MinimapConfiguration, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): AboutResponse_MinimapConfiguration {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseAboutResponse_MinimapConfiguration();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(_: any): AboutResponse_MinimapConfiguration {
    return {};
  }

  export function toJSON(_: AboutResponse_MinimapConfiguration): unknown {
    const obj: any = {};
    return obj;
  }

  export function create<I extends Exact<DeepPartial<AboutResponse_MinimapConfiguration>, I>>(
    base?: I,
  ): AboutResponse_MinimapConfiguration {
    return AboutResponse_MinimapConfiguration.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<AboutResponse_MinimapConfiguration>, I>>(
    _: I,
  ): AboutResponse_MinimapConfiguration {
    const message = createBaseAboutResponse_MinimapConfiguration();
    return message;
  }
}

function createBaseAboutResponse_MapConfiguration(): AboutResponse_MapConfiguration {
  return {
    minimapEnabled: false,
    sizes: [],
    satelliteView: undefined,
    parcelView: undefined,
    thumbnailView: undefined,
  };
}

export namespace AboutResponse_MapConfiguration {
  export function encode(
    message: AboutResponse_MapConfiguration,
    writer: _m0.Writer = _m0.Writer.create(),
  ): _m0.Writer {
    if (message.minimapEnabled === true) {
      writer.uint32(8).bool(message.minimapEnabled);
    }
    for (const v of message.sizes) {
      BorderRect.encode(v!, writer.uint32(18).fork()).ldelim();
    }
    if (message.satelliteView !== undefined) {
      AboutResponse_MapConfiguration_ImageViewWithZoomLevel.encode(message.satelliteView, writer.uint32(42).fork())
        .ldelim();
    }
    if (message.parcelView !== undefined) {
      AboutResponse_MapConfiguration_ParcelView.encode(message.parcelView, writer.uint32(50).fork()).ldelim();
    }
    if (message.thumbnailView !== undefined) {
      AboutResponse_MapConfiguration_ImageViewWithZoomLevel.encode(message.thumbnailView, writer.uint32(58).fork())
        .ldelim();
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): AboutResponse_MapConfiguration {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseAboutResponse_MapConfiguration();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 8) {
            break;
          }

          message.minimapEnabled = reader.bool();
          continue;
        case 2:
          if (tag !== 18) {
            break;
          }

          message.sizes.push(BorderRect.decode(reader, reader.uint32()));
          continue;
        case 5:
          if (tag !== 42) {
            break;
          }

          message.satelliteView = AboutResponse_MapConfiguration_ImageViewWithZoomLevel.decode(reader, reader.uint32());
          continue;
        case 6:
          if (tag !== 50) {
            break;
          }

          message.parcelView = AboutResponse_MapConfiguration_ParcelView.decode(reader, reader.uint32());
          continue;
        case 7:
          if (tag !== 58) {
            break;
          }

          message.thumbnailView = AboutResponse_MapConfiguration_ImageViewWithZoomLevel.decode(reader, reader.uint32());
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): AboutResponse_MapConfiguration {
    return {
      minimapEnabled: isSet(object.minimapEnabled) ? Boolean(object.minimapEnabled) : false,
      sizes: Array.isArray(object?.sizes) ? object.sizes.map((e: any) => BorderRect.fromJSON(e)) : [],
      satelliteView: isSet(object.satelliteView)
        ? AboutResponse_MapConfiguration_ImageViewWithZoomLevel.fromJSON(object.satelliteView)
        : undefined,
      parcelView: isSet(object.parcelView)
        ? AboutResponse_MapConfiguration_ParcelView.fromJSON(object.parcelView)
        : undefined,
      thumbnailView: isSet(object.thumbnailView)
        ? AboutResponse_MapConfiguration_ImageViewWithZoomLevel.fromJSON(object.thumbnailView)
        : undefined,
    };
  }

  export function toJSON(message: AboutResponse_MapConfiguration): unknown {
    const obj: any = {};
    message.minimapEnabled !== undefined && (obj.minimapEnabled = message.minimapEnabled);
    if (message.sizes) {
      obj.sizes = message.sizes.map((e) => e ? BorderRect.toJSON(e) : undefined);
    } else {
      obj.sizes = [];
    }
    message.satelliteView !== undefined &&
      (obj.satelliteView = message.satelliteView
        ? AboutResponse_MapConfiguration_ImageViewWithZoomLevel.toJSON(message.satelliteView)
        : undefined);
    message.parcelView !== undefined &&
      (obj.parcelView = message.parcelView
        ? AboutResponse_MapConfiguration_ParcelView.toJSON(message.parcelView)
        : undefined);
    message.thumbnailView !== undefined &&
      (obj.thumbnailView = message.thumbnailView
        ? AboutResponse_MapConfiguration_ImageViewWithZoomLevel.toJSON(message.thumbnailView)
        : undefined);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<AboutResponse_MapConfiguration>, I>>(
    base?: I,
  ): AboutResponse_MapConfiguration {
    return AboutResponse_MapConfiguration.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<AboutResponse_MapConfiguration>, I>>(
    object: I,
  ): AboutResponse_MapConfiguration {
    const message = createBaseAboutResponse_MapConfiguration();
    message.minimapEnabled = object.minimapEnabled ?? false;
    message.sizes = object.sizes?.map((e) => BorderRect.fromPartial(e)) || [];
    message.satelliteView = (object.satelliteView !== undefined && object.satelliteView !== null)
      ? AboutResponse_MapConfiguration_ImageViewWithZoomLevel.fromPartial(object.satelliteView)
      : undefined;
    message.parcelView = (object.parcelView !== undefined && object.parcelView !== null)
      ? AboutResponse_MapConfiguration_ParcelView.fromPartial(object.parcelView)
      : undefined;
    message.thumbnailView = (object.thumbnailView !== undefined && object.thumbnailView !== null)
      ? AboutResponse_MapConfiguration_ImageViewWithZoomLevel.fromPartial(object.thumbnailView)
      : undefined;
    return message;
  }
}

function createBaseAboutResponse_MapConfiguration_ImageViewWithZoomLevel(): AboutResponse_MapConfiguration_ImageViewWithZoomLevel {
  return { version: "", baseUrl: undefined, suffixUrl: undefined, topLeftOffset: undefined };
}

export namespace AboutResponse_MapConfiguration_ImageViewWithZoomLevel {
  export function encode(
    message: AboutResponse_MapConfiguration_ImageViewWithZoomLevel,
    writer: _m0.Writer = _m0.Writer.create(),
  ): _m0.Writer {
    if (message.version !== "") {
      writer.uint32(10).string(message.version);
    }
    if (message.baseUrl !== undefined) {
      writer.uint32(18).string(message.baseUrl);
    }
    if (message.suffixUrl !== undefined) {
      writer.uint32(26).string(message.suffixUrl);
    }
    if (message.topLeftOffset !== undefined) {
      Vector2.encode(message.topLeftOffset, writer.uint32(34).fork()).ldelim();
    }
    return writer;
  }

  export function decode(
    input: _m0.Reader | Uint8Array,
    length?: number,
  ): AboutResponse_MapConfiguration_ImageViewWithZoomLevel {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseAboutResponse_MapConfiguration_ImageViewWithZoomLevel();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.version = reader.string();
          continue;
        case 2:
          if (tag !== 18) {
            break;
          }

          message.baseUrl = reader.string();
          continue;
        case 3:
          if (tag !== 26) {
            break;
          }

          message.suffixUrl = reader.string();
          continue;
        case 4:
          if (tag !== 34) {
            break;
          }

          message.topLeftOffset = Vector2.decode(reader, reader.uint32());
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): AboutResponse_MapConfiguration_ImageViewWithZoomLevel {
    return {
      version: isSet(object.version) ? String(object.version) : "",
      baseUrl: isSet(object.baseUrl) ? String(object.baseUrl) : undefined,
      suffixUrl: isSet(object.suffixUrl) ? String(object.suffixUrl) : undefined,
      topLeftOffset: isSet(object.topLeftOffset) ? Vector2.fromJSON(object.topLeftOffset) : undefined,
    };
  }

  export function toJSON(message: AboutResponse_MapConfiguration_ImageViewWithZoomLevel): unknown {
    const obj: any = {};
    message.version !== undefined && (obj.version = message.version);
    message.baseUrl !== undefined && (obj.baseUrl = message.baseUrl);
    message.suffixUrl !== undefined && (obj.suffixUrl = message.suffixUrl);
    message.topLeftOffset !== undefined &&
      (obj.topLeftOffset = message.topLeftOffset ? Vector2.toJSON(message.topLeftOffset) : undefined);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<AboutResponse_MapConfiguration_ImageViewWithZoomLevel>, I>>(
    base?: I,
  ): AboutResponse_MapConfiguration_ImageViewWithZoomLevel {
    return AboutResponse_MapConfiguration_ImageViewWithZoomLevel.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<AboutResponse_MapConfiguration_ImageViewWithZoomLevel>, I>>(
    object: I,
  ): AboutResponse_MapConfiguration_ImageViewWithZoomLevel {
    const message = createBaseAboutResponse_MapConfiguration_ImageViewWithZoomLevel();
    message.version = object.version ?? "";
    message.baseUrl = object.baseUrl ?? undefined;
    message.suffixUrl = object.suffixUrl ?? undefined;
    message.topLeftOffset = (object.topLeftOffset !== undefined && object.topLeftOffset !== null)
      ? Vector2.fromPartial(object.topLeftOffset)
      : undefined;
    return message;
  }
}

function createBaseAboutResponse_MapConfiguration_ParcelView(): AboutResponse_MapConfiguration_ParcelView {
  return { version: "", imageUrl: undefined };
}

export namespace AboutResponse_MapConfiguration_ParcelView {
  export function encode(
    message: AboutResponse_MapConfiguration_ParcelView,
    writer: _m0.Writer = _m0.Writer.create(),
  ): _m0.Writer {
    if (message.version !== "") {
      writer.uint32(10).string(message.version);
    }
    if (message.imageUrl !== undefined) {
      writer.uint32(18).string(message.imageUrl);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): AboutResponse_MapConfiguration_ParcelView {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseAboutResponse_MapConfiguration_ParcelView();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.version = reader.string();
          continue;
        case 2:
          if (tag !== 18) {
            break;
          }

          message.imageUrl = reader.string();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): AboutResponse_MapConfiguration_ParcelView {
    return {
      version: isSet(object.version) ? String(object.version) : "",
      imageUrl: isSet(object.imageUrl) ? String(object.imageUrl) : undefined,
    };
  }

  export function toJSON(message: AboutResponse_MapConfiguration_ParcelView): unknown {
    const obj: any = {};
    message.version !== undefined && (obj.version = message.version);
    message.imageUrl !== undefined && (obj.imageUrl = message.imageUrl);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<AboutResponse_MapConfiguration_ParcelView>, I>>(
    base?: I,
  ): AboutResponse_MapConfiguration_ParcelView {
    return AboutResponse_MapConfiguration_ParcelView.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<AboutResponse_MapConfiguration_ParcelView>, I>>(
    object: I,
  ): AboutResponse_MapConfiguration_ParcelView {
    const message = createBaseAboutResponse_MapConfiguration_ParcelView();
    message.version = object.version ?? "";
    message.imageUrl = object.imageUrl ?? undefined;
    return message;
  }
}

function createBaseAboutResponse_SkyboxConfiguration(): AboutResponse_SkyboxConfiguration {
  return { fixedHour: undefined };
}

export namespace AboutResponse_SkyboxConfiguration {
  export function encode(
    message: AboutResponse_SkyboxConfiguration,
    writer: _m0.Writer = _m0.Writer.create(),
  ): _m0.Writer {
    if (message.fixedHour !== undefined) {
      writer.uint32(13).float(message.fixedHour);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): AboutResponse_SkyboxConfiguration {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseAboutResponse_SkyboxConfiguration();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 13) {
            break;
          }

          message.fixedHour = reader.float();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): AboutResponse_SkyboxConfiguration {
    return { fixedHour: isSet(object.fixedHour) ? Number(object.fixedHour) : undefined };
  }

  export function toJSON(message: AboutResponse_SkyboxConfiguration): unknown {
    const obj: any = {};
    message.fixedHour !== undefined && (obj.fixedHour = message.fixedHour);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<AboutResponse_SkyboxConfiguration>, I>>(
    base?: I,
  ): AboutResponse_SkyboxConfiguration {
    return AboutResponse_SkyboxConfiguration.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<AboutResponse_SkyboxConfiguration>, I>>(
    object: I,
  ): AboutResponse_SkyboxConfiguration {
    const message = createBaseAboutResponse_SkyboxConfiguration();
    message.fixedHour = object.fixedHour ?? undefined;
    return message;
  }
}

function createBaseAboutResponse_AboutConfiguration(): AboutResponse_AboutConfiguration {
  return {
    realmName: undefined,
    networkId: 0,
    globalScenesUrn: [],
    scenesUrn: [],
    skybox: undefined,
    cityLoaderContentServer: undefined,
    map: undefined,
  };
}

export namespace AboutResponse_AboutConfiguration {
  export function encode(
    message: AboutResponse_AboutConfiguration,
    writer: _m0.Writer = _m0.Writer.create(),
  ): _m0.Writer {
    if (message.realmName !== undefined) {
      writer.uint32(10).string(message.realmName);
    }
    if (message.networkId !== 0) {
      writer.uint32(16).uint32(message.networkId);
    }
    for (const v of message.globalScenesUrn) {
      writer.uint32(26).string(v!);
    }
    for (const v of message.scenesUrn) {
      writer.uint32(34).string(v!);
    }
    if (message.skybox !== undefined) {
      AboutResponse_SkyboxConfiguration.encode(message.skybox, writer.uint32(50).fork()).ldelim();
    }
    if (message.cityLoaderContentServer !== undefined) {
      writer.uint32(58).string(message.cityLoaderContentServer);
    }
    if (message.map !== undefined) {
      AboutResponse_MapConfiguration.encode(message.map, writer.uint32(66).fork()).ldelim();
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): AboutResponse_AboutConfiguration {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseAboutResponse_AboutConfiguration();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.realmName = reader.string();
          continue;
        case 2:
          if (tag !== 16) {
            break;
          }

          message.networkId = reader.uint32();
          continue;
        case 3:
          if (tag !== 26) {
            break;
          }

          message.globalScenesUrn.push(reader.string());
          continue;
        case 4:
          if (tag !== 34) {
            break;
          }

          message.scenesUrn.push(reader.string());
          continue;
        case 6:
          if (tag !== 50) {
            break;
          }

          message.skybox = AboutResponse_SkyboxConfiguration.decode(reader, reader.uint32());
          continue;
        case 7:
          if (tag !== 58) {
            break;
          }

          message.cityLoaderContentServer = reader.string();
          continue;
        case 8:
          if (tag !== 66) {
            break;
          }

          message.map = AboutResponse_MapConfiguration.decode(reader, reader.uint32());
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): AboutResponse_AboutConfiguration {
    return {
      realmName: isSet(object.realmName) ? String(object.realmName) : undefined,
      networkId: isSet(object.networkId) ? Number(object.networkId) : 0,
      globalScenesUrn: Array.isArray(object?.globalScenesUrn) ? object.globalScenesUrn.map((e: any) => String(e)) : [],
      scenesUrn: Array.isArray(object?.scenesUrn) ? object.scenesUrn.map((e: any) => String(e)) : [],
      skybox: isSet(object.skybox) ? AboutResponse_SkyboxConfiguration.fromJSON(object.skybox) : undefined,
      cityLoaderContentServer: isSet(object.cityLoaderContentServer)
        ? String(object.cityLoaderContentServer)
        : undefined,
      map: isSet(object.map) ? AboutResponse_MapConfiguration.fromJSON(object.map) : undefined,
    };
  }

  export function toJSON(message: AboutResponse_AboutConfiguration): unknown {
    const obj: any = {};
    message.realmName !== undefined && (obj.realmName = message.realmName);
    message.networkId !== undefined && (obj.networkId = Math.round(message.networkId));
    if (message.globalScenesUrn) {
      obj.globalScenesUrn = message.globalScenesUrn.map((e) => e);
    } else {
      obj.globalScenesUrn = [];
    }
    if (message.scenesUrn) {
      obj.scenesUrn = message.scenesUrn.map((e) => e);
    } else {
      obj.scenesUrn = [];
    }
    message.skybox !== undefined &&
      (obj.skybox = message.skybox ? AboutResponse_SkyboxConfiguration.toJSON(message.skybox) : undefined);
    message.cityLoaderContentServer !== undefined && (obj.cityLoaderContentServer = message.cityLoaderContentServer);
    message.map !== undefined &&
      (obj.map = message.map ? AboutResponse_MapConfiguration.toJSON(message.map) : undefined);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<AboutResponse_AboutConfiguration>, I>>(
    base?: I,
  ): AboutResponse_AboutConfiguration {
    return AboutResponse_AboutConfiguration.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<AboutResponse_AboutConfiguration>, I>>(
    object: I,
  ): AboutResponse_AboutConfiguration {
    const message = createBaseAboutResponse_AboutConfiguration();
    message.realmName = object.realmName ?? undefined;
    message.networkId = object.networkId ?? 0;
    message.globalScenesUrn = object.globalScenesUrn?.map((e) => e) || [];
    message.scenesUrn = object.scenesUrn?.map((e) => e) || [];
    message.skybox = (object.skybox !== undefined && object.skybox !== null)
      ? AboutResponse_SkyboxConfiguration.fromPartial(object.skybox)
      : undefined;
    message.cityLoaderContentServer = object.cityLoaderContentServer ?? undefined;
    message.map = (object.map !== undefined && object.map !== null)
      ? AboutResponse_MapConfiguration.fromPartial(object.map)
      : undefined;
    return message;
  }
}

function createBaseAboutResponse_ContentInfo(): AboutResponse_ContentInfo {
  return { healthy: false, version: undefined, commitHash: undefined, publicUrl: "" };
}

export namespace AboutResponse_ContentInfo {
  export function encode(message: AboutResponse_ContentInfo, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.healthy === true) {
      writer.uint32(8).bool(message.healthy);
    }
    if (message.version !== undefined) {
      writer.uint32(18).string(message.version);
    }
    if (message.commitHash !== undefined) {
      writer.uint32(26).string(message.commitHash);
    }
    if (message.publicUrl !== "") {
      writer.uint32(34).string(message.publicUrl);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): AboutResponse_ContentInfo {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseAboutResponse_ContentInfo();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 8) {
            break;
          }

          message.healthy = reader.bool();
          continue;
        case 2:
          if (tag !== 18) {
            break;
          }

          message.version = reader.string();
          continue;
        case 3:
          if (tag !== 26) {
            break;
          }

          message.commitHash = reader.string();
          continue;
        case 4:
          if (tag !== 34) {
            break;
          }

          message.publicUrl = reader.string();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): AboutResponse_ContentInfo {
    return {
      healthy: isSet(object.healthy) ? Boolean(object.healthy) : false,
      version: isSet(object.version) ? String(object.version) : undefined,
      commitHash: isSet(object.commitHash) ? String(object.commitHash) : undefined,
      publicUrl: isSet(object.publicUrl) ? String(object.publicUrl) : "",
    };
  }

  export function toJSON(message: AboutResponse_ContentInfo): unknown {
    const obj: any = {};
    message.healthy !== undefined && (obj.healthy = message.healthy);
    message.version !== undefined && (obj.version = message.version);
    message.commitHash !== undefined && (obj.commitHash = message.commitHash);
    message.publicUrl !== undefined && (obj.publicUrl = message.publicUrl);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<AboutResponse_ContentInfo>, I>>(
    base?: I,
  ): AboutResponse_ContentInfo {
    return AboutResponse_ContentInfo.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<AboutResponse_ContentInfo>, I>>(
    object: I,
  ): AboutResponse_ContentInfo {
    const message = createBaseAboutResponse_ContentInfo();
    message.healthy = object.healthy ?? false;
    message.version = object.version ?? undefined;
    message.commitHash = object.commitHash ?? undefined;
    message.publicUrl = object.publicUrl ?? "";
    return message;
  }
}

function createBaseAboutResponse_LambdasInfo(): AboutResponse_LambdasInfo {
  return { healthy: false, version: undefined, commitHash: undefined, publicUrl: "" };
}

export namespace AboutResponse_LambdasInfo {
  export function encode(message: AboutResponse_LambdasInfo, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.healthy === true) {
      writer.uint32(8).bool(message.healthy);
    }
    if (message.version !== undefined) {
      writer.uint32(18).string(message.version);
    }
    if (message.commitHash !== undefined) {
      writer.uint32(26).string(message.commitHash);
    }
    if (message.publicUrl !== "") {
      writer.uint32(34).string(message.publicUrl);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): AboutResponse_LambdasInfo {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseAboutResponse_LambdasInfo();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 8) {
            break;
          }

          message.healthy = reader.bool();
          continue;
        case 2:
          if (tag !== 18) {
            break;
          }

          message.version = reader.string();
          continue;
        case 3:
          if (tag !== 26) {
            break;
          }

          message.commitHash = reader.string();
          continue;
        case 4:
          if (tag !== 34) {
            break;
          }

          message.publicUrl = reader.string();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): AboutResponse_LambdasInfo {
    return {
      healthy: isSet(object.healthy) ? Boolean(object.healthy) : false,
      version: isSet(object.version) ? String(object.version) : undefined,
      commitHash: isSet(object.commitHash) ? String(object.commitHash) : undefined,
      publicUrl: isSet(object.publicUrl) ? String(object.publicUrl) : "",
    };
  }

  export function toJSON(message: AboutResponse_LambdasInfo): unknown {
    const obj: any = {};
    message.healthy !== undefined && (obj.healthy = message.healthy);
    message.version !== undefined && (obj.version = message.version);
    message.commitHash !== undefined && (obj.commitHash = message.commitHash);
    message.publicUrl !== undefined && (obj.publicUrl = message.publicUrl);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<AboutResponse_LambdasInfo>, I>>(
    base?: I,
  ): AboutResponse_LambdasInfo {
    return AboutResponse_LambdasInfo.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<AboutResponse_LambdasInfo>, I>>(
    object: I,
  ): AboutResponse_LambdasInfo {
    const message = createBaseAboutResponse_LambdasInfo();
    message.healthy = object.healthy ?? false;
    message.version = object.version ?? undefined;
    message.commitHash = object.commitHash ?? undefined;
    message.publicUrl = object.publicUrl ?? "";
    return message;
  }
}

function createBaseAboutResponse_CommsInfo(): AboutResponse_CommsInfo {
  return {
    healthy: false,
    version: undefined,
    commitHash: undefined,
    publicUrl: undefined,
    protocol: "",
    usersCount: undefined,
    fixedAdapter: undefined,
  };
}

export namespace AboutResponse_CommsInfo {
  export function encode(message: AboutResponse_CommsInfo, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.healthy === true) {
      writer.uint32(8).bool(message.healthy);
    }
    if (message.version !== undefined) {
      writer.uint32(18).string(message.version);
    }
    if (message.commitHash !== undefined) {
      writer.uint32(26).string(message.commitHash);
    }
    if (message.publicUrl !== undefined) {
      writer.uint32(34).string(message.publicUrl);
    }
    if (message.protocol !== "") {
      writer.uint32(402).string(message.protocol);
    }
    if (message.usersCount !== undefined) {
      writer.uint32(408).int32(message.usersCount);
    }
    if (message.fixedAdapter !== undefined) {
      writer.uint32(418).string(message.fixedAdapter);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): AboutResponse_CommsInfo {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseAboutResponse_CommsInfo();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 8) {
            break;
          }

          message.healthy = reader.bool();
          continue;
        case 2:
          if (tag !== 18) {
            break;
          }

          message.version = reader.string();
          continue;
        case 3:
          if (tag !== 26) {
            break;
          }

          message.commitHash = reader.string();
          continue;
        case 4:
          if (tag !== 34) {
            break;
          }

          message.publicUrl = reader.string();
          continue;
        case 50:
          if (tag !== 402) {
            break;
          }

          message.protocol = reader.string();
          continue;
        case 51:
          if (tag !== 408) {
            break;
          }

          message.usersCount = reader.int32();
          continue;
        case 52:
          if (tag !== 418) {
            break;
          }

          message.fixedAdapter = reader.string();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): AboutResponse_CommsInfo {
    return {
      healthy: isSet(object.healthy) ? Boolean(object.healthy) : false,
      version: isSet(object.version) ? String(object.version) : undefined,
      commitHash: isSet(object.commitHash) ? String(object.commitHash) : undefined,
      publicUrl: isSet(object.publicUrl) ? String(object.publicUrl) : undefined,
      protocol: isSet(object.protocol) ? String(object.protocol) : "",
      usersCount: isSet(object.usersCount) ? Number(object.usersCount) : undefined,
      fixedAdapter: isSet(object.fixedAdapter) ? String(object.fixedAdapter) : undefined,
    };
  }

  export function toJSON(message: AboutResponse_CommsInfo): unknown {
    const obj: any = {};
    message.healthy !== undefined && (obj.healthy = message.healthy);
    message.version !== undefined && (obj.version = message.version);
    message.commitHash !== undefined && (obj.commitHash = message.commitHash);
    message.publicUrl !== undefined && (obj.publicUrl = message.publicUrl);
    message.protocol !== undefined && (obj.protocol = message.protocol);
    message.usersCount !== undefined && (obj.usersCount = Math.round(message.usersCount));
    message.fixedAdapter !== undefined && (obj.fixedAdapter = message.fixedAdapter);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<AboutResponse_CommsInfo>, I>>(base?: I): AboutResponse_CommsInfo {
    return AboutResponse_CommsInfo.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<AboutResponse_CommsInfo>, I>>(
    object: I,
  ): AboutResponse_CommsInfo {
    const message = createBaseAboutResponse_CommsInfo();
    message.healthy = object.healthy ?? false;
    message.version = object.version ?? undefined;
    message.commitHash = object.commitHash ?? undefined;
    message.publicUrl = object.publicUrl ?? undefined;
    message.protocol = object.protocol ?? "";
    message.usersCount = object.usersCount ?? undefined;
    message.fixedAdapter = object.fixedAdapter ?? undefined;
    return message;
  }
}

function createBaseAboutResponse_BffInfo(): AboutResponse_BffInfo {
  return {
    healthy: false,
    version: undefined,
    commitHash: undefined,
    publicUrl: "",
    userCount: undefined,
    protocolVersion: undefined,
  };
}

export namespace AboutResponse_BffInfo {
  export function encode(message: AboutResponse_BffInfo, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.healthy === true) {
      writer.uint32(8).bool(message.healthy);
    }
    if (message.version !== undefined) {
      writer.uint32(18).string(message.version);
    }
    if (message.commitHash !== undefined) {
      writer.uint32(26).string(message.commitHash);
    }
    if (message.publicUrl !== "") {
      writer.uint32(34).string(message.publicUrl);
    }
    if (message.userCount !== undefined) {
      writer.uint32(408).int32(message.userCount);
    }
    if (message.protocolVersion !== undefined) {
      writer.uint32(418).string(message.protocolVersion);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): AboutResponse_BffInfo {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseAboutResponse_BffInfo();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 8) {
            break;
          }

          message.healthy = reader.bool();
          continue;
        case 2:
          if (tag !== 18) {
            break;
          }

          message.version = reader.string();
          continue;
        case 3:
          if (tag !== 26) {
            break;
          }

          message.commitHash = reader.string();
          continue;
        case 4:
          if (tag !== 34) {
            break;
          }

          message.publicUrl = reader.string();
          continue;
        case 51:
          if (tag !== 408) {
            break;
          }

          message.userCount = reader.int32();
          continue;
        case 52:
          if (tag !== 418) {
            break;
          }

          message.protocolVersion = reader.string();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): AboutResponse_BffInfo {
    return {
      healthy: isSet(object.healthy) ? Boolean(object.healthy) : false,
      version: isSet(object.version) ? String(object.version) : undefined,
      commitHash: isSet(object.commitHash) ? String(object.commitHash) : undefined,
      publicUrl: isSet(object.publicUrl) ? String(object.publicUrl) : "",
      userCount: isSet(object.userCount) ? Number(object.userCount) : undefined,
      protocolVersion: isSet(object.protocolVersion) ? String(object.protocolVersion) : undefined,
    };
  }

  export function toJSON(message: AboutResponse_BffInfo): unknown {
    const obj: any = {};
    message.healthy !== undefined && (obj.healthy = message.healthy);
    message.version !== undefined && (obj.version = message.version);
    message.commitHash !== undefined && (obj.commitHash = message.commitHash);
    message.publicUrl !== undefined && (obj.publicUrl = message.publicUrl);
    message.userCount !== undefined && (obj.userCount = Math.round(message.userCount));
    message.protocolVersion !== undefined && (obj.protocolVersion = message.protocolVersion);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<AboutResponse_BffInfo>, I>>(base?: I): AboutResponse_BffInfo {
    return AboutResponse_BffInfo.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<AboutResponse_BffInfo>, I>>(
    object: I,
  ): AboutResponse_BffInfo {
    const message = createBaseAboutResponse_BffInfo();
    message.healthy = object.healthy ?? false;
    message.version = object.version ?? undefined;
    message.commitHash = object.commitHash ?? undefined;
    message.publicUrl = object.publicUrl ?? "";
    message.userCount = object.userCount ?? undefined;
    message.protocolVersion = object.protocolVersion ?? undefined;
    return message;
  }
}

type Builtin = Date | Function | Uint8Array | string | number | boolean | undefined;

export type DeepPartial<T> = T extends Builtin ? T
  : T extends Array<infer U> ? Array<DeepPartial<U>> : T extends ReadonlyArray<infer U> ? ReadonlyArray<DeepPartial<U>>
  : T extends { $case: string } ? { [K in keyof Omit<T, "$case">]?: DeepPartial<T[K]> } & { $case: T["$case"] }
  : T extends {} ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & { [K in Exclude<keyof I, KeysOfUnion<P>>]: never };

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
