// @dcl/sdk-commands/dist/commands/init/project.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPxSceneJson = exports.getMinimalSceneJson = void 0;
const path_1 = __importDefault(require("path"));
const scene_validations_1 = require("../../logic/scene-validations");
function getMinimalSceneJson() {
    return {
        ecs7: true,
        runtimeVersion: '7',
        display: {
            title: 'SDK7 Scene Template',
            description: 'template scene with SDK7'
        },
        scene: {
            parcels: ['0,0'],
            base: '0,0'
        },
        main: 'bin/index.js'
    };
}
exports.getMinimalSceneJson = getMinimalSceneJson;
async function createPxSceneJson(dir, fs) {
    const scene = {
        ...getMinimalSceneJson(),
        isPortableExperience: true,
        display: {
            title: 'SDK7 Portable Experience Scene Template',
            description: 'portable experience template scene with SDK7'
        }
    };
    await fs.writeFile(path_1.default.resolve(dir, scene_validations_1.SCENE_FILE), JSON.stringify(scene, null, 2), 'utf-8');
}
exports.createPxSceneJson = createPxSceneJson;
