// @dcl/protocol/out-js/decentraland/sdk/components/common/camera_type.gen.js
"use strict";
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CameraType = exports.protobufPackage = void 0;
exports.cameraTypeFromJSON = cameraTypeFromJSON;
exports.cameraTypeToJSON = cameraTypeToJSON;
exports.protobufPackage = "decentraland.sdk.components.common";
var CameraType;
(function (CameraType) {
    CameraType[CameraType["CT_FIRST_PERSON"] = 0] = "CT_FIRST_PERSON";
    CameraType[CameraType["CT_THIRD_PERSON"] = 1] = "CT_THIRD_PERSON";
    /** CT_CINEMATIC - controlled by the scene */
    CameraType[CameraType["CT_CINEMATIC"] = 2] = "CT_CINEMATIC";
    CameraType[CameraType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(CameraType || (exports.CameraType = CameraType = {}));
function cameraTypeFromJSON(object) {
    switch (object) {
        case 0:
        case "CT_FIRST_PERSON":
            return CameraType.CT_FIRST_PERSON;
        case 1:
        case "CT_THIRD_PERSON":
            return CameraType.CT_THIRD_PERSON;
        case 2:
        case "CT_CINEMATIC":
            return CameraType.CT_CINEMATIC;
        case -1:
        case "UNRECOGNIZED":
        default:
            return CameraType.UNRECOGNIZED;
    }
}
function cameraTypeToJSON(object) {
    switch (object) {
        case CameraType.CT_FIRST_PERSON:
            return "CT_FIRST_PERSON";
        case CameraType.CT_THIRD_PERSON:
            return "CT_THIRD_PERSON";
        case CameraType.CT_CINEMATIC:
            return "CT_CINEMATIC";
        case CameraType.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
