// @dcl/protocol/out-js/decentraland/sdk/components/common/loading_state.gen.js
"use strict";
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoadingState = exports.protobufPackage = void 0;
exports.loadingStateFromJSON = loadingStateFromJSON;
exports.loadingStateToJSON = loadingStateToJSON;
exports.protobufPackage = "decentraland.sdk.components.common";
var LoadingState;
(function (LoadingState) {
    LoadingState[LoadingState["UNKNOWN"] = 0] = "UNKNOWN";
    LoadingState[LoadingState["LOADING"] = 1] = "LOADING";
    LoadingState[LoadingState["NOT_FOUND"] = 2] = "NOT_FOUND";
    LoadingState[LoadingState["FINISHED_WITH_ERROR"] = 3] = "FINISHED_WITH_ERROR";
    LoadingState[LoadingState["FINISHED"] = 4] = "FINISHED";
    LoadingState[LoadingState["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(LoadingState || (exports.LoadingState = LoadingState = {}));
function loadingStateFromJSON(object) {
    switch (object) {
        case 0:
        case "UNKNOWN":
            return LoadingState.UNKNOWN;
        case 1:
        case "LOADING":
            return LoadingState.LOADING;
        case 2:
        case "NOT_FOUND":
            return LoadingState.NOT_FOUND;
        case 3:
        case "FINISHED_WITH_ERROR":
            return LoadingState.FINISHED_WITH_ERROR;
        case 4:
        case "FINISHED":
            return LoadingState.FINISHED;
        case -1:
        case "UNRECOGNIZED":
        default:
            return LoadingState.UNRECOGNIZED;
    }
}
function loadingStateToJSON(object) {
    switch (object) {
        case LoadingState.UNKNOWN:
            return "UNKNOWN";
        case LoadingState.LOADING:
            return "LOADING";
        case LoadingState.NOT_FOUND:
            return "NOT_FOUND";
        case LoadingState.FINISHED_WITH_ERROR:
            return "FINISHED_WITH_ERROR";
        case LoadingState.FINISHED:
            return "FINISHED";
        case LoadingState.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
