// @dcl/schemas/dist/platform/notifications/notifications.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotificationType = void 0;
/**
 * Notification Types (the keys as stored in the notifications DB)
 * @alpha
 */
var NotificationType;
(function (NotificationType) {
    NotificationType["BID_ACCEPTED"] = "bid_accepted";
    NotificationType["BID_RECEIVED"] = "bid_received";
    NotificationType["EVENTS_STARTED"] = "events_started";
    NotificationType["EVENTS_STARTS_SOON"] = "events_starts_soon";
    NotificationType["GOVERNANCE_ANNOUNCEMENT"] = "governance_announcement";
    NotificationType["GOVERNANCE_AUTHORED_PROPOSAL_FINISHED"] = "governance_authored_proposal_finished";
    NotificationType["GOVERNANCE_COAUTHOR_REQUESTED"] = "governance_coauthor_requested";
    NotificationType["GOVERNANCE_NEW_COMMENT_ON_PROJECT_UPDATE"] = "governance_new_comment_on_project_update";
    NotificationType["GOVERNANCE_NEW_COMMENT_ON_PROPOSAL"] = "governance_new_comment_on_proposal";
    NotificationType["GOVERNANCE_PROPOSAL_ENACTED"] = "governance_proposal_enacted";
    NotificationType["GOVERNANCE_VOTING_ENDED_VOTER"] = "governance_voting_ended_voter";
    NotificationType["GOVERNANCE_PITCH_PASSED"] = "governance_pitch_passed";
    NotificationType["GOVERNANCE_TENDER_PASSED"] = "governance_tender_passed";
    NotificationType["ITEM_SOLD"] = "item_sold";
    NotificationType["LAND_RENTAL_ENDED"] = "rental_ended";
    NotificationType["LAND_RENTED"] = "rental_started";
    NotificationType["REWARD_ASSIGNED"] = "reward_assignment";
    NotificationType["ROYALTIES_EARNED"] = "royalties_earned";
    NotificationType["WORLDS_ACCESS_RESTORED"] = "worlds_access_restored";
    NotificationType["WORLDS_ACCESS_RESTRICTED"] = "worlds_access_restricted";
    NotificationType["WORLDS_MISSING_RESOURCES"] = "worlds_missing_resources";
    NotificationType["WORLDS_PERMISSION_GRANTED"] = "worlds_permission_granted";
    NotificationType["WORLDS_PERMISSION_REVOKED"] = "worlds_permission_revoked";
    NotificationType["BADGE_GRANTED"] = "badge_granted";
})(NotificationType = exports.NotificationType || (exports.NotificationType = {}));
