// @dcl/asset-packs/dist/action-types.js
import { Schemas } from '@dcl/ecs';
import { ComponentName, getComponent, } from './definitions';
export const EMPTY = {
    type: 'object',
    properties: {},
    serializationType: 'map',
};
export function addActionType(engine, type, schema) {
    const ActionTypes = getComponent(ComponentName.ACTION_TYPES, engine);
    const actionTypes = ActionTypes.getOrCreateMutable(engine.RootEntity);
    const actionType = {
        type,
        jsonSchema: JSON.stringify(schema?.jsonSchema || Schemas.Map({}).jsonSchema),
    };
    actionTypes.value = [
        ...actionTypes.value.filter(($) => $.type !== actionType.type),
        actionType,
    ];
}
export function getActionSchema(engine, type) {
    const ActionTypes = getComponent(ComponentName.ACTION_TYPES, engine);
    const actionTypes = ActionTypes.get(engine.RootEntity);
    const actionType = actionTypes.value.find(($) => $.type === type);
    const jsonSchema = actionType
        ? JSON.parse(actionType.jsonSchema)
        : EMPTY;
    return Schemas.fromJson(jsonSchema);
}
export function getActionTypes(engine) {
    const ActionTypes = getComponent(ComponentName.ACTION_TYPES, engine);
    const actionTypes = ActionTypes.get(engine.RootEntity);
    return actionTypes.value.map(($) => $.type);
}
export function getPayload(action) {
    return JSON.parse(action.jsonPayload);
}
export function getJson(payload) {
    return JSON.stringify(payload);
}
