// @dcl/ecs/dist-cjs/components/generated/pb/decentraland/sdk/components/common/texts.gen.js
"use strict";
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Font = exports.TextAlignMode = void 0;
const protobufPackageSarasa = "decentraland.sdk.components.common";
/**
 * @public
 */
var TextAlignMode;
(function (TextAlignMode) {
    TextAlignMode[TextAlignMode["TAM_TOP_LEFT"] = 0] = "TAM_TOP_LEFT";
    TextAlignMode[TextAlignMode["TAM_TOP_CENTER"] = 1] = "TAM_TOP_CENTER";
    TextAlignMode[TextAlignMode["TAM_TOP_RIGHT"] = 2] = "TAM_TOP_RIGHT";
    TextAlignMode[TextAlignMode["TAM_MIDDLE_LEFT"] = 3] = "TAM_MIDDLE_LEFT";
    TextAlignMode[TextAlignMode["TAM_MIDDLE_CENTER"] = 4] = "TAM_MIDDLE_CENTER";
    TextAlignMode[TextAlignMode["TAM_MIDDLE_RIGHT"] = 5] = "TAM_MIDDLE_RIGHT";
    TextAlignMode[TextAlignMode["TAM_BOTTOM_LEFT"] = 6] = "TAM_BOTTOM_LEFT";
    TextAlignMode[TextAlignMode["TAM_BOTTOM_CENTER"] = 7] = "TAM_BOTTOM_CENTER";
    TextAlignMode[TextAlignMode["TAM_BOTTOM_RIGHT"] = 8] = "TAM_BOTTOM_RIGHT";
})(TextAlignMode = exports.TextAlignMode || (exports.TextAlignMode = {}));
/**
 * @public
 */
var Font;
(function (Font) {
    Font[Font["F_SANS_SERIF"] = 0] = "F_SANS_SERIF";
    Font[Font["F_SERIF"] = 1] = "F_SERIF";
    Font[Font["F_MONOSPACE"] = 2] = "F_MONOSPACE";
})(Font = exports.Font || (exports.Font = {}));
