// @dcl/ecs/dist-cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkParent = exports.NetworkEntity = exports.SyncComponents = exports.InputModifier = exports.VirtualCamera = exports.Tween = exports.Name = exports.MeshCollider = exports.MeshRenderer = exports.Material = exports.AudioStream = exports.AudioSource = exports.Animator = exports.Transform = exports.components = exports.cyclicParentingChecker = void 0;
// The order of the following imports matters. Please do not auto-sort
__exportStar(require("./engine"), exports);
__exportStar(require("./schemas"), exports);
__exportStar(require("./runtime/initialization"), exports);
__exportStar(require("./runtime/types"), exports);
__exportStar(require("./runtime/helpers"), exports);
var cyclicParentingChecker_1 = require("./systems/cyclicParentingChecker");
Object.defineProperty(exports, "cyclicParentingChecker", { enumerable: true, get: function () { return cyclicParentingChecker_1.cyclicParentingChecker; } });
__exportStar(require("./systems/events"), exports);
__exportStar(require("./systems/raycast"), exports);
__exportStar(require("./systems/videoEvents"), exports);
__exportStar(require("./systems/async-task"), exports);
__exportStar(require("./systems/tween"), exports);
__exportStar(require("./engine/entity"), exports);
__exportStar(require("./components/types"), exports);
// @internal
const components = __importStar(require("./components"));
exports.components = components;
const initialization_1 = require("./runtime/initialization");
// export components for global engine
exports.Transform = components.Transform(initialization_1.engine);
exports.Animator = components.Animator(initialization_1.engine);
exports.AudioSource = components.AudioSource(initialization_1.engine);
exports.AudioStream = components.AudioStream(initialization_1.engine);
exports.Material = components.Material(initialization_1.engine);
exports.MeshRenderer = components.MeshRenderer(initialization_1.engine);
exports.MeshCollider = components.MeshCollider(initialization_1.engine);
exports.Name = components.Name(initialization_1.engine);
exports.Tween = components.Tween(initialization_1.engine);
exports.VirtualCamera = components.VirtualCamera(initialization_1.engine);
exports.InputModifier = components.InputModifier(initialization_1.engine);
/**
 * @alpha
 * This is going to be used for sync components through a server.
 * Can be modified in the future since it's still in research
 */
exports.SyncComponents = components.SyncComponents(initialization_1.engine);
/**
 * @alpha
 * Tag a entity to be syncronized through comms
 */
exports.NetworkEntity = components.NetworkEntity(initialization_1.engine);
/**
 * @alpha
 * Tag a entity to be syncronized through comms
 */
exports.NetworkParent = components.NetworkParent(initialization_1.engine);
// export components for global engine
__exportStar(require("./components/generated/global.gen"), exports);
__exportStar(require("./components/generated/types.gen"), exports);
__exportStar(require("./serialization/crdt"), exports);
__exportStar(require("./composite"), exports);
