// @dcl-sdk/utils/dist/debug.js
import { Transform, TextShape, Billboard, MeshRenderer, MeshCollider, Material, pointerEventsSystem, engine } from '@dcl/sdk/ecs';
import { Vector3, Color4 } from '@dcl/sdk/math';
export function addLabel(text, parent, billboard, color, size, textOffset) {
    let label = engine.addEntity();
    Transform.create(label, {
        position: textOffset ? textOffset : Vector3.create(0, 1.5, 0),
        parent: parent
    });
    let textShape = TextShape.create(label);
    textShape.text = text;
    textShape.fontSize = size ? size : 3;
    textShape.textColor = color ? color : Color4.Black();
    if (billboard) {
        Billboard.create(label);
    }
    return label;
}
export function addTestCube(transform, triggeredFunction, label, color, sphere, noCollider) {
    let cube = engine.addEntity();
    Transform.create(cube, transform);
    if (sphere) {
        let sphereTransform = Transform.getMutable(cube);
        sphereTransform.scale = Vector3.multiplyByFloats(sphereTransform.scale, 0.5, 0.5, 0.5);
        MeshRenderer.setSphere(cube);
        if (!noCollider)
            MeshCollider.setSphere(cube);
    }
    else {
        MeshRenderer.setBox(cube);
        if (!noCollider)
            MeshCollider.setBox(cube);
    }
    if (color) {
        Material.setPbrMaterial(cube, { albedoColor: color });
    }
    if (label) {
        addLabel(label, cube, true);
    }
    if (triggeredFunction) {
        pointerEventsSystem.onPointerDown(cube, triggeredFunction, {
            button: 0,
            hoverText: label ? label : 'click'
        });
    }
    return cube;
}
