// @dcl-sdk/utils/dist/timer.js
import { engine } from '@dcl/sdk/ecs';
import { priority } from './priority';
function createTimers(targetEngine) {
    const timers = new Map();
    let timerIdCounter = 0;
    function system(dt) {
        let deadTimers = [];
        let callbacks = [];
        for (let [timerId, timerData] of timers) {
            timerData.accumulatedTime += 1000 * dt;
            if (timerData.accumulatedTime < timerData.interval)
                continue;
            callbacks.push(timerData.callback);
            if (timerData.recurrent) {
                timerData.accumulatedTime -= Math.floor(timerData.accumulatedTime / timerData.interval) * timerData.interval;
            }
            else {
                deadTimers.push(timerId);
            }
        }
        for (let timerId of deadTimers)
            timers.delete(timerId);
        for (let callback of callbacks)
            callback();
    }
    targetEngine.addSystem(system, priority.TimerSystemPriority);
    return {
        setTimeout(callback, milliseconds) {
            let timerId = timerIdCounter++;
            timers.set(timerId, { callback: callback, interval: milliseconds, recurrent: false, accumulatedTime: 0 });
            return timerId;
        },
        clearTimeout(timer) {
            timers.delete(timer);
        },
        setInterval(callback, milliseconds) {
            let timerId = timerIdCounter++;
            timers.set(timerId, { callback: callback, interval: milliseconds, recurrent: true, accumulatedTime: 0 });
            return timerId;
        },
        clearInterval(timer) {
            timers.delete(timer);
        }
    };
}
export const timers = createTimers(engine);
