// @dcl-sdk/utils/dist/toggle.js
import { engine, EntityState, Schemas } from '@dcl/sdk/ecs';
import { timers } from './timer';
export var ToggleState;
(function (ToggleState) {
    ToggleState[ToggleState["Off"] = 0] = "Off";
    ToggleState[ToggleState["On"] = 1] = "On";
})(ToggleState || (ToggleState = {}));
function createToggles(targetEngine, timers) {
    const Toggle = targetEngine.defineComponent('dcl.utils.Toggle', {
        state: Schemas.EnumNumber(ToggleState, ToggleState.Off)
    });
    let toggles = new Map();
    timers.setInterval(function () {
        for (const entity of toggles.keys()) {
            if (targetEngine.getEntityState(entity) == EntityState.Removed || !Toggle.has(entity)) {
                toggles.delete(entity);
            }
        }
    }, 5000);
    return {
        addToggle(entity, state, callback) {
            toggles.set(entity, callback);
            Toggle.createOrReplace(entity, { state: state });
        },
        removeToggle(entity) {
            toggles.delete(entity);
            Toggle.deleteFrom(entity);
        },
        setCallback(entity, callback) {
            toggles.set(entity, callback);
        },
        set(entity, state) {
            const oldState = Toggle.get(entity).state;
            if (oldState != state) {
                Toggle.getMutable(entity).state = state;
                const callback = toggles.get(entity);
                if (callback)
                    callback(state);
            }
        },
        flip(entity) {
            this.set(entity, 1 - Toggle.get(entity).state);
        },
        isOn(entity) {
            return Toggle.get(entity).state == ToggleState.On;
        }
    };
}
export const toggles = createToggles(engine, timers);
