// @dcl-sdk/utils/dist/tween.js
import { engine, EntityState, Tween, } from "@dcl/sdk/ecs";
import { priority } from "./priority";
import { InterpolationType } from "./math";
import { getEasingFunctionFromInterpolation } from "./helpers";
function createTweens(targetEngine) {
    const tweenMap = new Map();
    function makeSystem(dt) {
        const deadTweens = [];
        for (const [entity, tweenData] of tweenMap.entries()) {
            if (targetEngine.getEntityState(entity) == EntityState.Removed ||
                !Tween.has(entity)) {
                tweenMap.delete(entity);
                continue;
            }
            const tween = Tween.get(entity);
            tweenData.normalizedTime += dt * 1000;
            if (tweenData.normalizedTime >= tween.duration) {
                deadTweens.push(entity);
            }
        }
        for (const entity of deadTweens) {
            const callback = tweenMap.get(entity)?.callback;
            Tween.deleteFrom(entity);
            tweenMap.delete(entity);
            if (callback)
                callback();
        }
    }
    function makeStop(entity) {
        Tween.deleteFrom(entity);
        tweenMap.delete(entity);
    }
    function makeStart(mode) {
        return function (entity, start, end, duration, interpolationType = InterpolationType.LINEAR, onFinish) {
            const currentTime = duration === 0 ? 1 : 0;
            tweenMap.set(entity, { normalizedTime: currentTime, callback: onFinish });
            Tween.createOrReplace(entity, {
                duration: duration * 1000,
                easingFunction: getEasingFunctionFromInterpolation(interpolationType),
                currentTime,
                mode: Tween.Mode[mode]({ start: start, end: end }),
            });
        };
    }
    function makeGetOnFinishCallback(entity) {
        if (!tweenMap.has(entity)) {
            throw new Error(`Entity ${entity} is not registered with tweens system`);
        }
        return tweenMap.get(entity);
    }
    targetEngine.addSystem(makeSystem, priority.TweenSystemPriority);
    return {
        startTranslation: makeStart("Move"),
        stopTranslation: makeStop,
        startRotation: makeStart("Rotate"),
        stopRotation: makeStop,
        startScaling: makeStart("Scale"),
        stopScaling: makeStop,
        getTranslationOnFinishCallback: makeGetOnFinishCallback,
        getRotationOnFinishCallback: makeGetOnFinishCallback,
        getScalingOnFinishCallback: makeGetOnFinishCallback,
    };
}
export const tweens = createTweens(engine);
