// @dcl/sdk/internal/transports/rendererTransport.js
import { MAX_STATIC_COMPONENT } from '@dcl/ecs/dist/components/component-number';
export function createRendererTransport(engineApi) {
    async function sendToRenderer(message) {
        const response = await engineApi.crdtSendToRenderer({
            data: new Uint8Array(message)
        });
        if (response && response.data && response.data.length) {
            if (rendererTransport.onmessage) {
                for (const byteArray of response.data) {
                    rendererTransport.onmessage(byteArray);
                }
            }
        }
    }
    const rendererTransport = {
        async send(message) {
            try {
                await sendToRenderer(message);
            }
            catch (error) {
                // this is the console.error of the scene
                // eslint-disable-next-line no-console
                console.error(error);
                debugger;
            }
        },
        filter(message) {
            // Only send renderer components (Proto Generated)
            if (
            // filter out messages for non-core components
            message.componentId > MAX_STATIC_COMPONENT) {
                return false;
            }
            return !!message;
        },
        type: 'renderer'
    };
    return rendererTransport;
}
