// @dcl/ecs/dist/components/generated/global.gen.js
import { engine } from '../../runtime/initialization';
import * as components from './index.gen';
export * from './index.gen';
/** @public */ export const AudioEvent = /* @__PURE__ */ components.AudioEvent(engine);
/** @public */ export const AudioSource = /* @__PURE__ */ components.AudioSource(engine);
/** @public */ export const AudioStream = /* @__PURE__ */ components.AudioStream(engine);
/** @public */ export const AvatarAttach = /* @__PURE__ */ components.AvatarAttach(engine);
/** @public */ export const AvatarBase = /* @__PURE__ */ components.AvatarBase(engine);
/** @public */ export const AvatarEmoteCommand = /* @__PURE__ */ components.AvatarEmoteCommand(engine);
/** @public */ export const AvatarEquippedData = /* @__PURE__ */ components.AvatarEquippedData(engine);
/** @public */ export const AvatarModifierArea = /* @__PURE__ */ components.AvatarModifierArea(engine);
/** @public */ export const AvatarShape = /* @__PURE__ */ components.AvatarShape(engine);
/** @public */ export const Billboard = /* @__PURE__ */ components.Billboard(engine);
/** @public */ export const CameraMode = /* @__PURE__ */ components.CameraMode(engine);
/** @public */ export const CameraModeArea = /* @__PURE__ */ components.CameraModeArea(engine);
/** @public */ export const EngineInfo = /* @__PURE__ */ components.EngineInfo(engine);
/** @public */ export const GltfContainer = /* @__PURE__ */ components.GltfContainer(engine);
/** @public */ export const GltfContainerLoadingState = /* @__PURE__ */ components.GltfContainerLoadingState(engine);
/** @public */ export const InputModifier = /* @__PURE__ */ components.InputModifier(engine);
/** @public */ export const MainCamera = /* @__PURE__ */ components.MainCamera(engine);
/** @public */ export const NftShape = /* @__PURE__ */ components.NftShape(engine);
/** @public */ export const PlayerIdentityData = /* @__PURE__ */ components.PlayerIdentityData(engine);
/** @public */ export const PointerEvents = /* @__PURE__ */ components.PointerEvents(engine);
/** @public */ export const PointerEventsResult = /* @__PURE__ */ components.PointerEventsResult(engine);
/** @public */ export const PointerLock = /* @__PURE__ */ components.PointerLock(engine);
/** @public */ export const Raycast = /* @__PURE__ */ components.Raycast(engine);
/** @public */ export const RaycastResult = /* @__PURE__ */ components.RaycastResult(engine);
/** @public */ export const RealmInfo = /* @__PURE__ */ components.RealmInfo(engine);
/** @public */ export const TextShape = /* @__PURE__ */ components.TextShape(engine);
/** @public */ export const TweenSequence = /* @__PURE__ */ components.TweenSequence(engine);
/** @public */ export const TweenState = /* @__PURE__ */ components.TweenState(engine);
/** @public */ export const UiBackground = /* @__PURE__ */ components.UiBackground(engine);
/** @public */ export const UiCanvasInformation = /* @__PURE__ */ components.UiCanvasInformation(engine);
/** @public */ export const UiDropdown = /* @__PURE__ */ components.UiDropdown(engine);
/** @public */ export const UiDropdownResult = /* @__PURE__ */ components.UiDropdownResult(engine);
/** @public */ export const UiInput = /* @__PURE__ */ components.UiInput(engine);
/** @public */ export const UiInputResult = /* @__PURE__ */ components.UiInputResult(engine);
/** @public */ export const UiText = /* @__PURE__ */ components.UiText(engine);
/** @public */ export const UiTransform = /* @__PURE__ */ components.UiTransform(engine);
/** @public */ export const VideoEvent = /* @__PURE__ */ components.VideoEvent(engine);
/** @public */ export const VideoPlayer = /* @__PURE__ */ components.VideoPlayer(engine);
/** @public */ export const VirtualCamera = /* @__PURE__ */ components.VirtualCamera(engine);
/** @public */ export const VisibilityComponent = /* @__PURE__ */ components.VisibilityComponent(engine);
