// @dcl/protocol/out-ts/decentraland/kernel/apis/dev_tools.gen.ts
/* eslint-disable */
import _m0 from "protobufjs/minimal";

export const protobufPackage = "decentraland.kernel.apis";

export interface DevToolsBody {
  type: string;
  jsonPayload: string;
}

export interface EventResponse {
}

function createBaseDevToolsBody(): DevToolsBody {
  return { type: "", jsonPayload: "" };
}

export namespace DevToolsBody {
  export function encode(message: DevToolsBody, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.type !== "") {
      writer.uint32(10).string(message.type);
    }
    if (message.jsonPayload !== "") {
      writer.uint32(18).string(message.jsonPayload);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): DevToolsBody {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseDevToolsBody();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.type = reader.string();
          continue;
        case 2:
          if (tag !== 18) {
            break;
          }

          message.jsonPayload = reader.string();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): DevToolsBody {
    return {
      type: isSet(object.type) ? String(object.type) : "",
      jsonPayload: isSet(object.jsonPayload) ? String(object.jsonPayload) : "",
    };
  }

  export function toJSON(message: DevToolsBody): unknown {
    const obj: any = {};
    message.type !== undefined && (obj.type = message.type);
    message.jsonPayload !== undefined && (obj.jsonPayload = message.jsonPayload);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<DevToolsBody>, I>>(base?: I): DevToolsBody {
    return DevToolsBody.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<DevToolsBody>, I>>(object: I): DevToolsBody {
    const message = createBaseDevToolsBody();
    message.type = object.type ?? "";
    message.jsonPayload = object.jsonPayload ?? "";
    return message;
  }
}

function createBaseEventResponse(): EventResponse {
  return {};
}

export namespace EventResponse {
  export function encode(_: EventResponse, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): EventResponse {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseEventResponse();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(_: any): EventResponse {
    return {};
  }

  export function toJSON(_: EventResponse): unknown {
    const obj: any = {};
    return obj;
  }

  export function create<I extends Exact<DeepPartial<EventResponse>, I>>(base?: I): EventResponse {
    return EventResponse.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<EventResponse>, I>>(_: I): EventResponse {
    const message = createBaseEventResponse();
    return message;
  }
}

/** @deprecated */
export type DevToolsServiceDefinition = typeof DevToolsServiceDefinition;
export const DevToolsServiceDefinition = {
  name: "DevToolsService",
  fullName: "decentraland.kernel.apis.DevToolsService",
  methods: {
    /** @deprecated */
    event: {
      name: "Event",
      requestType: DevToolsBody,
      requestStream: false,
      responseType: EventResponse,
      responseStream: false,
      options: {},
    },
  },
} as const;

type Builtin = Date | Function | Uint8Array | string | number | boolean | undefined;

export type DeepPartial<T> = T extends Builtin ? T
  : T extends Array<infer U> ? Array<DeepPartial<U>> : T extends ReadonlyArray<infer U> ? ReadonlyArray<DeepPartial<U>>
  : T extends { $case: string } ? { [K in keyof Omit<T, "$case">]?: DeepPartial<T[K]> } & { $case: T["$case"] }
  : T extends {} ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & { [K in Exclude<keyof I, KeysOfUnion<P>>]: never };

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
