// @dcl/ecs/dist-cjs/components/generated/index.gen.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.componentDefinitionByName = exports.VisibilityComponent = exports.VirtualCamera = exports.VideoPlayer = exports.VideoEvent = exports.UiTransform = exports.UiText = exports.UiInputResult = exports.UiInput = exports.UiDropdownResult = exports.UiDropdown = exports.UiCanvasInformation = exports.UiBackground = exports.TweenState = exports.TweenSequence = exports.Tween = exports.TextShape = exports.RealmInfo = exports.RaycastResult = exports.Raycast = exports.PointerLock = exports.PointerEventsResult = exports.PointerEvents = exports.PlayerIdentityData = exports.NftShape = exports.MeshRenderer = exports.MeshCollider = exports.Material = exports.MainCamera = exports.InputModifier = exports.GltfContainerLoadingState = exports.GltfContainer = exports.EngineInfo = exports.CameraModeArea = exports.CameraMode = exports.Billboard = exports.AvatarShape = exports.AvatarModifierArea = exports.AvatarEquippedData = exports.AvatarEmoteCommand = exports.AvatarBase = exports.AvatarAttach = exports.AudioStream = exports.AudioSource = exports.AudioEvent = exports.Animator = void 0;
const Animator_gen_1 = require("./Animator.gen");
const AudioEvent_gen_1 = require("./AudioEvent.gen");
const AudioSource_gen_1 = require("./AudioSource.gen");
const AudioStream_gen_1 = require("./AudioStream.gen");
const AvatarAttach_gen_1 = require("./AvatarAttach.gen");
const AvatarBase_gen_1 = require("./AvatarBase.gen");
const AvatarEmoteCommand_gen_1 = require("./AvatarEmoteCommand.gen");
const AvatarEquippedData_gen_1 = require("./AvatarEquippedData.gen");
const AvatarModifierArea_gen_1 = require("./AvatarModifierArea.gen");
const AvatarShape_gen_1 = require("./AvatarShape.gen");
const Billboard_gen_1 = require("./Billboard.gen");
const CameraMode_gen_1 = require("./CameraMode.gen");
const CameraModeArea_gen_1 = require("./CameraModeArea.gen");
const EngineInfo_gen_1 = require("./EngineInfo.gen");
const GltfContainer_gen_1 = require("./GltfContainer.gen");
const GltfContainerLoadingState_gen_1 = require("./GltfContainerLoadingState.gen");
const InputModifier_gen_1 = require("./InputModifier.gen");
const MainCamera_gen_1 = require("./MainCamera.gen");
const Material_gen_1 = require("./Material.gen");
const MeshCollider_gen_1 = require("./MeshCollider.gen");
const MeshRenderer_gen_1 = require("./MeshRenderer.gen");
const NftShape_gen_1 = require("./NftShape.gen");
const PlayerIdentityData_gen_1 = require("./PlayerIdentityData.gen");
const PointerEvents_gen_1 = require("./PointerEvents.gen");
const PointerEventsResult_gen_1 = require("./PointerEventsResult.gen");
const PointerLock_gen_1 = require("./PointerLock.gen");
const Raycast_gen_1 = require("./Raycast.gen");
const RaycastResult_gen_1 = require("./RaycastResult.gen");
const RealmInfo_gen_1 = require("./RealmInfo.gen");
const TextShape_gen_1 = require("./TextShape.gen");
const Tween_gen_1 = require("./Tween.gen");
const TweenSequence_gen_1 = require("./TweenSequence.gen");
const TweenState_gen_1 = require("./TweenState.gen");
const UiBackground_gen_1 = require("./UiBackground.gen");
const UiCanvasInformation_gen_1 = require("./UiCanvasInformation.gen");
const UiDropdown_gen_1 = require("./UiDropdown.gen");
const UiDropdownResult_gen_1 = require("./UiDropdownResult.gen");
const UiInput_gen_1 = require("./UiInput.gen");
const UiInputResult_gen_1 = require("./UiInputResult.gen");
const UiText_gen_1 = require("./UiText.gen");
const UiTransform_gen_1 = require("./UiTransform.gen");
const VideoEvent_gen_1 = require("./VideoEvent.gen");
const VideoPlayer_gen_1 = require("./VideoPlayer.gen");
const VirtualCamera_gen_1 = require("./VirtualCamera.gen");
const VisibilityComponent_gen_1 = require("./VisibilityComponent.gen");
__exportStar(require("./pb/decentraland/sdk/components/animator.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/audio_event.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/audio_source.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/audio_stream.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/avatar_attach.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/avatar_base.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/avatar_emote_command.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/avatar_equipped_data.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/avatar_modifier_area.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/avatar_shape.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/billboard.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/camera_mode.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/camera_mode_area.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/engine_info.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/gltf_container.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/gltf_container_loading_state.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/input_modifier.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/main_camera.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/material.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/mesh_collider.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/mesh_renderer.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/nft_shape.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/player_identity_data.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/pointer_events.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/pointer_events_result.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/pointer_lock.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/raycast.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/raycast_result.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/realm_info.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/text_shape.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/tween.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/tween_sequence.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/tween_state.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/ui_background.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/ui_canvas_information.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/ui_dropdown.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/ui_dropdown_result.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/ui_input.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/ui_input_result.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/ui_text.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/ui_transform.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/video_event.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/video_player.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/virtual_camera.gen"), exports);
__exportStar(require("./pb/decentraland/sdk/components/visibility_component.gen"), exports);
/** @public */ const Animator = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::Animator", Animator_gen_1.AnimatorSchema);
exports.Animator = Animator;
/** @public */ const AudioEvent = (engine) => /* @__PURE__ */ engine.defineValueSetComponentFromSchema("core::AudioEvent", AudioEvent_gen_1.AudioEventSchema, {
    timestampFunction: (t) => t.timestamp,
    maxElements: 100
});
exports.AudioEvent = AudioEvent;
/** @public */ const AudioSource = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::AudioSource", AudioSource_gen_1.AudioSourceSchema);
exports.AudioSource = AudioSource;
/** @public */ const AudioStream = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::AudioStream", AudioStream_gen_1.AudioStreamSchema);
exports.AudioStream = AudioStream;
/** @public */ const AvatarAttach = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::AvatarAttach", AvatarAttach_gen_1.AvatarAttachSchema);
exports.AvatarAttach = AvatarAttach;
/** @public */ const AvatarBase = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::AvatarBase", AvatarBase_gen_1.AvatarBaseSchema);
exports.AvatarBase = AvatarBase;
/** @public */ const AvatarEmoteCommand = (engine) => /* @__PURE__ */ engine.defineValueSetComponentFromSchema("core::AvatarEmoteCommand", AvatarEmoteCommand_gen_1.AvatarEmoteCommandSchema, {
    timestampFunction: (t) => t.timestamp,
    maxElements: 100
});
exports.AvatarEmoteCommand = AvatarEmoteCommand;
/** @public */ const AvatarEquippedData = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::AvatarEquippedData", AvatarEquippedData_gen_1.AvatarEquippedDataSchema);
exports.AvatarEquippedData = AvatarEquippedData;
/** @public */ const AvatarModifierArea = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::AvatarModifierArea", AvatarModifierArea_gen_1.AvatarModifierAreaSchema);
exports.AvatarModifierArea = AvatarModifierArea;
/** @public */ const AvatarShape = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::AvatarShape", AvatarShape_gen_1.AvatarShapeSchema);
exports.AvatarShape = AvatarShape;
/** @public */ const Billboard = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::Billboard", Billboard_gen_1.BillboardSchema);
exports.Billboard = Billboard;
/** @public */ const CameraMode = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::CameraMode", CameraMode_gen_1.CameraModeSchema);
exports.CameraMode = CameraMode;
/** @public */ const CameraModeArea = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::CameraModeArea", CameraModeArea_gen_1.CameraModeAreaSchema);
exports.CameraModeArea = CameraModeArea;
/** @public */ const EngineInfo = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::EngineInfo", EngineInfo_gen_1.EngineInfoSchema);
exports.EngineInfo = EngineInfo;
/** @public */ const GltfContainer = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::GltfContainer", GltfContainer_gen_1.GltfContainerSchema);
exports.GltfContainer = GltfContainer;
/** @public */ const GltfContainerLoadingState = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::GltfContainerLoadingState", GltfContainerLoadingState_gen_1.GltfContainerLoadingStateSchema);
exports.GltfContainerLoadingState = GltfContainerLoadingState;
/** @public */ const InputModifier = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::InputModifier", InputModifier_gen_1.InputModifierSchema);
exports.InputModifier = InputModifier;
/** @public */ const MainCamera = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::MainCamera", MainCamera_gen_1.MainCameraSchema);
exports.MainCamera = MainCamera;
/** @public */ const Material = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::Material", Material_gen_1.MaterialSchema);
exports.Material = Material;
/** @public */ const MeshCollider = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::MeshCollider", MeshCollider_gen_1.MeshColliderSchema);
exports.MeshCollider = MeshCollider;
/** @public */ const MeshRenderer = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::MeshRenderer", MeshRenderer_gen_1.MeshRendererSchema);
exports.MeshRenderer = MeshRenderer;
/** @public */ const NftShape = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::NftShape", NftShape_gen_1.NftShapeSchema);
exports.NftShape = NftShape;
/** @public */ const PlayerIdentityData = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::PlayerIdentityData", PlayerIdentityData_gen_1.PlayerIdentityDataSchema);
exports.PlayerIdentityData = PlayerIdentityData;
/** @public */ const PointerEvents = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::PointerEvents", PointerEvents_gen_1.PointerEventsSchema);
exports.PointerEvents = PointerEvents;
/** @public */ const PointerEventsResult = (engine) => /* @__PURE__ */ engine.defineValueSetComponentFromSchema("core::PointerEventsResult", PointerEventsResult_gen_1.PointerEventsResultSchema, {
    timestampFunction: (t) => t.timestamp,
    maxElements: 100
});
exports.PointerEventsResult = PointerEventsResult;
/** @public */ const PointerLock = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::PointerLock", PointerLock_gen_1.PointerLockSchema);
exports.PointerLock = PointerLock;
/** @public */ const Raycast = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::Raycast", Raycast_gen_1.RaycastSchema);
exports.Raycast = Raycast;
/** @public */ const RaycastResult = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::RaycastResult", RaycastResult_gen_1.RaycastResultSchema);
exports.RaycastResult = RaycastResult;
/** @public */ const RealmInfo = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::RealmInfo", RealmInfo_gen_1.RealmInfoSchema);
exports.RealmInfo = RealmInfo;
/** @public */ const TextShape = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::TextShape", TextShape_gen_1.TextShapeSchema);
exports.TextShape = TextShape;
/** @public */ const Tween = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::Tween", Tween_gen_1.TweenSchema);
exports.Tween = Tween;
/** @public */ const TweenSequence = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::TweenSequence", TweenSequence_gen_1.TweenSequenceSchema);
exports.TweenSequence = TweenSequence;
/** @public */ const TweenState = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::TweenState", TweenState_gen_1.TweenStateSchema);
exports.TweenState = TweenState;
/** @public */ const UiBackground = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::UiBackground", UiBackground_gen_1.UiBackgroundSchema);
exports.UiBackground = UiBackground;
/** @public */ const UiCanvasInformation = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::UiCanvasInformation", UiCanvasInformation_gen_1.UiCanvasInformationSchema);
exports.UiCanvasInformation = UiCanvasInformation;
/** @public */ const UiDropdown = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::UiDropdown", UiDropdown_gen_1.UiDropdownSchema);
exports.UiDropdown = UiDropdown;
/** @public */ const UiDropdownResult = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::UiDropdownResult", UiDropdownResult_gen_1.UiDropdownResultSchema);
exports.UiDropdownResult = UiDropdownResult;
/** @public */ const UiInput = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::UiInput", UiInput_gen_1.UiInputSchema);
exports.UiInput = UiInput;
/** @public */ const UiInputResult = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::UiInputResult", UiInputResult_gen_1.UiInputResultSchema);
exports.UiInputResult = UiInputResult;
/** @public */ const UiText = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::UiText", UiText_gen_1.UiTextSchema);
exports.UiText = UiText;
/** @public */ const UiTransform = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::UiTransform", UiTransform_gen_1.UiTransformSchema);
exports.UiTransform = UiTransform;
/** @public */ const VideoEvent = (engine) => /* @__PURE__ */ engine.defineValueSetComponentFromSchema("core::VideoEvent", VideoEvent_gen_1.VideoEventSchema, {
    timestampFunction: (t) => t.timestamp,
    maxElements: 100
});
exports.VideoEvent = VideoEvent;
/** @public */ const VideoPlayer = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::VideoPlayer", VideoPlayer_gen_1.VideoPlayerSchema);
exports.VideoPlayer = VideoPlayer;
/** @public */ const VirtualCamera = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::VirtualCamera", VirtualCamera_gen_1.VirtualCameraSchema);
exports.VirtualCamera = VirtualCamera;
/** @public */ const VisibilityComponent = engine => 
/* @__PURE__ */ engine.defineComponentFromSchema("core::VisibilityComponent", VisibilityComponent_gen_1.VisibilityComponentSchema);
exports.VisibilityComponent = VisibilityComponent;
/** public */
exports.componentDefinitionByName = {
    "core::Animator": exports.Animator,
    "core::AudioEvent": exports.AudioEvent,
    "core::AudioSource": exports.AudioSource,
    "core::AudioStream": exports.AudioStream,
    "core::AvatarAttach": exports.AvatarAttach,
    "core::AvatarBase": exports.AvatarBase,
    "core::AvatarEmoteCommand": exports.AvatarEmoteCommand,
    "core::AvatarEquippedData": exports.AvatarEquippedData,
    "core::AvatarModifierArea": exports.AvatarModifierArea,
    "core::AvatarShape": exports.AvatarShape,
    "core::Billboard": exports.Billboard,
    "core::CameraMode": exports.CameraMode,
    "core::CameraModeArea": exports.CameraModeArea,
    "core::EngineInfo": exports.EngineInfo,
    "core::GltfContainer": exports.GltfContainer,
    "core::GltfContainerLoadingState": exports.GltfContainerLoadingState,
    "core::InputModifier": exports.InputModifier,
    "core::MainCamera": exports.MainCamera,
    "core::Material": exports.Material,
    "core::MeshCollider": exports.MeshCollider,
    "core::MeshRenderer": exports.MeshRenderer,
    "core::NftShape": exports.NftShape,
    "core::PlayerIdentityData": exports.PlayerIdentityData,
    "core::PointerEvents": exports.PointerEvents,
    "core::PointerEventsResult": exports.PointerEventsResult,
    "core::PointerLock": exports.PointerLock,
    "core::Raycast": exports.Raycast,
    "core::RaycastResult": exports.RaycastResult,
    "core::RealmInfo": exports.RealmInfo,
    "core::TextShape": exports.TextShape,
    "core::Tween": exports.Tween,
    "core::TweenSequence": exports.TweenSequence,
    "core::TweenState": exports.TweenState,
    "core::UiBackground": exports.UiBackground,
    "core::UiCanvasInformation": exports.UiCanvasInformation,
    "core::UiDropdown": exports.UiDropdown,
    "core::UiDropdownResult": exports.UiDropdownResult,
    "core::UiInput": exports.UiInput,
    "core::UiInputResult": exports.UiInputResult,
    "core::UiText": exports.UiText,
    "core::UiTransform": exports.UiTransform,
    "core::VideoEvent": exports.VideoEvent,
    "core::VideoPlayer": exports.VideoPlayer,
    "core::VirtualCamera": exports.VirtualCamera,
    "core::VisibilityComponent": exports.VisibilityComponent,
};
