// @dcl/protocol/out-js/decentraland/sdk/components/avatar_attach.gen.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PBAvatarAttach = exports.AvatarAnchorPointType = exports.protobufPackage = void 0;
exports.avatarAnchorPointTypeFromJSON = avatarAnchorPointTypeFromJSON;
exports.avatarAnchorPointTypeToJSON = avatarAnchorPointTypeToJSON;
/* eslint-disable */
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "decentraland.sdk.components";
/** AvatarAnchorPointType determines the part of the avatar's body that anchors the Entity. */
var AvatarAnchorPointType;
(function (AvatarAnchorPointType) {
    /** AAPT_POSITION - @deprecated consider parenting to `engine.PlayerEntity`, this will attach to player position with an arbitrary offset */
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_POSITION"] = 0] = "AAPT_POSITION";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_NAME_TAG"] = 1] = "AAPT_NAME_TAG";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_HEAD"] = 4] = "AAPT_HEAD";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_NECK"] = 5] = "AAPT_NECK";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_SPINE"] = 6] = "AAPT_SPINE";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_SPINE1"] = 7] = "AAPT_SPINE1";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_SPINE2"] = 8] = "AAPT_SPINE2";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_HIP"] = 9] = "AAPT_HIP";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_LEFT_SHOULDER"] = 10] = "AAPT_LEFT_SHOULDER";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_LEFT_ARM"] = 11] = "AAPT_LEFT_ARM";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_LEFT_FOREARM"] = 12] = "AAPT_LEFT_FOREARM";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_LEFT_HAND"] = 2] = "AAPT_LEFT_HAND";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_LEFT_HAND_INDEX"] = 13] = "AAPT_LEFT_HAND_INDEX";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_RIGHT_SHOULDER"] = 14] = "AAPT_RIGHT_SHOULDER";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_RIGHT_ARM"] = 15] = "AAPT_RIGHT_ARM";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_RIGHT_FOREARM"] = 16] = "AAPT_RIGHT_FOREARM";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_RIGHT_HAND"] = 3] = "AAPT_RIGHT_HAND";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_RIGHT_HAND_INDEX"] = 17] = "AAPT_RIGHT_HAND_INDEX";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_LEFT_UP_LEG"] = 18] = "AAPT_LEFT_UP_LEG";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_LEFT_LEG"] = 19] = "AAPT_LEFT_LEG";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_LEFT_FOOT"] = 20] = "AAPT_LEFT_FOOT";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_LEFT_TOE_BASE"] = 21] = "AAPT_LEFT_TOE_BASE";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_RIGHT_UP_LEG"] = 22] = "AAPT_RIGHT_UP_LEG";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_RIGHT_LEG"] = 23] = "AAPT_RIGHT_LEG";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_RIGHT_FOOT"] = 24] = "AAPT_RIGHT_FOOT";
    AvatarAnchorPointType[AvatarAnchorPointType["AAPT_RIGHT_TOE_BASE"] = 25] = "AAPT_RIGHT_TOE_BASE";
    AvatarAnchorPointType[AvatarAnchorPointType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(AvatarAnchorPointType || (exports.AvatarAnchorPointType = AvatarAnchorPointType = {}));
function avatarAnchorPointTypeFromJSON(object) {
    switch (object) {
        case 0:
        case "AAPT_POSITION":
            return AvatarAnchorPointType.AAPT_POSITION;
        case 1:
        case "AAPT_NAME_TAG":
            return AvatarAnchorPointType.AAPT_NAME_TAG;
        case 4:
        case "AAPT_HEAD":
            return AvatarAnchorPointType.AAPT_HEAD;
        case 5:
        case "AAPT_NECK":
            return AvatarAnchorPointType.AAPT_NECK;
        case 6:
        case "AAPT_SPINE":
            return AvatarAnchorPointType.AAPT_SPINE;
        case 7:
        case "AAPT_SPINE1":
            return AvatarAnchorPointType.AAPT_SPINE1;
        case 8:
        case "AAPT_SPINE2":
            return AvatarAnchorPointType.AAPT_SPINE2;
        case 9:
        case "AAPT_HIP":
            return AvatarAnchorPointType.AAPT_HIP;
        case 10:
        case "AAPT_LEFT_SHOULDER":
            return AvatarAnchorPointType.AAPT_LEFT_SHOULDER;
        case 11:
        case "AAPT_LEFT_ARM":
            return AvatarAnchorPointType.AAPT_LEFT_ARM;
        case 12:
        case "AAPT_LEFT_FOREARM":
            return AvatarAnchorPointType.AAPT_LEFT_FOREARM;
        case 2:
        case "AAPT_LEFT_HAND":
            return AvatarAnchorPointType.AAPT_LEFT_HAND;
        case 13:
        case "AAPT_LEFT_HAND_INDEX":
            return AvatarAnchorPointType.AAPT_LEFT_HAND_INDEX;
        case 14:
        case "AAPT_RIGHT_SHOULDER":
            return AvatarAnchorPointType.AAPT_RIGHT_SHOULDER;
        case 15:
        case "AAPT_RIGHT_ARM":
            return AvatarAnchorPointType.AAPT_RIGHT_ARM;
        case 16:
        case "AAPT_RIGHT_FOREARM":
            return AvatarAnchorPointType.AAPT_RIGHT_FOREARM;
        case 3:
        case "AAPT_RIGHT_HAND":
            return AvatarAnchorPointType.AAPT_RIGHT_HAND;
        case 17:
        case "AAPT_RIGHT_HAND_INDEX":
            return AvatarAnchorPointType.AAPT_RIGHT_HAND_INDEX;
        case 18:
        case "AAPT_LEFT_UP_LEG":
            return AvatarAnchorPointType.AAPT_LEFT_UP_LEG;
        case 19:
        case "AAPT_LEFT_LEG":
            return AvatarAnchorPointType.AAPT_LEFT_LEG;
        case 20:
        case "AAPT_LEFT_FOOT":
            return AvatarAnchorPointType.AAPT_LEFT_FOOT;
        case 21:
        case "AAPT_LEFT_TOE_BASE":
            return AvatarAnchorPointType.AAPT_LEFT_TOE_BASE;
        case 22:
        case "AAPT_RIGHT_UP_LEG":
            return AvatarAnchorPointType.AAPT_RIGHT_UP_LEG;
        case 23:
        case "AAPT_RIGHT_LEG":
            return AvatarAnchorPointType.AAPT_RIGHT_LEG;
        case 24:
        case "AAPT_RIGHT_FOOT":
            return AvatarAnchorPointType.AAPT_RIGHT_FOOT;
        case 25:
        case "AAPT_RIGHT_TOE_BASE":
            return AvatarAnchorPointType.AAPT_RIGHT_TOE_BASE;
        case -1:
        case "UNRECOGNIZED":
        default:
            return AvatarAnchorPointType.UNRECOGNIZED;
    }
}
function avatarAnchorPointTypeToJSON(object) {
    switch (object) {
        case AvatarAnchorPointType.AAPT_POSITION:
            return "AAPT_POSITION";
        case AvatarAnchorPointType.AAPT_NAME_TAG:
            return "AAPT_NAME_TAG";
        case AvatarAnchorPointType.AAPT_HEAD:
            return "AAPT_HEAD";
        case AvatarAnchorPointType.AAPT_NECK:
            return "AAPT_NECK";
        case AvatarAnchorPointType.AAPT_SPINE:
            return "AAPT_SPINE";
        case AvatarAnchorPointType.AAPT_SPINE1:
            return "AAPT_SPINE1";
        case AvatarAnchorPointType.AAPT_SPINE2:
            return "AAPT_SPINE2";
        case AvatarAnchorPointType.AAPT_HIP:
            return "AAPT_HIP";
        case AvatarAnchorPointType.AAPT_LEFT_SHOULDER:
            return "AAPT_LEFT_SHOULDER";
        case AvatarAnchorPointType.AAPT_LEFT_ARM:
            return "AAPT_LEFT_ARM";
        case AvatarAnchorPointType.AAPT_LEFT_FOREARM:
            return "AAPT_LEFT_FOREARM";
        case AvatarAnchorPointType.AAPT_LEFT_HAND:
            return "AAPT_LEFT_HAND";
        case AvatarAnchorPointType.AAPT_LEFT_HAND_INDEX:
            return "AAPT_LEFT_HAND_INDEX";
        case AvatarAnchorPointType.AAPT_RIGHT_SHOULDER:
            return "AAPT_RIGHT_SHOULDER";
        case AvatarAnchorPointType.AAPT_RIGHT_ARM:
            return "AAPT_RIGHT_ARM";
        case AvatarAnchorPointType.AAPT_RIGHT_FOREARM:
            return "AAPT_RIGHT_FOREARM";
        case AvatarAnchorPointType.AAPT_RIGHT_HAND:
            return "AAPT_RIGHT_HAND";
        case AvatarAnchorPointType.AAPT_RIGHT_HAND_INDEX:
            return "AAPT_RIGHT_HAND_INDEX";
        case AvatarAnchorPointType.AAPT_LEFT_UP_LEG:
            return "AAPT_LEFT_UP_LEG";
        case AvatarAnchorPointType.AAPT_LEFT_LEG:
            return "AAPT_LEFT_LEG";
        case AvatarAnchorPointType.AAPT_LEFT_FOOT:
            return "AAPT_LEFT_FOOT";
        case AvatarAnchorPointType.AAPT_LEFT_TOE_BASE:
            return "AAPT_LEFT_TOE_BASE";
        case AvatarAnchorPointType.AAPT_RIGHT_UP_LEG:
            return "AAPT_RIGHT_UP_LEG";
        case AvatarAnchorPointType.AAPT_RIGHT_LEG:
            return "AAPT_RIGHT_LEG";
        case AvatarAnchorPointType.AAPT_RIGHT_FOOT:
            return "AAPT_RIGHT_FOOT";
        case AvatarAnchorPointType.AAPT_RIGHT_TOE_BASE:
            return "AAPT_RIGHT_TOE_BASE";
        case AvatarAnchorPointType.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
function createBasePBAvatarAttach() {
    return { avatarId: undefined, anchorPointId: 0 };
}
var PBAvatarAttach;
(function (PBAvatarAttach) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.avatarId !== undefined) {
            writer.uint32(10).string(message.avatarId);
        }
        if (message.anchorPointId !== 0) {
            writer.uint32(16).int32(message.anchorPointId);
        }
        return writer;
    }
    PBAvatarAttach.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePBAvatarAttach();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.avatarId = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.anchorPointId = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    PBAvatarAttach.decode = decode;
    function fromJSON(object) {
        return {
            avatarId: isSet(object.avatarId) ? String(object.avatarId) : undefined,
            anchorPointId: isSet(object.anchorPointId) ? avatarAnchorPointTypeFromJSON(object.anchorPointId) : 0,
        };
    }
    PBAvatarAttach.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.avatarId !== undefined && (obj.avatarId = message.avatarId);
        message.anchorPointId !== undefined && (obj.anchorPointId = avatarAnchorPointTypeToJSON(message.anchorPointId));
        return obj;
    }
    PBAvatarAttach.toJSON = toJSON;
    function create(base) {
        return PBAvatarAttach.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    PBAvatarAttach.create = create;
    function fromPartial(object) {
        var _a, _b;
        const message = createBasePBAvatarAttach();
        message.avatarId = (_a = object.avatarId) !== null && _a !== void 0 ? _a : undefined;
        message.anchorPointId = (_b = object.anchorPointId) !== null && _b !== void 0 ? _b : 0;
        return message;
    }
    PBAvatarAttach.fromPartial = fromPartial;
})(PBAvatarAttach || (exports.PBAvatarAttach = PBAvatarAttach = {}));
function isSet(value) {
    return value !== null && value !== undefined;
}
