// @dcl/protocol/out-js/decentraland/sdk/components/billboard.gen.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PBBillboard = exports.BillboardMode = exports.protobufPackage = void 0;
exports.billboardModeFromJSON = billboardModeFromJSON;
exports.billboardModeToJSON = billboardModeToJSON;
/* eslint-disable */
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "decentraland.sdk.components";
/**
 * BillboardMode indicates one or more axis for automatic rotation, in OR-able bit flag form.
 * Only the values below and the (BM_X | BM_Y) combination are valid.
 */
var BillboardMode;
(function (BillboardMode) {
    BillboardMode[BillboardMode["BM_NONE"] = 0] = "BM_NONE";
    BillboardMode[BillboardMode["BM_X"] = 1] = "BM_X";
    BillboardMode[BillboardMode["BM_Y"] = 2] = "BM_Y";
    BillboardMode[BillboardMode["BM_Z"] = 4] = "BM_Z";
    /** BM_ALL - bitwise combination BM_X | BM_Y | BM_Z */
    BillboardMode[BillboardMode["BM_ALL"] = 7] = "BM_ALL";
    BillboardMode[BillboardMode["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(BillboardMode || (exports.BillboardMode = BillboardMode = {}));
function billboardModeFromJSON(object) {
    switch (object) {
        case 0:
        case "BM_NONE":
            return BillboardMode.BM_NONE;
        case 1:
        case "BM_X":
            return BillboardMode.BM_X;
        case 2:
        case "BM_Y":
            return BillboardMode.BM_Y;
        case 4:
        case "BM_Z":
            return BillboardMode.BM_Z;
        case 7:
        case "BM_ALL":
            return BillboardMode.BM_ALL;
        case -1:
        case "UNRECOGNIZED":
        default:
            return BillboardMode.UNRECOGNIZED;
    }
}
function billboardModeToJSON(object) {
    switch (object) {
        case BillboardMode.BM_NONE:
            return "BM_NONE";
        case BillboardMode.BM_X:
            return "BM_X";
        case BillboardMode.BM_Y:
            return "BM_Y";
        case BillboardMode.BM_Z:
            return "BM_Z";
        case BillboardMode.BM_ALL:
            return "BM_ALL";
        case BillboardMode.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
function createBasePBBillboard() {
    return { billboardMode: undefined };
}
var PBBillboard;
(function (PBBillboard) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.billboardMode !== undefined) {
            writer.uint32(8).int32(message.billboardMode);
        }
        return writer;
    }
    PBBillboard.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePBBillboard();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.billboardMode = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    PBBillboard.decode = decode;
    function fromJSON(object) {
        return { billboardMode: isSet(object.billboardMode) ? billboardModeFromJSON(object.billboardMode) : undefined };
    }
    PBBillboard.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.billboardMode !== undefined &&
            (obj.billboardMode = message.billboardMode !== undefined
                ? billboardModeToJSON(message.billboardMode)
                : undefined);
        return obj;
    }
    PBBillboard.toJSON = toJSON;
    function create(base) {
        return PBBillboard.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    PBBillboard.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBasePBBillboard();
        message.billboardMode = (_a = object.billboardMode) !== null && _a !== void 0 ? _a : undefined;
        return message;
    }
    PBBillboard.fromPartial = fromPartial;
})(PBBillboard || (exports.PBBillboard = PBBillboard = {}));
function isSet(value) {
    return value !== null && value !== undefined;
}
