// @dcl/protocol/out-js/decentraland/sdk/components/camera_mode.gen.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PBCameraMode = exports.protobufPackage = void 0;
/* eslint-disable */
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const camera_type_gen_1 = require("./common/camera_type.gen");
exports.protobufPackage = "decentraland.sdk.components";
function createBasePBCameraMode() {
    return { mode: 0 };
}
var PBCameraMode;
(function (PBCameraMode) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.mode !== 0) {
            writer.uint32(8).int32(message.mode);
        }
        return writer;
    }
    PBCameraMode.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePBCameraMode();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.mode = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    PBCameraMode.decode = decode;
    function fromJSON(object) {
        return { mode: isSet(object.mode) ? (0, camera_type_gen_1.cameraTypeFromJSON)(object.mode) : 0 };
    }
    PBCameraMode.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.mode !== undefined && (obj.mode = (0, camera_type_gen_1.cameraTypeToJSON)(message.mode));
        return obj;
    }
    PBCameraMode.toJSON = toJSON;
    function create(base) {
        return PBCameraMode.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    PBCameraMode.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBasePBCameraMode();
        message.mode = (_a = object.mode) !== null && _a !== void 0 ? _a : 0;
        return message;
    }
    PBCameraMode.fromPartial = fromPartial;
})(PBCameraMode || (exports.PBCameraMode = PBCameraMode = {}));
function isSet(value) {
    return value !== null && value !== undefined;
}
