// @dcl/protocol/out-js/decentraland/sdk/components/player_identity_data.gen.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PBPlayerIdentityData = exports.protobufPackage = void 0;
/* eslint-disable */
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "decentraland.sdk.components";
function createBasePBPlayerIdentityData() {
    return { address: "", isGuest: false };
}
var PBPlayerIdentityData;
(function (PBPlayerIdentityData) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.address !== "") {
            writer.uint32(10).string(message.address);
        }
        if (message.isGuest === true) {
            writer.uint32(24).bool(message.isGuest);
        }
        return writer;
    }
    PBPlayerIdentityData.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePBPlayerIdentityData();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.address = reader.string();
                    continue;
                case 3:
                    if (tag !== 24) {
                        break;
                    }
                    message.isGuest = reader.bool();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    PBPlayerIdentityData.decode = decode;
    function fromJSON(object) {
        return {
            address: isSet(object.address) ? String(object.address) : "",
            isGuest: isSet(object.isGuest) ? Boolean(object.isGuest) : false,
        };
    }
    PBPlayerIdentityData.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.address !== undefined && (obj.address = message.address);
        message.isGuest !== undefined && (obj.isGuest = message.isGuest);
        return obj;
    }
    PBPlayerIdentityData.toJSON = toJSON;
    function create(base) {
        return PBPlayerIdentityData.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    PBPlayerIdentityData.create = create;
    function fromPartial(object) {
        var _a, _b;
        const message = createBasePBPlayerIdentityData();
        message.address = (_a = object.address) !== null && _a !== void 0 ? _a : "";
        message.isGuest = (_b = object.isGuest) !== null && _b !== void 0 ? _b : false;
        return message;
    }
    PBPlayerIdentityData.fromPartial = fromPartial;
})(PBPlayerIdentityData || (exports.PBPlayerIdentityData = PBPlayerIdentityData = {}));
function isSet(value) {
    return value !== null && value !== undefined;
}
