// @dcl/protocol/out-js/decentraland/sdk/components/pointer_lock.gen.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PBPointerLock = exports.protobufPackage = void 0;
/* eslint-disable */
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "decentraland.sdk.components";
function createBasePBPointerLock() {
    return { isPointerLocked: false };
}
var PBPointerLock;
(function (PBPointerLock) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.isPointerLocked === true) {
            writer.uint32(8).bool(message.isPointerLocked);
        }
        return writer;
    }
    PBPointerLock.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePBPointerLock();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.isPointerLocked = reader.bool();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    PBPointerLock.decode = decode;
    function fromJSON(object) {
        return { isPointerLocked: isSet(object.isPointerLocked) ? Boolean(object.isPointerLocked) : false };
    }
    PBPointerLock.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.isPointerLocked !== undefined && (obj.isPointerLocked = message.isPointerLocked);
        return obj;
    }
    PBPointerLock.toJSON = toJSON;
    function create(base) {
        return PBPointerLock.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    PBPointerLock.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBasePBPointerLock();
        message.isPointerLocked = (_a = object.isPointerLocked) !== null && _a !== void 0 ? _a : false;
        return message;
    }
    PBPointerLock.fromPartial = fromPartial;
})(PBPointerLock || (exports.PBPointerLock = PBPointerLock = {}));
function isSet(value) {
    return value !== null && value !== undefined;
}
