// @dcl/protocol/out-js/decentraland/sdk/components/ui_dropdown_result.gen.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PBUiDropdownResult = exports.protobufPackage = void 0;
/* eslint-disable */
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "decentraland.sdk.components";
function createBasePBUiDropdownResult() {
    return { value: 0 };
}
var PBUiDropdownResult;
(function (PBUiDropdownResult) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.value !== 0) {
            writer.uint32(8).int32(message.value);
        }
        return writer;
    }
    PBUiDropdownResult.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePBUiDropdownResult();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.value = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    PBUiDropdownResult.decode = decode;
    function fromJSON(object) {
        return { value: isSet(object.value) ? Number(object.value) : 0 };
    }
    PBUiDropdownResult.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.value !== undefined && (obj.value = Math.round(message.value));
        return obj;
    }
    PBUiDropdownResult.toJSON = toJSON;
    function create(base) {
        return PBUiDropdownResult.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    PBUiDropdownResult.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBasePBUiDropdownResult();
        message.value = (_a = object.value) !== null && _a !== void 0 ? _a : 0;
        return message;
    }
    PBUiDropdownResult.fromPartial = fromPartial;
})(PBUiDropdownResult || (exports.PBUiDropdownResult = PBUiDropdownResult = {}));
function isSet(value) {
    return value !== null && value !== undefined;
}
