// @dcl/protocol/out-js/decentraland/sdk/components/ui_input_result.gen.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PBUiInputResult = exports.protobufPackage = void 0;
/* eslint-disable */
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "decentraland.sdk.components";
function createBasePBUiInputResult() {
    return { value: "", isSubmit: undefined };
}
var PBUiInputResult;
(function (PBUiInputResult) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.value !== "") {
            writer.uint32(10).string(message.value);
        }
        if (message.isSubmit !== undefined) {
            writer.uint32(16).bool(message.isSubmit);
        }
        return writer;
    }
    PBUiInputResult.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePBUiInputResult();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.value = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.isSubmit = reader.bool();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    PBUiInputResult.decode = decode;
    function fromJSON(object) {
        return {
            value: isSet(object.value) ? String(object.value) : "",
            isSubmit: isSet(object.isSubmit) ? Boolean(object.isSubmit) : undefined,
        };
    }
    PBUiInputResult.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.value !== undefined && (obj.value = message.value);
        message.isSubmit !== undefined && (obj.isSubmit = message.isSubmit);
        return obj;
    }
    PBUiInputResult.toJSON = toJSON;
    function create(base) {
        return PBUiInputResult.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    PBUiInputResult.create = create;
    function fromPartial(object) {
        var _a, _b;
        const message = createBasePBUiInputResult();
        message.value = (_a = object.value) !== null && _a !== void 0 ? _a : "";
        message.isSubmit = (_b = object.isSubmit) !== null && _b !== void 0 ? _b : undefined;
        return message;
    }
    PBUiInputResult.fromPartial = fromPartial;
})(PBUiInputResult || (exports.PBUiInputResult = PBUiInputResult = {}));
function isSet(value) {
    return value !== null && value !== undefined;
}
