// @dcl/protocol/out-js/decentraland/sdk/components/ui_text.gen.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PBUiText = exports.TextWrap = exports.protobufPackage = void 0;
exports.textWrapFromJSON = textWrapFromJSON;
exports.textWrapToJSON = textWrapToJSON;
/* eslint-disable */
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const colors_gen_1 = require("../../common/colors.gen");
const texts_gen_1 = require("./common/texts.gen");
exports.protobufPackage = "decentraland.sdk.components";
var TextWrap;
(function (TextWrap) {
    TextWrap[TextWrap["TW_WRAP"] = 0] = "TW_WRAP";
    TextWrap[TextWrap["TW_NO_WRAP"] = 1] = "TW_NO_WRAP";
    TextWrap[TextWrap["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(TextWrap || (exports.TextWrap = TextWrap = {}));
function textWrapFromJSON(object) {
    switch (object) {
        case 0:
        case "TW_WRAP":
            return TextWrap.TW_WRAP;
        case 1:
        case "TW_NO_WRAP":
            return TextWrap.TW_NO_WRAP;
        case -1:
        case "UNRECOGNIZED":
        default:
            return TextWrap.UNRECOGNIZED;
    }
}
function textWrapToJSON(object) {
    switch (object) {
        case TextWrap.TW_WRAP:
            return "TW_WRAP";
        case TextWrap.TW_NO_WRAP:
            return "TW_NO_WRAP";
        case TextWrap.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
function createBasePBUiText() {
    return {
        value: "",
        color: undefined,
        textAlign: undefined,
        font: undefined,
        fontSize: undefined,
        textWrap: undefined,
    };
}
var PBUiText;
(function (PBUiText) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.value !== "") {
            writer.uint32(10).string(message.value);
        }
        if (message.color !== undefined) {
            colors_gen_1.Color4.encode(message.color, writer.uint32(18).fork()).ldelim();
        }
        if (message.textAlign !== undefined) {
            writer.uint32(24).int32(message.textAlign);
        }
        if (message.font !== undefined) {
            writer.uint32(32).int32(message.font);
        }
        if (message.fontSize !== undefined) {
            writer.uint32(40).int32(message.fontSize);
        }
        if (message.textWrap !== undefined) {
            writer.uint32(48).int32(message.textWrap);
        }
        return writer;
    }
    PBUiText.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePBUiText();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.value = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.color = colors_gen_1.Color4.decode(reader, reader.uint32());
                    continue;
                case 3:
                    if (tag !== 24) {
                        break;
                    }
                    message.textAlign = reader.int32();
                    continue;
                case 4:
                    if (tag !== 32) {
                        break;
                    }
                    message.font = reader.int32();
                    continue;
                case 5:
                    if (tag !== 40) {
                        break;
                    }
                    message.fontSize = reader.int32();
                    continue;
                case 6:
                    if (tag !== 48) {
                        break;
                    }
                    message.textWrap = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    PBUiText.decode = decode;
    function fromJSON(object) {
        return {
            value: isSet(object.value) ? String(object.value) : "",
            color: isSet(object.color) ? colors_gen_1.Color4.fromJSON(object.color) : undefined,
            textAlign: isSet(object.textAlign) ? (0, texts_gen_1.textAlignModeFromJSON)(object.textAlign) : undefined,
            font: isSet(object.font) ? (0, texts_gen_1.fontFromJSON)(object.font) : undefined,
            fontSize: isSet(object.fontSize) ? Number(object.fontSize) : undefined,
            textWrap: isSet(object.textWrap) ? textWrapFromJSON(object.textWrap) : undefined,
        };
    }
    PBUiText.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.value !== undefined && (obj.value = message.value);
        message.color !== undefined && (obj.color = message.color ? colors_gen_1.Color4.toJSON(message.color) : undefined);
        message.textAlign !== undefined &&
            (obj.textAlign = message.textAlign !== undefined ? (0, texts_gen_1.textAlignModeToJSON)(message.textAlign) : undefined);
        message.font !== undefined && (obj.font = message.font !== undefined ? (0, texts_gen_1.fontToJSON)(message.font) : undefined);
        message.fontSize !== undefined && (obj.fontSize = Math.round(message.fontSize));
        message.textWrap !== undefined &&
            (obj.textWrap = message.textWrap !== undefined ? textWrapToJSON(message.textWrap) : undefined);
        return obj;
    }
    PBUiText.toJSON = toJSON;
    function create(base) {
        return PBUiText.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    PBUiText.create = create;
    function fromPartial(object) {
        var _a, _b, _c, _d, _e;
        const message = createBasePBUiText();
        message.value = (_a = object.value) !== null && _a !== void 0 ? _a : "";
        message.color = (object.color !== undefined && object.color !== null)
            ? colors_gen_1.Color4.fromPartial(object.color)
            : undefined;
        message.textAlign = (_b = object.textAlign) !== null && _b !== void 0 ? _b : undefined;
        message.font = (_c = object.font) !== null && _c !== void 0 ? _c : undefined;
        message.fontSize = (_d = object.fontSize) !== null && _d !== void 0 ? _d : undefined;
        message.textWrap = (_e = object.textWrap) !== null && _e !== void 0 ? _e : undefined;
        return message;
    }
    PBUiText.fromPartial = fromPartial;
})(PBUiText || (exports.PBUiText = PBUiText = {}));
function isSet(value) {
    return value !== null && value !== undefined;
}
