// @dcl/protocol/out-js/decentraland/kernel/apis/runtime.gen.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RuntimeServiceDefinition = exports.GetExplorerInformationResponse_ConfigurationsEntry = exports.GetExplorerInformationResponse = exports.GetExplorerInformationRequest = exports.CurrentSceneEntityResponse = exports.CurrentSceneEntityRequest = exports.ReadFileResponse = exports.ReadFileRequest = exports.GetWorldTimeRequest = exports.GetRealmRequest = exports.GetWorldTimeResponse = exports.GetRealmResponse = exports.protobufPackage = void 0;
/* eslint-disable */
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const content_mapping_gen_1 = require("../../common/content_mapping.gen");
const realm_info_gen_1 = require("../../sdk/components/realm_info.gen");
exports.protobufPackage = "decentraland.kernel.apis";
function createBaseGetRealmResponse() {
    return { realmInfo: undefined };
}
var GetRealmResponse;
(function (GetRealmResponse) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.realmInfo !== undefined) {
            realm_info_gen_1.PBRealmInfo.encode(message.realmInfo, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    }
    GetRealmResponse.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetRealmResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.realmInfo = realm_info_gen_1.PBRealmInfo.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    GetRealmResponse.decode = decode;
    function fromJSON(object) {
        return { realmInfo: isSet(object.realmInfo) ? realm_info_gen_1.PBRealmInfo.fromJSON(object.realmInfo) : undefined };
    }
    GetRealmResponse.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.realmInfo !== undefined &&
            (obj.realmInfo = message.realmInfo ? realm_info_gen_1.PBRealmInfo.toJSON(message.realmInfo) : undefined);
        return obj;
    }
    GetRealmResponse.toJSON = toJSON;
    function create(base) {
        return GetRealmResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    GetRealmResponse.create = create;
    function fromPartial(object) {
        const message = createBaseGetRealmResponse();
        message.realmInfo = (object.realmInfo !== undefined && object.realmInfo !== null)
            ? realm_info_gen_1.PBRealmInfo.fromPartial(object.realmInfo)
            : undefined;
        return message;
    }
    GetRealmResponse.fromPartial = fromPartial;
})(GetRealmResponse || (exports.GetRealmResponse = GetRealmResponse = {}));
function createBaseGetWorldTimeResponse() {
    return { seconds: 0 };
}
var GetWorldTimeResponse;
(function (GetWorldTimeResponse) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.seconds !== 0) {
            writer.uint32(8).int32(message.seconds);
        }
        return writer;
    }
    GetWorldTimeResponse.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetWorldTimeResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.seconds = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    GetWorldTimeResponse.decode = decode;
    function fromJSON(object) {
        return { seconds: isSet(object.seconds) ? Number(object.seconds) : 0 };
    }
    GetWorldTimeResponse.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.seconds !== undefined && (obj.seconds = Math.round(message.seconds));
        return obj;
    }
    GetWorldTimeResponse.toJSON = toJSON;
    function create(base) {
        return GetWorldTimeResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    GetWorldTimeResponse.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseGetWorldTimeResponse();
        message.seconds = (_a = object.seconds) !== null && _a !== void 0 ? _a : 0;
        return message;
    }
    GetWorldTimeResponse.fromPartial = fromPartial;
})(GetWorldTimeResponse || (exports.GetWorldTimeResponse = GetWorldTimeResponse = {}));
function createBaseGetRealmRequest() {
    return {};
}
var GetRealmRequest;
(function (GetRealmRequest) {
    function encode(_, writer = minimal_1.default.Writer.create()) {
        return writer;
    }
    GetRealmRequest.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetRealmRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    GetRealmRequest.decode = decode;
    function fromJSON(_) {
        return {};
    }
    GetRealmRequest.fromJSON = fromJSON;
    function toJSON(_) {
        const obj = {};
        return obj;
    }
    GetRealmRequest.toJSON = toJSON;
    function create(base) {
        return GetRealmRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    GetRealmRequest.create = create;
    function fromPartial(_) {
        const message = createBaseGetRealmRequest();
        return message;
    }
    GetRealmRequest.fromPartial = fromPartial;
})(GetRealmRequest || (exports.GetRealmRequest = GetRealmRequest = {}));
function createBaseGetWorldTimeRequest() {
    return {};
}
var GetWorldTimeRequest;
(function (GetWorldTimeRequest) {
    function encode(_, writer = minimal_1.default.Writer.create()) {
        return writer;
    }
    GetWorldTimeRequest.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetWorldTimeRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    GetWorldTimeRequest.decode = decode;
    function fromJSON(_) {
        return {};
    }
    GetWorldTimeRequest.fromJSON = fromJSON;
    function toJSON(_) {
        const obj = {};
        return obj;
    }
    GetWorldTimeRequest.toJSON = toJSON;
    function create(base) {
        return GetWorldTimeRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    GetWorldTimeRequest.create = create;
    function fromPartial(_) {
        const message = createBaseGetWorldTimeRequest();
        return message;
    }
    GetWorldTimeRequest.fromPartial = fromPartial;
})(GetWorldTimeRequest || (exports.GetWorldTimeRequest = GetWorldTimeRequest = {}));
function createBaseReadFileRequest() {
    return { fileName: "" };
}
var ReadFileRequest;
(function (ReadFileRequest) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.fileName !== "") {
            writer.uint32(10).string(message.fileName);
        }
        return writer;
    }
    ReadFileRequest.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseReadFileRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.fileName = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    ReadFileRequest.decode = decode;
    function fromJSON(object) {
        return { fileName: isSet(object.fileName) ? String(object.fileName) : "" };
    }
    ReadFileRequest.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.fileName !== undefined && (obj.fileName = message.fileName);
        return obj;
    }
    ReadFileRequest.toJSON = toJSON;
    function create(base) {
        return ReadFileRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    ReadFileRequest.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseReadFileRequest();
        message.fileName = (_a = object.fileName) !== null && _a !== void 0 ? _a : "";
        return message;
    }
    ReadFileRequest.fromPartial = fromPartial;
})(ReadFileRequest || (exports.ReadFileRequest = ReadFileRequest = {}));
function createBaseReadFileResponse() {
    return { content: new Uint8Array(0), hash: "" };
}
var ReadFileResponse;
(function (ReadFileResponse) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.content.length !== 0) {
            writer.uint32(10).bytes(message.content);
        }
        if (message.hash !== "") {
            writer.uint32(18).string(message.hash);
        }
        return writer;
    }
    ReadFileResponse.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseReadFileResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.content = reader.bytes();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.hash = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    ReadFileResponse.decode = decode;
    function fromJSON(object) {
        return {
            content: isSet(object.content) ? bytesFromBase64(object.content) : new Uint8Array(0),
            hash: isSet(object.hash) ? String(object.hash) : "",
        };
    }
    ReadFileResponse.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.content !== undefined &&
            (obj.content = base64FromBytes(message.content !== undefined ? message.content : new Uint8Array(0)));
        message.hash !== undefined && (obj.hash = message.hash);
        return obj;
    }
    ReadFileResponse.toJSON = toJSON;
    function create(base) {
        return ReadFileResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    ReadFileResponse.create = create;
    function fromPartial(object) {
        var _a, _b;
        const message = createBaseReadFileResponse();
        message.content = (_a = object.content) !== null && _a !== void 0 ? _a : new Uint8Array(0);
        message.hash = (_b = object.hash) !== null && _b !== void 0 ? _b : "";
        return message;
    }
    ReadFileResponse.fromPartial = fromPartial;
})(ReadFileResponse || (exports.ReadFileResponse = ReadFileResponse = {}));
function createBaseCurrentSceneEntityRequest() {
    return {};
}
var CurrentSceneEntityRequest;
(function (CurrentSceneEntityRequest) {
    function encode(_, writer = minimal_1.default.Writer.create()) {
        return writer;
    }
    CurrentSceneEntityRequest.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCurrentSceneEntityRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    CurrentSceneEntityRequest.decode = decode;
    function fromJSON(_) {
        return {};
    }
    CurrentSceneEntityRequest.fromJSON = fromJSON;
    function toJSON(_) {
        const obj = {};
        return obj;
    }
    CurrentSceneEntityRequest.toJSON = toJSON;
    function create(base) {
        return CurrentSceneEntityRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    CurrentSceneEntityRequest.create = create;
    function fromPartial(_) {
        const message = createBaseCurrentSceneEntityRequest();
        return message;
    }
    CurrentSceneEntityRequest.fromPartial = fromPartial;
})(CurrentSceneEntityRequest || (exports.CurrentSceneEntityRequest = CurrentSceneEntityRequest = {}));
function createBaseCurrentSceneEntityResponse() {
    return { urn: "", content: [], metadataJson: "", baseUrl: "" };
}
var CurrentSceneEntityResponse;
(function (CurrentSceneEntityResponse) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.urn !== "") {
            writer.uint32(10).string(message.urn);
        }
        for (const v of message.content) {
            content_mapping_gen_1.ContentMapping.encode(v, writer.uint32(18).fork()).ldelim();
        }
        if (message.metadataJson !== "") {
            writer.uint32(26).string(message.metadataJson);
        }
        if (message.baseUrl !== "") {
            writer.uint32(34).string(message.baseUrl);
        }
        return writer;
    }
    CurrentSceneEntityResponse.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCurrentSceneEntityResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.urn = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.content.push(content_mapping_gen_1.ContentMapping.decode(reader, reader.uint32()));
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.metadataJson = reader.string();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.baseUrl = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    CurrentSceneEntityResponse.decode = decode;
    function fromJSON(object) {
        return {
            urn: isSet(object.urn) ? String(object.urn) : "",
            content: Array.isArray(object === null || object === void 0 ? void 0 : object.content) ? object.content.map((e) => content_mapping_gen_1.ContentMapping.fromJSON(e)) : [],
            metadataJson: isSet(object.metadataJson) ? String(object.metadataJson) : "",
            baseUrl: isSet(object.baseUrl) ? String(object.baseUrl) : "",
        };
    }
    CurrentSceneEntityResponse.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.urn !== undefined && (obj.urn = message.urn);
        if (message.content) {
            obj.content = message.content.map((e) => e ? content_mapping_gen_1.ContentMapping.toJSON(e) : undefined);
        }
        else {
            obj.content = [];
        }
        message.metadataJson !== undefined && (obj.metadataJson = message.metadataJson);
        message.baseUrl !== undefined && (obj.baseUrl = message.baseUrl);
        return obj;
    }
    CurrentSceneEntityResponse.toJSON = toJSON;
    function create(base) {
        return CurrentSceneEntityResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    CurrentSceneEntityResponse.create = create;
    function fromPartial(object) {
        var _a, _b, _c, _d;
        const message = createBaseCurrentSceneEntityResponse();
        message.urn = (_a = object.urn) !== null && _a !== void 0 ? _a : "";
        message.content = ((_b = object.content) === null || _b === void 0 ? void 0 : _b.map((e) => content_mapping_gen_1.ContentMapping.fromPartial(e))) || [];
        message.metadataJson = (_c = object.metadataJson) !== null && _c !== void 0 ? _c : "";
        message.baseUrl = (_d = object.baseUrl) !== null && _d !== void 0 ? _d : "";
        return message;
    }
    CurrentSceneEntityResponse.fromPartial = fromPartial;
})(CurrentSceneEntityResponse || (exports.CurrentSceneEntityResponse = CurrentSceneEntityResponse = {}));
function createBaseGetExplorerInformationRequest() {
    return {};
}
var GetExplorerInformationRequest;
(function (GetExplorerInformationRequest) {
    function encode(_, writer = minimal_1.default.Writer.create()) {
        return writer;
    }
    GetExplorerInformationRequest.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetExplorerInformationRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    GetExplorerInformationRequest.decode = decode;
    function fromJSON(_) {
        return {};
    }
    GetExplorerInformationRequest.fromJSON = fromJSON;
    function toJSON(_) {
        const obj = {};
        return obj;
    }
    GetExplorerInformationRequest.toJSON = toJSON;
    function create(base) {
        return GetExplorerInformationRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    GetExplorerInformationRequest.create = create;
    function fromPartial(_) {
        const message = createBaseGetExplorerInformationRequest();
        return message;
    }
    GetExplorerInformationRequest.fromPartial = fromPartial;
})(GetExplorerInformationRequest || (exports.GetExplorerInformationRequest = GetExplorerInformationRequest = {}));
function createBaseGetExplorerInformationResponse() {
    return { agent: "", platform: "", configurations: {} };
}
var GetExplorerInformationResponse;
(function (GetExplorerInformationResponse) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.agent !== "") {
            writer.uint32(10).string(message.agent);
        }
        if (message.platform !== "") {
            writer.uint32(18).string(message.platform);
        }
        Object.entries(message.configurations).forEach(([key, value]) => {
            GetExplorerInformationResponse_ConfigurationsEntry.encode({ key: key, value }, writer.uint32(26).fork())
                .ldelim();
        });
        return writer;
    }
    GetExplorerInformationResponse.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetExplorerInformationResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.agent = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.platform = reader.string();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    const entry3 = GetExplorerInformationResponse_ConfigurationsEntry.decode(reader, reader.uint32());
                    if (entry3.value !== undefined) {
                        message.configurations[entry3.key] = entry3.value;
                    }
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    GetExplorerInformationResponse.decode = decode;
    function fromJSON(object) {
        return {
            agent: isSet(object.agent) ? String(object.agent) : "",
            platform: isSet(object.platform) ? String(object.platform) : "",
            configurations: isObject(object.configurations)
                ? Object.entries(object.configurations).reduce((acc, [key, value]) => {
                    acc[key] = String(value);
                    return acc;
                }, {})
                : {},
        };
    }
    GetExplorerInformationResponse.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.agent !== undefined && (obj.agent = message.agent);
        message.platform !== undefined && (obj.platform = message.platform);
        obj.configurations = {};
        if (message.configurations) {
            Object.entries(message.configurations).forEach(([k, v]) => {
                obj.configurations[k] = v;
            });
        }
        return obj;
    }
    GetExplorerInformationResponse.toJSON = toJSON;
    function create(base) {
        return GetExplorerInformationResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    GetExplorerInformationResponse.create = create;
    function fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseGetExplorerInformationResponse();
        message.agent = (_a = object.agent) !== null && _a !== void 0 ? _a : "";
        message.platform = (_b = object.platform) !== null && _b !== void 0 ? _b : "";
        message.configurations = Object.entries((_c = object.configurations) !== null && _c !== void 0 ? _c : {}).reduce((acc, [key, value]) => {
            if (value !== undefined) {
                acc[key] = String(value);
            }
            return acc;
        }, {});
        return message;
    }
    GetExplorerInformationResponse.fromPartial = fromPartial;
})(GetExplorerInformationResponse || (exports.GetExplorerInformationResponse = GetExplorerInformationResponse = {}));
function createBaseGetExplorerInformationResponse_ConfigurationsEntry() {
    return { key: "", value: "" };
}
var GetExplorerInformationResponse_ConfigurationsEntry;
(function (GetExplorerInformationResponse_ConfigurationsEntry) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        if (message.value !== "") {
            writer.uint32(18).string(message.value);
        }
        return writer;
    }
    GetExplorerInformationResponse_ConfigurationsEntry.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetExplorerInformationResponse_ConfigurationsEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.key = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.value = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    GetExplorerInformationResponse_ConfigurationsEntry.decode = decode;
    function fromJSON(object) {
        return { key: isSet(object.key) ? String(object.key) : "", value: isSet(object.value) ? String(object.value) : "" };
    }
    GetExplorerInformationResponse_ConfigurationsEntry.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.key !== undefined && (obj.key = message.key);
        message.value !== undefined && (obj.value = message.value);
        return obj;
    }
    GetExplorerInformationResponse_ConfigurationsEntry.toJSON = toJSON;
    function create(base) {
        return GetExplorerInformationResponse_ConfigurationsEntry.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    GetExplorerInformationResponse_ConfigurationsEntry.create = create;
    function fromPartial(object) {
        var _a, _b;
        const message = createBaseGetExplorerInformationResponse_ConfigurationsEntry();
        message.key = (_a = object.key) !== null && _a !== void 0 ? _a : "";
        message.value = (_b = object.value) !== null && _b !== void 0 ? _b : "";
        return message;
    }
    GetExplorerInformationResponse_ConfigurationsEntry.fromPartial = fromPartial;
})(GetExplorerInformationResponse_ConfigurationsEntry || (exports.GetExplorerInformationResponse_ConfigurationsEntry = GetExplorerInformationResponse_ConfigurationsEntry = {}));
exports.RuntimeServiceDefinition = {
    name: "RuntimeService",
    fullName: "decentraland.kernel.apis.RuntimeService",
    methods: {
        /** Provides information about the current realm */
        getRealm: {
            name: "GetRealm",
            requestType: GetRealmRequest,
            requestStream: false,
            responseType: GetRealmResponse,
            responseStream: false,
            options: {},
        },
        /**
         * Provides information about the Decentraland Time, which is coordinated
         * across players.
         */
        getWorldTime: {
            name: "GetWorldTime",
            requestType: GetWorldTimeRequest,
            requestStream: false,
            responseType: GetWorldTimeResponse,
            responseStream: false,
            options: {},
        },
        /**
         * Returns the file content of a deployed asset. If the file doesn't
         * exist or cannot be retrieved, the RPC call throws an error.
         * This method is called to load any assets deployed among the scene,
         * runtime may cache this response much more than the provided "fetch" function.
         */
        readFile: {
            name: "ReadFile",
            requestType: ReadFileRequest,
            requestStream: false,
            responseType: ReadFileResponse,
            responseStream: false,
            options: {},
        },
        /** Returns information about the current scene. This is the replacement of GetBootstrapData */
        getSceneInformation: {
            name: "GetSceneInformation",
            requestType: CurrentSceneEntityRequest,
            requestStream: false,
            responseType: CurrentSceneEntityResponse,
            responseStream: false,
            options: {},
        },
        /** Provides information about the explorer */
        getExplorerInformation: {
            name: "GetExplorerInformation",
            requestType: GetExplorerInformationRequest,
            requestStream: false,
            responseType: GetExplorerInformationResponse,
            responseStream: false,
            options: {},
        },
    },
};
const tsProtoGlobalThis = (() => {
    if (typeof globalThis !== "undefined") {
        return globalThis;
    }
    if (typeof self !== "undefined") {
        return self;
    }
    if (typeof window !== "undefined") {
        return window;
    }
    if (typeof global !== "undefined") {
        return global;
    }
    throw "Unable to locate global object";
})();
function bytesFromBase64(b64) {
    if (tsProtoGlobalThis.Buffer) {
        return Uint8Array.from(tsProtoGlobalThis.Buffer.from(b64, "base64"));
    }
    else {
        const bin = tsProtoGlobalThis.atob(b64);
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (tsProtoGlobalThis.Buffer) {
        return tsProtoGlobalThis.Buffer.from(arr).toString("base64");
    }
    else {
        const bin = [];
        arr.forEach((byte) => {
            bin.push(String.fromCharCode(byte));
        });
        return tsProtoGlobalThis.btoa(bin.join(""));
    }
}
function isObject(value) {
    return typeof value === "object" && value !== null;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
