// @dcl/protocol/out-js/decentraland/kernel/apis/testing.gen.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestingServiceDefinition = exports.SetCameraTransformTestCommandResponse = exports.SetCameraTransformTestCommand_Quaternion = exports.SetCameraTransformTestCommand_Vector3 = exports.SetCameraTransformTestCommand = exports.TestPlanResponse = exports.TestPlan_TestPlanEntry = exports.TestPlan = exports.TestResultResponse = exports.TestResult = exports.TakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult = exports.TakeAndCompareScreenshotResponse = exports.TakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff = exports.TakeAndCompareScreenshotRequest = exports.TakeAndCompareScreenshotRequest_SnapshotMode = exports.protobufPackage = void 0;
exports.takeAndCompareScreenshotRequest_SnapshotModeFromJSON = takeAndCompareScreenshotRequest_SnapshotModeFromJSON;
exports.takeAndCompareScreenshotRequest_SnapshotModeToJSON = takeAndCompareScreenshotRequest_SnapshotModeToJSON;
/* eslint-disable */
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const vectors_gen_1 = require("../../common/vectors.gen");
exports.protobufPackage = "decentraland.kernel.apis";
var TakeAndCompareScreenshotRequest_SnapshotMode;
(function (TakeAndCompareScreenshotRequest_SnapshotMode) {
    /** SM_3D_AND_UI - only visible 3d scene and ui scene */
    TakeAndCompareScreenshotRequest_SnapshotMode[TakeAndCompareScreenshotRequest_SnapshotMode["SM_3D_AND_UI"] = 0] = "SM_3D_AND_UI";
    /** SM_3D_ONLY - only visible 3d scene */
    TakeAndCompareScreenshotRequest_SnapshotMode[TakeAndCompareScreenshotRequest_SnapshotMode["SM_3D_ONLY"] = 1] = "SM_3D_ONLY";
    /** SM_UI_ONLY - only visible ui scene */
    TakeAndCompareScreenshotRequest_SnapshotMode[TakeAndCompareScreenshotRequest_SnapshotMode["SM_UI_ONLY"] = 2] = "SM_UI_ONLY";
    TakeAndCompareScreenshotRequest_SnapshotMode[TakeAndCompareScreenshotRequest_SnapshotMode["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(TakeAndCompareScreenshotRequest_SnapshotMode || (exports.TakeAndCompareScreenshotRequest_SnapshotMode = TakeAndCompareScreenshotRequest_SnapshotMode = {}));
function takeAndCompareScreenshotRequest_SnapshotModeFromJSON(object) {
    switch (object) {
        case 0:
        case "SM_3D_AND_UI":
            return TakeAndCompareScreenshotRequest_SnapshotMode.SM_3D_AND_UI;
        case 1:
        case "SM_3D_ONLY":
            return TakeAndCompareScreenshotRequest_SnapshotMode.SM_3D_ONLY;
        case 2:
        case "SM_UI_ONLY":
            return TakeAndCompareScreenshotRequest_SnapshotMode.SM_UI_ONLY;
        case -1:
        case "UNRECOGNIZED":
        default:
            return TakeAndCompareScreenshotRequest_SnapshotMode.UNRECOGNIZED;
    }
}
function takeAndCompareScreenshotRequest_SnapshotModeToJSON(object) {
    switch (object) {
        case TakeAndCompareScreenshotRequest_SnapshotMode.SM_3D_AND_UI:
            return "SM_3D_AND_UI";
        case TakeAndCompareScreenshotRequest_SnapshotMode.SM_3D_ONLY:
            return "SM_3D_ONLY";
        case TakeAndCompareScreenshotRequest_SnapshotMode.SM_UI_ONLY:
            return "SM_UI_ONLY";
        case TakeAndCompareScreenshotRequest_SnapshotMode.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
function createBaseTakeAndCompareScreenshotRequest() {
    return {
        srcStoredSnapshot: "",
        cameraPosition: undefined,
        cameraTarget: undefined,
        screenshotSize: undefined,
        comparisonMethod: undefined,
        snapshotMode: 0,
    };
}
var TakeAndCompareScreenshotRequest;
(function (TakeAndCompareScreenshotRequest) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        var _a;
        if (message.srcStoredSnapshot !== "") {
            writer.uint32(10).string(message.srcStoredSnapshot);
        }
        if (message.cameraPosition !== undefined) {
            vectors_gen_1.Vector3.encode(message.cameraPosition, writer.uint32(18).fork()).ldelim();
        }
        if (message.cameraTarget !== undefined) {
            vectors_gen_1.Vector3.encode(message.cameraTarget, writer.uint32(26).fork()).ldelim();
        }
        if (message.screenshotSize !== undefined) {
            vectors_gen_1.Vector2.encode(message.screenshotSize, writer.uint32(34).fork()).ldelim();
        }
        switch ((_a = message.comparisonMethod) === null || _a === void 0 ? void 0 : _a.$case) {
            case "greyPixelDiff":
                TakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff.encode(message.comparisonMethod.greyPixelDiff, writer.uint32(42).fork()).ldelim();
                break;
        }
        if (message.snapshotMode !== 0) {
            writer.uint32(48).int32(message.snapshotMode);
        }
        return writer;
    }
    TakeAndCompareScreenshotRequest.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTakeAndCompareScreenshotRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.srcStoredSnapshot = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.cameraPosition = vectors_gen_1.Vector3.decode(reader, reader.uint32());
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.cameraTarget = vectors_gen_1.Vector3.decode(reader, reader.uint32());
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.screenshotSize = vectors_gen_1.Vector2.decode(reader, reader.uint32());
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.comparisonMethod = {
                        $case: "greyPixelDiff",
                        greyPixelDiff: TakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff.decode(reader, reader.uint32()),
                    };
                    continue;
                case 6:
                    if (tag !== 48) {
                        break;
                    }
                    message.snapshotMode = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    TakeAndCompareScreenshotRequest.decode = decode;
    function fromJSON(object) {
        return {
            srcStoredSnapshot: isSet(object.srcStoredSnapshot) ? String(object.srcStoredSnapshot) : "",
            cameraPosition: isSet(object.cameraPosition) ? vectors_gen_1.Vector3.fromJSON(object.cameraPosition) : undefined,
            cameraTarget: isSet(object.cameraTarget) ? vectors_gen_1.Vector3.fromJSON(object.cameraTarget) : undefined,
            screenshotSize: isSet(object.screenshotSize) ? vectors_gen_1.Vector2.fromJSON(object.screenshotSize) : undefined,
            comparisonMethod: isSet(object.greyPixelDiff)
                ? {
                    $case: "greyPixelDiff",
                    greyPixelDiff: TakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff.fromJSON(object.greyPixelDiff),
                }
                : undefined,
            snapshotMode: isSet(object.snapshotMode)
                ? takeAndCompareScreenshotRequest_SnapshotModeFromJSON(object.snapshotMode)
                : 0,
        };
    }
    TakeAndCompareScreenshotRequest.fromJSON = fromJSON;
    function toJSON(message) {
        var _a, _b, _c;
        const obj = {};
        message.srcStoredSnapshot !== undefined && (obj.srcStoredSnapshot = message.srcStoredSnapshot);
        message.cameraPosition !== undefined &&
            (obj.cameraPosition = message.cameraPosition ? vectors_gen_1.Vector3.toJSON(message.cameraPosition) : undefined);
        message.cameraTarget !== undefined &&
            (obj.cameraTarget = message.cameraTarget ? vectors_gen_1.Vector3.toJSON(message.cameraTarget) : undefined);
        message.screenshotSize !== undefined &&
            (obj.screenshotSize = message.screenshotSize ? vectors_gen_1.Vector2.toJSON(message.screenshotSize) : undefined);
        ((_a = message.comparisonMethod) === null || _a === void 0 ? void 0 : _a.$case) === "greyPixelDiff" &&
            (obj.greyPixelDiff = ((_b = message.comparisonMethod) === null || _b === void 0 ? void 0 : _b.greyPixelDiff)
                ? TakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff.toJSON((_c = message.comparisonMethod) === null || _c === void 0 ? void 0 : _c.greyPixelDiff)
                : undefined);
        message.snapshotMode !== undefined &&
            (obj.snapshotMode = takeAndCompareScreenshotRequest_SnapshotModeToJSON(message.snapshotMode));
        return obj;
    }
    TakeAndCompareScreenshotRequest.toJSON = toJSON;
    function create(base) {
        return TakeAndCompareScreenshotRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    TakeAndCompareScreenshotRequest.create = create;
    function fromPartial(object) {
        var _a, _b, _c, _d, _e;
        const message = createBaseTakeAndCompareScreenshotRequest();
        message.srcStoredSnapshot = (_a = object.srcStoredSnapshot) !== null && _a !== void 0 ? _a : "";
        message.cameraPosition = (object.cameraPosition !== undefined && object.cameraPosition !== null)
            ? vectors_gen_1.Vector3.fromPartial(object.cameraPosition)
            : undefined;
        message.cameraTarget = (object.cameraTarget !== undefined && object.cameraTarget !== null)
            ? vectors_gen_1.Vector3.fromPartial(object.cameraTarget)
            : undefined;
        message.screenshotSize = (object.screenshotSize !== undefined && object.screenshotSize !== null)
            ? vectors_gen_1.Vector2.fromPartial(object.screenshotSize)
            : undefined;
        if (((_b = object.comparisonMethod) === null || _b === void 0 ? void 0 : _b.$case) === "greyPixelDiff" &&
            ((_c = object.comparisonMethod) === null || _c === void 0 ? void 0 : _c.greyPixelDiff) !== undefined &&
            ((_d = object.comparisonMethod) === null || _d === void 0 ? void 0 : _d.greyPixelDiff) !== null) {
            message.comparisonMethod = {
                $case: "greyPixelDiff",
                greyPixelDiff: TakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff.fromPartial(object.comparisonMethod.greyPixelDiff),
            };
        }
        message.snapshotMode = (_e = object.snapshotMode) !== null && _e !== void 0 ? _e : 0;
        return message;
    }
    TakeAndCompareScreenshotRequest.fromPartial = fromPartial;
})(TakeAndCompareScreenshotRequest || (exports.TakeAndCompareScreenshotRequest = TakeAndCompareScreenshotRequest = {}));
function createBaseTakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff() {
    return {};
}
var TakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff;
(function (TakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff) {
    function encode(_, writer = minimal_1.default.Writer.create()) {
        return writer;
    }
    TakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    TakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff.decode = decode;
    function fromJSON(_) {
        return {};
    }
    TakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff.fromJSON = fromJSON;
    function toJSON(_) {
        const obj = {};
        return obj;
    }
    TakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff.toJSON = toJSON;
    function create(base) {
        return TakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    TakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff.create = create;
    function fromPartial(_) {
        const message = createBaseTakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff();
        return message;
    }
    TakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff.fromPartial = fromPartial;
})(TakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff || (exports.TakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff = TakeAndCompareScreenshotRequest_ComparisonMethodGreyPixelDiff = {}));
function createBaseTakeAndCompareScreenshotResponse() {
    return { storedSnapshotFound: false, comparisonMethodResult: undefined };
}
var TakeAndCompareScreenshotResponse;
(function (TakeAndCompareScreenshotResponse) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        var _a;
        if (message.storedSnapshotFound === true) {
            writer.uint32(8).bool(message.storedSnapshotFound);
        }
        switch ((_a = message.comparisonMethodResult) === null || _a === void 0 ? void 0 : _a.$case) {
            case "greyPixelDiff":
                TakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult.encode(message.comparisonMethodResult.greyPixelDiff, writer.uint32(18).fork()).ldelim();
                break;
        }
        return writer;
    }
    TakeAndCompareScreenshotResponse.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTakeAndCompareScreenshotResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.storedSnapshotFound = reader.bool();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.comparisonMethodResult = {
                        $case: "greyPixelDiff",
                        greyPixelDiff: TakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult.decode(reader, reader.uint32()),
                    };
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    TakeAndCompareScreenshotResponse.decode = decode;
    function fromJSON(object) {
        return {
            storedSnapshotFound: isSet(object.storedSnapshotFound) ? Boolean(object.storedSnapshotFound) : false,
            comparisonMethodResult: isSet(object.greyPixelDiff)
                ? {
                    $case: "greyPixelDiff",
                    greyPixelDiff: TakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult.fromJSON(object.greyPixelDiff),
                }
                : undefined,
        };
    }
    TakeAndCompareScreenshotResponse.fromJSON = fromJSON;
    function toJSON(message) {
        var _a, _b, _c;
        const obj = {};
        message.storedSnapshotFound !== undefined && (obj.storedSnapshotFound = message.storedSnapshotFound);
        ((_a = message.comparisonMethodResult) === null || _a === void 0 ? void 0 : _a.$case) === "greyPixelDiff" &&
            (obj.greyPixelDiff = ((_b = message.comparisonMethodResult) === null || _b === void 0 ? void 0 : _b.greyPixelDiff)
                ? TakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult.toJSON((_c = message.comparisonMethodResult) === null || _c === void 0 ? void 0 : _c.greyPixelDiff)
                : undefined);
        return obj;
    }
    TakeAndCompareScreenshotResponse.toJSON = toJSON;
    function create(base) {
        return TakeAndCompareScreenshotResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    TakeAndCompareScreenshotResponse.create = create;
    function fromPartial(object) {
        var _a, _b, _c, _d;
        const message = createBaseTakeAndCompareScreenshotResponse();
        message.storedSnapshotFound = (_a = object.storedSnapshotFound) !== null && _a !== void 0 ? _a : false;
        if (((_b = object.comparisonMethodResult) === null || _b === void 0 ? void 0 : _b.$case) === "greyPixelDiff" &&
            ((_c = object.comparisonMethodResult) === null || _c === void 0 ? void 0 : _c.greyPixelDiff) !== undefined &&
            ((_d = object.comparisonMethodResult) === null || _d === void 0 ? void 0 : _d.greyPixelDiff) !== null) {
            message.comparisonMethodResult = {
                $case: "greyPixelDiff",
                greyPixelDiff: TakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult.fromPartial(object.comparisonMethodResult.greyPixelDiff),
            };
        }
        return message;
    }
    TakeAndCompareScreenshotResponse.fromPartial = fromPartial;
})(TakeAndCompareScreenshotResponse || (exports.TakeAndCompareScreenshotResponse = TakeAndCompareScreenshotResponse = {}));
function createBaseTakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult() {
    return { similarity: 0 };
}
var TakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult;
(function (TakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.similarity !== 0) {
            writer.uint32(13).float(message.similarity);
        }
        return writer;
    }
    TakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 13) {
                        break;
                    }
                    message.similarity = reader.float();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    TakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult.decode = decode;
    function fromJSON(object) {
        return { similarity: isSet(object.similarity) ? Number(object.similarity) : 0 };
    }
    TakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.similarity !== undefined && (obj.similarity = message.similarity);
        return obj;
    }
    TakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult.toJSON = toJSON;
    function create(base) {
        return TakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    TakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseTakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult();
        message.similarity = (_a = object.similarity) !== null && _a !== void 0 ? _a : 0;
        return message;
    }
    TakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult.fromPartial = fromPartial;
})(TakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult || (exports.TakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult = TakeAndCompareScreenshotResponse_ComparisonMethodGreyPixelDiffResult = {}));
function createBaseTestResult() {
    return { name: "", ok: false, error: undefined, stack: undefined, totalFrames: 0, totalTime: 0 };
}
var TestResult;
(function (TestResult) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.name !== "") {
            writer.uint32(10).string(message.name);
        }
        if (message.ok === true) {
            writer.uint32(16).bool(message.ok);
        }
        if (message.error !== undefined) {
            writer.uint32(26).string(message.error);
        }
        if (message.stack !== undefined) {
            writer.uint32(34).string(message.stack);
        }
        if (message.totalFrames !== 0) {
            writer.uint32(40).uint32(message.totalFrames);
        }
        if (message.totalTime !== 0) {
            writer.uint32(53).float(message.totalTime);
        }
        return writer;
    }
    TestResult.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTestResult();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.ok = reader.bool();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.error = reader.string();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.stack = reader.string();
                    continue;
                case 5:
                    if (tag !== 40) {
                        break;
                    }
                    message.totalFrames = reader.uint32();
                    continue;
                case 6:
                    if (tag !== 53) {
                        break;
                    }
                    message.totalTime = reader.float();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    TestResult.decode = decode;
    function fromJSON(object) {
        return {
            name: isSet(object.name) ? String(object.name) : "",
            ok: isSet(object.ok) ? Boolean(object.ok) : false,
            error: isSet(object.error) ? String(object.error) : undefined,
            stack: isSet(object.stack) ? String(object.stack) : undefined,
            totalFrames: isSet(object.totalFrames) ? Number(object.totalFrames) : 0,
            totalTime: isSet(object.totalTime) ? Number(object.totalTime) : 0,
        };
    }
    TestResult.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.name !== undefined && (obj.name = message.name);
        message.ok !== undefined && (obj.ok = message.ok);
        message.error !== undefined && (obj.error = message.error);
        message.stack !== undefined && (obj.stack = message.stack);
        message.totalFrames !== undefined && (obj.totalFrames = Math.round(message.totalFrames));
        message.totalTime !== undefined && (obj.totalTime = message.totalTime);
        return obj;
    }
    TestResult.toJSON = toJSON;
    function create(base) {
        return TestResult.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    TestResult.create = create;
    function fromPartial(object) {
        var _a, _b, _c, _d, _e, _f;
        const message = createBaseTestResult();
        message.name = (_a = object.name) !== null && _a !== void 0 ? _a : "";
        message.ok = (_b = object.ok) !== null && _b !== void 0 ? _b : false;
        message.error = (_c = object.error) !== null && _c !== void 0 ? _c : undefined;
        message.stack = (_d = object.stack) !== null && _d !== void 0 ? _d : undefined;
        message.totalFrames = (_e = object.totalFrames) !== null && _e !== void 0 ? _e : 0;
        message.totalTime = (_f = object.totalTime) !== null && _f !== void 0 ? _f : 0;
        return message;
    }
    TestResult.fromPartial = fromPartial;
})(TestResult || (exports.TestResult = TestResult = {}));
function createBaseTestResultResponse() {
    return {};
}
var TestResultResponse;
(function (TestResultResponse) {
    function encode(_, writer = minimal_1.default.Writer.create()) {
        return writer;
    }
    TestResultResponse.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTestResultResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    TestResultResponse.decode = decode;
    function fromJSON(_) {
        return {};
    }
    TestResultResponse.fromJSON = fromJSON;
    function toJSON(_) {
        const obj = {};
        return obj;
    }
    TestResultResponse.toJSON = toJSON;
    function create(base) {
        return TestResultResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    TestResultResponse.create = create;
    function fromPartial(_) {
        const message = createBaseTestResultResponse();
        return message;
    }
    TestResultResponse.fromPartial = fromPartial;
})(TestResultResponse || (exports.TestResultResponse = TestResultResponse = {}));
function createBaseTestPlan() {
    return { tests: [] };
}
var TestPlan;
(function (TestPlan) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.tests) {
            TestPlan_TestPlanEntry.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    }
    TestPlan.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTestPlan();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.tests.push(TestPlan_TestPlanEntry.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    TestPlan.decode = decode;
    function fromJSON(object) {
        return {
            tests: Array.isArray(object === null || object === void 0 ? void 0 : object.tests) ? object.tests.map((e) => TestPlan_TestPlanEntry.fromJSON(e)) : [],
        };
    }
    TestPlan.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        if (message.tests) {
            obj.tests = message.tests.map((e) => e ? TestPlan_TestPlanEntry.toJSON(e) : undefined);
        }
        else {
            obj.tests = [];
        }
        return obj;
    }
    TestPlan.toJSON = toJSON;
    function create(base) {
        return TestPlan.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    TestPlan.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseTestPlan();
        message.tests = ((_a = object.tests) === null || _a === void 0 ? void 0 : _a.map((e) => TestPlan_TestPlanEntry.fromPartial(e))) || [];
        return message;
    }
    TestPlan.fromPartial = fromPartial;
})(TestPlan || (exports.TestPlan = TestPlan = {}));
function createBaseTestPlan_TestPlanEntry() {
    return { name: "" };
}
var TestPlan_TestPlanEntry;
(function (TestPlan_TestPlanEntry) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.name !== "") {
            writer.uint32(10).string(message.name);
        }
        return writer;
    }
    TestPlan_TestPlanEntry.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTestPlan_TestPlanEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    TestPlan_TestPlanEntry.decode = decode;
    function fromJSON(object) {
        return { name: isSet(object.name) ? String(object.name) : "" };
    }
    TestPlan_TestPlanEntry.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.name !== undefined && (obj.name = message.name);
        return obj;
    }
    TestPlan_TestPlanEntry.toJSON = toJSON;
    function create(base) {
        return TestPlan_TestPlanEntry.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    TestPlan_TestPlanEntry.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseTestPlan_TestPlanEntry();
        message.name = (_a = object.name) !== null && _a !== void 0 ? _a : "";
        return message;
    }
    TestPlan_TestPlanEntry.fromPartial = fromPartial;
})(TestPlan_TestPlanEntry || (exports.TestPlan_TestPlanEntry = TestPlan_TestPlanEntry = {}));
function createBaseTestPlanResponse() {
    return {};
}
var TestPlanResponse;
(function (TestPlanResponse) {
    function encode(_, writer = minimal_1.default.Writer.create()) {
        return writer;
    }
    TestPlanResponse.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTestPlanResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    TestPlanResponse.decode = decode;
    function fromJSON(_) {
        return {};
    }
    TestPlanResponse.fromJSON = fromJSON;
    function toJSON(_) {
        const obj = {};
        return obj;
    }
    TestPlanResponse.toJSON = toJSON;
    function create(base) {
        return TestPlanResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    TestPlanResponse.create = create;
    function fromPartial(_) {
        const message = createBaseTestPlanResponse();
        return message;
    }
    TestPlanResponse.fromPartial = fromPartial;
})(TestPlanResponse || (exports.TestPlanResponse = TestPlanResponse = {}));
function createBaseSetCameraTransformTestCommand() {
    return { position: undefined, rotation: undefined };
}
var SetCameraTransformTestCommand;
(function (SetCameraTransformTestCommand) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.position !== undefined) {
            SetCameraTransformTestCommand_Vector3.encode(message.position, writer.uint32(10).fork()).ldelim();
        }
        if (message.rotation !== undefined) {
            SetCameraTransformTestCommand_Quaternion.encode(message.rotation, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    }
    SetCameraTransformTestCommand.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSetCameraTransformTestCommand();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.position = SetCameraTransformTestCommand_Vector3.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.rotation = SetCameraTransformTestCommand_Quaternion.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    SetCameraTransformTestCommand.decode = decode;
    function fromJSON(object) {
        return {
            position: isSet(object.position) ? SetCameraTransformTestCommand_Vector3.fromJSON(object.position) : undefined,
            rotation: isSet(object.rotation) ? SetCameraTransformTestCommand_Quaternion.fromJSON(object.rotation) : undefined,
        };
    }
    SetCameraTransformTestCommand.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.position !== undefined &&
            (obj.position = message.position ? SetCameraTransformTestCommand_Vector3.toJSON(message.position) : undefined);
        message.rotation !== undefined &&
            (obj.rotation = message.rotation ? SetCameraTransformTestCommand_Quaternion.toJSON(message.rotation) : undefined);
        return obj;
    }
    SetCameraTransformTestCommand.toJSON = toJSON;
    function create(base) {
        return SetCameraTransformTestCommand.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    SetCameraTransformTestCommand.create = create;
    function fromPartial(object) {
        const message = createBaseSetCameraTransformTestCommand();
        message.position = (object.position !== undefined && object.position !== null)
            ? SetCameraTransformTestCommand_Vector3.fromPartial(object.position)
            : undefined;
        message.rotation = (object.rotation !== undefined && object.rotation !== null)
            ? SetCameraTransformTestCommand_Quaternion.fromPartial(object.rotation)
            : undefined;
        return message;
    }
    SetCameraTransformTestCommand.fromPartial = fromPartial;
})(SetCameraTransformTestCommand || (exports.SetCameraTransformTestCommand = SetCameraTransformTestCommand = {}));
function createBaseSetCameraTransformTestCommand_Vector3() {
    return { x: 0, y: 0, z: 0 };
}
var SetCameraTransformTestCommand_Vector3;
(function (SetCameraTransformTestCommand_Vector3) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.x !== 0) {
            writer.uint32(13).float(message.x);
        }
        if (message.y !== 0) {
            writer.uint32(21).float(message.y);
        }
        if (message.z !== 0) {
            writer.uint32(29).float(message.z);
        }
        return writer;
    }
    SetCameraTransformTestCommand_Vector3.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSetCameraTransformTestCommand_Vector3();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 13) {
                        break;
                    }
                    message.x = reader.float();
                    continue;
                case 2:
                    if (tag !== 21) {
                        break;
                    }
                    message.y = reader.float();
                    continue;
                case 3:
                    if (tag !== 29) {
                        break;
                    }
                    message.z = reader.float();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    SetCameraTransformTestCommand_Vector3.decode = decode;
    function fromJSON(object) {
        return {
            x: isSet(object.x) ? Number(object.x) : 0,
            y: isSet(object.y) ? Number(object.y) : 0,
            z: isSet(object.z) ? Number(object.z) : 0,
        };
    }
    SetCameraTransformTestCommand_Vector3.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.x !== undefined && (obj.x = message.x);
        message.y !== undefined && (obj.y = message.y);
        message.z !== undefined && (obj.z = message.z);
        return obj;
    }
    SetCameraTransformTestCommand_Vector3.toJSON = toJSON;
    function create(base) {
        return SetCameraTransformTestCommand_Vector3.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    SetCameraTransformTestCommand_Vector3.create = create;
    function fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseSetCameraTransformTestCommand_Vector3();
        message.x = (_a = object.x) !== null && _a !== void 0 ? _a : 0;
        message.y = (_b = object.y) !== null && _b !== void 0 ? _b : 0;
        message.z = (_c = object.z) !== null && _c !== void 0 ? _c : 0;
        return message;
    }
    SetCameraTransformTestCommand_Vector3.fromPartial = fromPartial;
})(SetCameraTransformTestCommand_Vector3 || (exports.SetCameraTransformTestCommand_Vector3 = SetCameraTransformTestCommand_Vector3 = {}));
function createBaseSetCameraTransformTestCommand_Quaternion() {
    return { x: 0, y: 0, z: 0, w: 0 };
}
var SetCameraTransformTestCommand_Quaternion;
(function (SetCameraTransformTestCommand_Quaternion) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.x !== 0) {
            writer.uint32(13).float(message.x);
        }
        if (message.y !== 0) {
            writer.uint32(21).float(message.y);
        }
        if (message.z !== 0) {
            writer.uint32(29).float(message.z);
        }
        if (message.w !== 0) {
            writer.uint32(37).float(message.w);
        }
        return writer;
    }
    SetCameraTransformTestCommand_Quaternion.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSetCameraTransformTestCommand_Quaternion();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 13) {
                        break;
                    }
                    message.x = reader.float();
                    continue;
                case 2:
                    if (tag !== 21) {
                        break;
                    }
                    message.y = reader.float();
                    continue;
                case 3:
                    if (tag !== 29) {
                        break;
                    }
                    message.z = reader.float();
                    continue;
                case 4:
                    if (tag !== 37) {
                        break;
                    }
                    message.w = reader.float();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    SetCameraTransformTestCommand_Quaternion.decode = decode;
    function fromJSON(object) {
        return {
            x: isSet(object.x) ? Number(object.x) : 0,
            y: isSet(object.y) ? Number(object.y) : 0,
            z: isSet(object.z) ? Number(object.z) : 0,
            w: isSet(object.w) ? Number(object.w) : 0,
        };
    }
    SetCameraTransformTestCommand_Quaternion.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.x !== undefined && (obj.x = message.x);
        message.y !== undefined && (obj.y = message.y);
        message.z !== undefined && (obj.z = message.z);
        message.w !== undefined && (obj.w = message.w);
        return obj;
    }
    SetCameraTransformTestCommand_Quaternion.toJSON = toJSON;
    function create(base) {
        return SetCameraTransformTestCommand_Quaternion.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    SetCameraTransformTestCommand_Quaternion.create = create;
    function fromPartial(object) {
        var _a, _b, _c, _d;
        const message = createBaseSetCameraTransformTestCommand_Quaternion();
        message.x = (_a = object.x) !== null && _a !== void 0 ? _a : 0;
        message.y = (_b = object.y) !== null && _b !== void 0 ? _b : 0;
        message.z = (_c = object.z) !== null && _c !== void 0 ? _c : 0;
        message.w = (_d = object.w) !== null && _d !== void 0 ? _d : 0;
        return message;
    }
    SetCameraTransformTestCommand_Quaternion.fromPartial = fromPartial;
})(SetCameraTransformTestCommand_Quaternion || (exports.SetCameraTransformTestCommand_Quaternion = SetCameraTransformTestCommand_Quaternion = {}));
function createBaseSetCameraTransformTestCommandResponse() {
    return {};
}
var SetCameraTransformTestCommandResponse;
(function (SetCameraTransformTestCommandResponse) {
    function encode(_, writer = minimal_1.default.Writer.create()) {
        return writer;
    }
    SetCameraTransformTestCommandResponse.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSetCameraTransformTestCommandResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    SetCameraTransformTestCommandResponse.decode = decode;
    function fromJSON(_) {
        return {};
    }
    SetCameraTransformTestCommandResponse.fromJSON = fromJSON;
    function toJSON(_) {
        const obj = {};
        return obj;
    }
    SetCameraTransformTestCommandResponse.toJSON = toJSON;
    function create(base) {
        return SetCameraTransformTestCommandResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    SetCameraTransformTestCommandResponse.create = create;
    function fromPartial(_) {
        const message = createBaseSetCameraTransformTestCommandResponse();
        return message;
    }
    SetCameraTransformTestCommandResponse.fromPartial = fromPartial;
})(SetCameraTransformTestCommandResponse || (exports.SetCameraTransformTestCommandResponse = SetCameraTransformTestCommandResponse = {}));
exports.TestingServiceDefinition = {
    name: "TestingService",
    fullName: "decentraland.kernel.apis.TestingService",
    methods: {
        /** sends a test result to the test runner */
        logTestResult: {
            name: "LogTestResult",
            requestType: TestResult,
            requestStream: false,
            responseType: TestResultResponse,
            responseStream: false,
            options: {},
        },
        /** send a list of all planned tests to the test runner */
        plan: {
            name: "Plan",
            requestType: TestPlan,
            requestStream: false,
            responseType: TestPlanResponse,
            responseStream: false,
            options: {},
        },
        /** sets the camera position and rotation in the engine */
        setCameraTransform: {
            name: "SetCameraTransform",
            requestType: SetCameraTransformTestCommand,
            requestStream: false,
            responseType: SetCameraTransformTestCommandResponse,
            responseStream: false,
            options: {},
        },
        /**
         * @internal
         * take a screenshot and compare it with a stored one
         *  it hides the explorer hud and ui, primary player and players of avatar scenes
         */
        takeAndCompareScreenshot: {
            name: "TakeAndCompareScreenshot",
            requestType: TakeAndCompareScreenshotRequest,
            requestStream: false,
            responseType: TakeAndCompareScreenshotResponse,
            responseStream: false,
            options: {},
        },
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
