// @dcl/protocol/out-ts/decentraland/common/colors.gen.ts
/* eslint-disable */
import _m0 from "protobufjs/minimal";

export const protobufPackage = "decentraland.common";

export interface Color3 {
  r: number;
  g: number;
  b: number;
}

export interface Color4 {
  r: number;
  g: number;
  b: number;
  a: number;
}

function createBaseColor3(): Color3 {
  return { r: 0, g: 0, b: 0 };
}

export namespace Color3 {
  export function encode(message: Color3, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.r !== 0) {
      writer.uint32(13).float(message.r);
    }
    if (message.g !== 0) {
      writer.uint32(21).float(message.g);
    }
    if (message.b !== 0) {
      writer.uint32(29).float(message.b);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): Color3 {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseColor3();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 13) {
            break;
          }

          message.r = reader.float();
          continue;
        case 2:
          if (tag !== 21) {
            break;
          }

          message.g = reader.float();
          continue;
        case 3:
          if (tag !== 29) {
            break;
          }

          message.b = reader.float();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): Color3 {
    return {
      r: isSet(object.r) ? Number(object.r) : 0,
      g: isSet(object.g) ? Number(object.g) : 0,
      b: isSet(object.b) ? Number(object.b) : 0,
    };
  }

  export function toJSON(message: Color3): unknown {
    const obj: any = {};
    message.r !== undefined && (obj.r = message.r);
    message.g !== undefined && (obj.g = message.g);
    message.b !== undefined && (obj.b = message.b);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<Color3>, I>>(base?: I): Color3 {
    return Color3.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<Color3>, I>>(object: I): Color3 {
    const message = createBaseColor3();
    message.r = object.r ?? 0;
    message.g = object.g ?? 0;
    message.b = object.b ?? 0;
    return message;
  }
}

function createBaseColor4(): Color4 {
  return { r: 0, g: 0, b: 0, a: 0 };
}

export namespace Color4 {
  export function encode(message: Color4, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.r !== 0) {
      writer.uint32(13).float(message.r);
    }
    if (message.g !== 0) {
      writer.uint32(21).float(message.g);
    }
    if (message.b !== 0) {
      writer.uint32(29).float(message.b);
    }
    if (message.a !== 0) {
      writer.uint32(37).float(message.a);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): Color4 {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseColor4();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 13) {
            break;
          }

          message.r = reader.float();
          continue;
        case 2:
          if (tag !== 21) {
            break;
          }

          message.g = reader.float();
          continue;
        case 3:
          if (tag !== 29) {
            break;
          }

          message.b = reader.float();
          continue;
        case 4:
          if (tag !== 37) {
            break;
          }

          message.a = reader.float();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): Color4 {
    return {
      r: isSet(object.r) ? Number(object.r) : 0,
      g: isSet(object.g) ? Number(object.g) : 0,
      b: isSet(object.b) ? Number(object.b) : 0,
      a: isSet(object.a) ? Number(object.a) : 0,
    };
  }

  export function toJSON(message: Color4): unknown {
    const obj: any = {};
    message.r !== undefined && (obj.r = message.r);
    message.g !== undefined && (obj.g = message.g);
    message.b !== undefined && (obj.b = message.b);
    message.a !== undefined && (obj.a = message.a);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<Color4>, I>>(base?: I): Color4 {
    return Color4.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<Color4>, I>>(object: I): Color4 {
    const message = createBaseColor4();
    message.r = object.r ?? 0;
    message.g = object.g ?? 0;
    message.b = object.b ?? 0;
    message.a = object.a ?? 0;
    return message;
  }
}

type Builtin = Date | Function | Uint8Array | string | number | boolean | undefined;

export type DeepPartial<T> = T extends Builtin ? T
  : T extends Array<infer U> ? Array<DeepPartial<U>> : T extends ReadonlyArray<infer U> ? ReadonlyArray<DeepPartial<U>>
  : T extends { $case: string } ? { [K in keyof Omit<T, "$case">]?: DeepPartial<T[K]> } & { $case: T["$case"] }
  : T extends {} ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & { [K in Exclude<keyof I, KeysOfUnion<P>>]: never };

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
