// @dcl/quests-client/dist/protocol/google/protobuf/empty.gen.js
import _m0 from "protobufjs/minimal";
export const protobufPackage = "google.protobuf";
function createBaseEmpty() {
    return {};
}
export var Empty;
(function (Empty) {
    function encode(_, writer = _m0.Writer.create()) {
        return writer;
    }
    Empty.encode = encode;
    function decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseEmpty();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    Empty.decode = decode;
    function fromJSON(_) {
        return {};
    }
    Empty.fromJSON = fromJSON;
    function toJSON(_) {
        const obj = {};
        return obj;
    }
    Empty.toJSON = toJSON;
    function create(base) {
        return Empty.fromPartial(base ?? {});
    }
    Empty.create = create;
    function fromPartial(_) {
        const message = createBaseEmpty();
        return message;
    }
    Empty.fromPartial = fromPartial;
})(Empty || (Empty = {}));
