// @dcl/protocol/out-ts/decentraland/sdk/components/avatar_emote_command.gen.ts
/* eslint-disable */
import _m0 from "protobufjs/minimal";

export const protobufPackage = "decentraland.sdk.components";

/**
 * AvatarEmoteCommand is a grow only value set, used to signal the renderer about
 * avatar emotes playback.
 */
export interface PBAvatarEmoteCommand {
  emoteUrn: string;
  loop: boolean;
  /** monotonic counter */
  timestamp: number;
}

function createBasePBAvatarEmoteCommand(): PBAvatarEmoteCommand {
  return { emoteUrn: "", loop: false, timestamp: 0 };
}

export namespace PBAvatarEmoteCommand {
  export function encode(message: PBAvatarEmoteCommand, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.emoteUrn !== "") {
      writer.uint32(10).string(message.emoteUrn);
    }
    if (message.loop === true) {
      writer.uint32(16).bool(message.loop);
    }
    if (message.timestamp !== 0) {
      writer.uint32(24).uint32(message.timestamp);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): PBAvatarEmoteCommand {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBasePBAvatarEmoteCommand();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.emoteUrn = reader.string();
          continue;
        case 2:
          if (tag !== 16) {
            break;
          }

          message.loop = reader.bool();
          continue;
        case 3:
          if (tag !== 24) {
            break;
          }

          message.timestamp = reader.uint32();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): PBAvatarEmoteCommand {
    return {
      emoteUrn: isSet(object.emoteUrn) ? String(object.emoteUrn) : "",
      loop: isSet(object.loop) ? Boolean(object.loop) : false,
      timestamp: isSet(object.timestamp) ? Number(object.timestamp) : 0,
    };
  }

  export function toJSON(message: PBAvatarEmoteCommand): unknown {
    const obj: any = {};
    message.emoteUrn !== undefined && (obj.emoteUrn = message.emoteUrn);
    message.loop !== undefined && (obj.loop = message.loop);
    message.timestamp !== undefined && (obj.timestamp = Math.round(message.timestamp));
    return obj;
  }

  export function create<I extends Exact<DeepPartial<PBAvatarEmoteCommand>, I>>(base?: I): PBAvatarEmoteCommand {
    return PBAvatarEmoteCommand.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<PBAvatarEmoteCommand>, I>>(object: I): PBAvatarEmoteCommand {
    const message = createBasePBAvatarEmoteCommand();
    message.emoteUrn = object.emoteUrn ?? "";
    message.loop = object.loop ?? false;
    message.timestamp = object.timestamp ?? 0;
    return message;
  }
}

type Builtin = Date | Function | Uint8Array | string | number | boolean | undefined;

export type DeepPartial<T> = T extends Builtin ? T
  : T extends Array<infer U> ? Array<DeepPartial<U>> : T extends ReadonlyArray<infer U> ? ReadonlyArray<DeepPartial<U>>
  : T extends { $case: string } ? { [K in keyof Omit<T, "$case">]?: DeepPartial<T[K]> } & { $case: T["$case"] }
  : T extends {} ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & { [K in Exclude<keyof I, KeysOfUnion<P>>]: never };

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
