// @dcl/protocol/out-ts/decentraland/sdk/components/camera_mode.gen.ts
/* eslint-disable */
import _m0 from "protobufjs/minimal";
import { CameraType, cameraTypeFromJSON, cameraTypeToJSON } from "./common/camera_type.gen";

export const protobufPackage = "decentraland.sdk.components";

/**
 * The CameraMode component can be used to determine whether the player is using a first-person o
 * third-person view.
 */
export interface PBCameraMode {
  /** the camera type (1st-person or 3rd-person) */
  mode: CameraType;
}

function createBasePBCameraMode(): PBCameraMode {
  return { mode: 0 };
}

export namespace PBCameraMode {
  export function encode(message: PBCameraMode, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.mode !== 0) {
      writer.uint32(8).int32(message.mode);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): PBCameraMode {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBasePBCameraMode();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 8) {
            break;
          }

          message.mode = reader.int32() as any;
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): PBCameraMode {
    return { mode: isSet(object.mode) ? cameraTypeFromJSON(object.mode) : 0 };
  }

  export function toJSON(message: PBCameraMode): unknown {
    const obj: any = {};
    message.mode !== undefined && (obj.mode = cameraTypeToJSON(message.mode));
    return obj;
  }

  export function create<I extends Exact<DeepPartial<PBCameraMode>, I>>(base?: I): PBCameraMode {
    return PBCameraMode.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<PBCameraMode>, I>>(object: I): PBCameraMode {
    const message = createBasePBCameraMode();
    message.mode = object.mode ?? 0;
    return message;
  }
}

type Builtin = Date | Function | Uint8Array | string | number | boolean | undefined;

export type DeepPartial<T> = T extends Builtin ? T
  : T extends Array<infer U> ? Array<DeepPartial<U>> : T extends ReadonlyArray<infer U> ? ReadonlyArray<DeepPartial<U>>
  : T extends { $case: string } ? { [K in keyof Omit<T, "$case">]?: DeepPartial<T[K]> } & { $case: T["$case"] }
  : T extends {} ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & { [K in Exclude<keyof I, KeysOfUnion<P>>]: never };

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
