// @dcl/protocol/out-ts/decentraland/sdk/components/player_identity_data.gen.ts
/* eslint-disable */
import _m0 from "protobufjs/minimal";

export const protobufPackage = "decentraland.sdk.components";

/**
 * PlayerIdentityData is used to read the information about the avatar's
 * identity. this component is written by the engine using the communications
 * transports' data.
 */
export interface PBPlayerIdentityData {
  /** ethereum address of this player */
  address: string;
  isGuest: boolean;
}

function createBasePBPlayerIdentityData(): PBPlayerIdentityData {
  return { address: "", isGuest: false };
}

export namespace PBPlayerIdentityData {
  export function encode(message: PBPlayerIdentityData, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.address !== "") {
      writer.uint32(10).string(message.address);
    }
    if (message.isGuest === true) {
      writer.uint32(24).bool(message.isGuest);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): PBPlayerIdentityData {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBasePBPlayerIdentityData();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.address = reader.string();
          continue;
        case 3:
          if (tag !== 24) {
            break;
          }

          message.isGuest = reader.bool();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): PBPlayerIdentityData {
    return {
      address: isSet(object.address) ? String(object.address) : "",
      isGuest: isSet(object.isGuest) ? Boolean(object.isGuest) : false,
    };
  }

  export function toJSON(message: PBPlayerIdentityData): unknown {
    const obj: any = {};
    message.address !== undefined && (obj.address = message.address);
    message.isGuest !== undefined && (obj.isGuest = message.isGuest);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<PBPlayerIdentityData>, I>>(base?: I): PBPlayerIdentityData {
    return PBPlayerIdentityData.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<PBPlayerIdentityData>, I>>(object: I): PBPlayerIdentityData {
    const message = createBasePBPlayerIdentityData();
    message.address = object.address ?? "";
    message.isGuest = object.isGuest ?? false;
    return message;
  }
}

type Builtin = Date | Function | Uint8Array | string | number | boolean | undefined;

export type DeepPartial<T> = T extends Builtin ? T
  : T extends Array<infer U> ? Array<DeepPartial<U>> : T extends ReadonlyArray<infer U> ? ReadonlyArray<DeepPartial<U>>
  : T extends { $case: string } ? { [K in keyof Omit<T, "$case">]?: DeepPartial<T[K]> } & { $case: T["$case"] }
  : T extends {} ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & { [K in Exclude<keyof I, KeysOfUnion<P>>]: never };

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
