// @dcl/protocol/out-ts/decentraland/sdk/components/pointer_lock.gen.ts
/* eslint-disable */
import _m0 from "protobufjs/minimal";

export const protobufPackage = "decentraland.sdk.components";

/**
 * PointerLock indicates whether the mouse pointer is automatically following the camera’s point of
 * focus (locked), or can move freely on the screen (unlocked).
 */
export interface PBPointerLock {
  /** whether the pointer is locked */
  isPointerLocked: boolean;
}

function createBasePBPointerLock(): PBPointerLock {
  return { isPointerLocked: false };
}

export namespace PBPointerLock {
  export function encode(message: PBPointerLock, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.isPointerLocked === true) {
      writer.uint32(8).bool(message.isPointerLocked);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): PBPointerLock {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBasePBPointerLock();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 8) {
            break;
          }

          message.isPointerLocked = reader.bool();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): PBPointerLock {
    return { isPointerLocked: isSet(object.isPointerLocked) ? Boolean(object.isPointerLocked) : false };
  }

  export function toJSON(message: PBPointerLock): unknown {
    const obj: any = {};
    message.isPointerLocked !== undefined && (obj.isPointerLocked = message.isPointerLocked);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<PBPointerLock>, I>>(base?: I): PBPointerLock {
    return PBPointerLock.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<PBPointerLock>, I>>(object: I): PBPointerLock {
    const message = createBasePBPointerLock();
    message.isPointerLocked = object.isPointerLocked ?? false;
    return message;
  }
}

type Builtin = Date | Function | Uint8Array | string | number | boolean | undefined;

export type DeepPartial<T> = T extends Builtin ? T
  : T extends Array<infer U> ? Array<DeepPartial<U>> : T extends ReadonlyArray<infer U> ? ReadonlyArray<DeepPartial<U>>
  : T extends { $case: string } ? { [K in keyof Omit<T, "$case">]?: DeepPartial<T[K]> } & { $case: T["$case"] }
  : T extends {} ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & { [K in Exclude<keyof I, KeysOfUnion<P>>]: never };

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
