// @dcl/protocol/out-ts/decentraland/sdk/components/ui_dropdown_result.gen.ts
/* eslint-disable */
import _m0 from "protobufjs/minimal";

export const protobufPackage = "decentraland.sdk.components";

export interface PBUiDropdownResult {
  value: number;
}

function createBasePBUiDropdownResult(): PBUiDropdownResult {
  return { value: 0 };
}

export namespace PBUiDropdownResult {
  export function encode(message: PBUiDropdownResult, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.value !== 0) {
      writer.uint32(8).int32(message.value);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): PBUiDropdownResult {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBasePBUiDropdownResult();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 8) {
            break;
          }

          message.value = reader.int32();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): PBUiDropdownResult {
    return { value: isSet(object.value) ? Number(object.value) : 0 };
  }

  export function toJSON(message: PBUiDropdownResult): unknown {
    const obj: any = {};
    message.value !== undefined && (obj.value = Math.round(message.value));
    return obj;
  }

  export function create<I extends Exact<DeepPartial<PBUiDropdownResult>, I>>(base?: I): PBUiDropdownResult {
    return PBUiDropdownResult.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<PBUiDropdownResult>, I>>(object: I): PBUiDropdownResult {
    const message = createBasePBUiDropdownResult();
    message.value = object.value ?? 0;
    return message;
  }
}

type Builtin = Date | Function | Uint8Array | string | number | boolean | undefined;

export type DeepPartial<T> = T extends Builtin ? T
  : T extends Array<infer U> ? Array<DeepPartial<U>> : T extends ReadonlyArray<infer U> ? ReadonlyArray<DeepPartial<U>>
  : T extends { $case: string } ? { [K in keyof Omit<T, "$case">]?: DeepPartial<T[K]> } & { $case: T["$case"] }
  : T extends {} ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & { [K in Exclude<keyof I, KeysOfUnion<P>>]: never };

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
