// @dcl/protocol/out-ts/decentraland/sdk/components/ui_dropdown.gen.ts
/* eslint-disable */
import _m0 from "protobufjs/minimal";
import { Color4 } from "../../common/colors.gen";
import {
  Font,
  fontFromJSON,
  fontToJSON,
  TextAlignMode,
  textAlignModeFromJSON,
  textAlignModeToJSON,
} from "./common/texts.gen";

export const protobufPackage = "decentraland.sdk.components";

export interface PBUiDropdown {
  acceptEmpty: boolean;
  emptyLabel?: string | undefined;
  options: string[];
  /** default=null when accept_empty==true; default=0 when accept_empty==false */
  selectedIndex?: number | undefined;
  disabled: boolean;
  /** default=(0.0,0.0,0.0,1.0) */
  color?:
    | Color4
    | undefined;
  /** default='center' */
  textAlign?:
    | TextAlignMode
    | undefined;
  /** default=0 */
  font?:
    | Font
    | undefined;
  /** default=10 */
  fontSize?: number | undefined;
}

function createBasePBUiDropdown(): PBUiDropdown {
  return {
    acceptEmpty: false,
    emptyLabel: undefined,
    options: [],
    selectedIndex: undefined,
    disabled: false,
    color: undefined,
    textAlign: undefined,
    font: undefined,
    fontSize: undefined,
  };
}

export namespace PBUiDropdown {
  export function encode(message: PBUiDropdown, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.acceptEmpty === true) {
      writer.uint32(8).bool(message.acceptEmpty);
    }
    if (message.emptyLabel !== undefined) {
      writer.uint32(18).string(message.emptyLabel);
    }
    for (const v of message.options) {
      writer.uint32(26).string(v!);
    }
    if (message.selectedIndex !== undefined) {
      writer.uint32(32).int32(message.selectedIndex);
    }
    if (message.disabled === true) {
      writer.uint32(40).bool(message.disabled);
    }
    if (message.color !== undefined) {
      Color4.encode(message.color, writer.uint32(50).fork()).ldelim();
    }
    if (message.textAlign !== undefined) {
      writer.uint32(80).int32(message.textAlign);
    }
    if (message.font !== undefined) {
      writer.uint32(88).int32(message.font);
    }
    if (message.fontSize !== undefined) {
      writer.uint32(96).int32(message.fontSize);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): PBUiDropdown {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBasePBUiDropdown();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 8) {
            break;
          }

          message.acceptEmpty = reader.bool();
          continue;
        case 2:
          if (tag !== 18) {
            break;
          }

          message.emptyLabel = reader.string();
          continue;
        case 3:
          if (tag !== 26) {
            break;
          }

          message.options.push(reader.string());
          continue;
        case 4:
          if (tag !== 32) {
            break;
          }

          message.selectedIndex = reader.int32();
          continue;
        case 5:
          if (tag !== 40) {
            break;
          }

          message.disabled = reader.bool();
          continue;
        case 6:
          if (tag !== 50) {
            break;
          }

          message.color = Color4.decode(reader, reader.uint32());
          continue;
        case 10:
          if (tag !== 80) {
            break;
          }

          message.textAlign = reader.int32() as any;
          continue;
        case 11:
          if (tag !== 88) {
            break;
          }

          message.font = reader.int32() as any;
          continue;
        case 12:
          if (tag !== 96) {
            break;
          }

          message.fontSize = reader.int32();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): PBUiDropdown {
    return {
      acceptEmpty: isSet(object.acceptEmpty) ? Boolean(object.acceptEmpty) : false,
      emptyLabel: isSet(object.emptyLabel) ? String(object.emptyLabel) : undefined,
      options: Array.isArray(object?.options) ? object.options.map((e: any) => String(e)) : [],
      selectedIndex: isSet(object.selectedIndex) ? Number(object.selectedIndex) : undefined,
      disabled: isSet(object.disabled) ? Boolean(object.disabled) : false,
      color: isSet(object.color) ? Color4.fromJSON(object.color) : undefined,
      textAlign: isSet(object.textAlign) ? textAlignModeFromJSON(object.textAlign) : undefined,
      font: isSet(object.font) ? fontFromJSON(object.font) : undefined,
      fontSize: isSet(object.fontSize) ? Number(object.fontSize) : undefined,
    };
  }

  export function toJSON(message: PBUiDropdown): unknown {
    const obj: any = {};
    message.acceptEmpty !== undefined && (obj.acceptEmpty = message.acceptEmpty);
    message.emptyLabel !== undefined && (obj.emptyLabel = message.emptyLabel);
    if (message.options) {
      obj.options = message.options.map((e) => e);
    } else {
      obj.options = [];
    }
    message.selectedIndex !== undefined && (obj.selectedIndex = Math.round(message.selectedIndex));
    message.disabled !== undefined && (obj.disabled = message.disabled);
    message.color !== undefined && (obj.color = message.color ? Color4.toJSON(message.color) : undefined);
    message.textAlign !== undefined &&
      (obj.textAlign = message.textAlign !== undefined ? textAlignModeToJSON(message.textAlign) : undefined);
    message.font !== undefined && (obj.font = message.font !== undefined ? fontToJSON(message.font) : undefined);
    message.fontSize !== undefined && (obj.fontSize = Math.round(message.fontSize));
    return obj;
  }

  export function create<I extends Exact<DeepPartial<PBUiDropdown>, I>>(base?: I): PBUiDropdown {
    return PBUiDropdown.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<PBUiDropdown>, I>>(object: I): PBUiDropdown {
    const message = createBasePBUiDropdown();
    message.acceptEmpty = object.acceptEmpty ?? false;
    message.emptyLabel = object.emptyLabel ?? undefined;
    message.options = object.options?.map((e) => e) || [];
    message.selectedIndex = object.selectedIndex ?? undefined;
    message.disabled = object.disabled ?? false;
    message.color = (object.color !== undefined && object.color !== null)
      ? Color4.fromPartial(object.color)
      : undefined;
    message.textAlign = object.textAlign ?? undefined;
    message.font = object.font ?? undefined;
    message.fontSize = object.fontSize ?? undefined;
    return message;
  }
}

type Builtin = Date | Function | Uint8Array | string | number | boolean | undefined;

export type DeepPartial<T> = T extends Builtin ? T
  : T extends Array<infer U> ? Array<DeepPartial<U>> : T extends ReadonlyArray<infer U> ? ReadonlyArray<DeepPartial<U>>
  : T extends { $case: string } ? { [K in keyof Omit<T, "$case">]?: DeepPartial<T[K]> } & { $case: T["$case"] }
  : T extends {} ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & { [K in Exclude<keyof I, KeysOfUnion<P>>]: never };

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
