// @dcl/protocol/out-ts/decentraland/sdk/components/ui_input.gen.ts
/* eslint-disable */
import _m0 from "protobufjs/minimal";
import { Color4 } from "../../common/colors.gen";
import {
  Font,
  fontFromJSON,
  fontToJSON,
  TextAlignMode,
  textAlignModeFromJSON,
  textAlignModeToJSON,
} from "./common/texts.gen";

export const protobufPackage = "decentraland.sdk.components";

export interface PBUiInput {
  placeholder: string;
  /** default=(0.0,0.0,0.0,1.0) */
  color?:
    | Color4
    | undefined;
  /** default=(0.3,0.3,0.3,1.0) */
  placeholderColor?: Color4 | undefined;
  disabled: boolean;
  /** default='center' */
  textAlign?:
    | TextAlignMode
    | undefined;
  /** default=0 */
  font?:
    | Font
    | undefined;
  /** default=10 */
  fontSize?: number | undefined;
  value?: string | undefined;
}

function createBasePBUiInput(): PBUiInput {
  return {
    placeholder: "",
    color: undefined,
    placeholderColor: undefined,
    disabled: false,
    textAlign: undefined,
    font: undefined,
    fontSize: undefined,
    value: undefined,
  };
}

export namespace PBUiInput {
  export function encode(message: PBUiInput, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.placeholder !== "") {
      writer.uint32(10).string(message.placeholder);
    }
    if (message.color !== undefined) {
      Color4.encode(message.color, writer.uint32(18).fork()).ldelim();
    }
    if (message.placeholderColor !== undefined) {
      Color4.encode(message.placeholderColor, writer.uint32(26).fork()).ldelim();
    }
    if (message.disabled === true) {
      writer.uint32(32).bool(message.disabled);
    }
    if (message.textAlign !== undefined) {
      writer.uint32(80).int32(message.textAlign);
    }
    if (message.font !== undefined) {
      writer.uint32(88).int32(message.font);
    }
    if (message.fontSize !== undefined) {
      writer.uint32(96).int32(message.fontSize);
    }
    if (message.value !== undefined) {
      writer.uint32(106).string(message.value);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): PBUiInput {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBasePBUiInput();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.placeholder = reader.string();
          continue;
        case 2:
          if (tag !== 18) {
            break;
          }

          message.color = Color4.decode(reader, reader.uint32());
          continue;
        case 3:
          if (tag !== 26) {
            break;
          }

          message.placeholderColor = Color4.decode(reader, reader.uint32());
          continue;
        case 4:
          if (tag !== 32) {
            break;
          }

          message.disabled = reader.bool();
          continue;
        case 10:
          if (tag !== 80) {
            break;
          }

          message.textAlign = reader.int32() as any;
          continue;
        case 11:
          if (tag !== 88) {
            break;
          }

          message.font = reader.int32() as any;
          continue;
        case 12:
          if (tag !== 96) {
            break;
          }

          message.fontSize = reader.int32();
          continue;
        case 13:
          if (tag !== 106) {
            break;
          }

          message.value = reader.string();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): PBUiInput {
    return {
      placeholder: isSet(object.placeholder) ? String(object.placeholder) : "",
      color: isSet(object.color) ? Color4.fromJSON(object.color) : undefined,
      placeholderColor: isSet(object.placeholderColor) ? Color4.fromJSON(object.placeholderColor) : undefined,
      disabled: isSet(object.disabled) ? Boolean(object.disabled) : false,
      textAlign: isSet(object.textAlign) ? textAlignModeFromJSON(object.textAlign) : undefined,
      font: isSet(object.font) ? fontFromJSON(object.font) : undefined,
      fontSize: isSet(object.fontSize) ? Number(object.fontSize) : undefined,
      value: isSet(object.value) ? String(object.value) : undefined,
    };
  }

  export function toJSON(message: PBUiInput): unknown {
    const obj: any = {};
    message.placeholder !== undefined && (obj.placeholder = message.placeholder);
    message.color !== undefined && (obj.color = message.color ? Color4.toJSON(message.color) : undefined);
    message.placeholderColor !== undefined &&
      (obj.placeholderColor = message.placeholderColor ? Color4.toJSON(message.placeholderColor) : undefined);
    message.disabled !== undefined && (obj.disabled = message.disabled);
    message.textAlign !== undefined &&
      (obj.textAlign = message.textAlign !== undefined ? textAlignModeToJSON(message.textAlign) : undefined);
    message.font !== undefined && (obj.font = message.font !== undefined ? fontToJSON(message.font) : undefined);
    message.fontSize !== undefined && (obj.fontSize = Math.round(message.fontSize));
    message.value !== undefined && (obj.value = message.value);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<PBUiInput>, I>>(base?: I): PBUiInput {
    return PBUiInput.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<PBUiInput>, I>>(object: I): PBUiInput {
    const message = createBasePBUiInput();
    message.placeholder = object.placeholder ?? "";
    message.color = (object.color !== undefined && object.color !== null)
      ? Color4.fromPartial(object.color)
      : undefined;
    message.placeholderColor = (object.placeholderColor !== undefined && object.placeholderColor !== null)
      ? Color4.fromPartial(object.placeholderColor)
      : undefined;
    message.disabled = object.disabled ?? false;
    message.textAlign = object.textAlign ?? undefined;
    message.font = object.font ?? undefined;
    message.fontSize = object.fontSize ?? undefined;
    message.value = object.value ?? undefined;
    return message;
  }
}

type Builtin = Date | Function | Uint8Array | string | number | boolean | undefined;

export type DeepPartial<T> = T extends Builtin ? T
  : T extends Array<infer U> ? Array<DeepPartial<U>> : T extends ReadonlyArray<infer U> ? ReadonlyArray<DeepPartial<U>>
  : T extends { $case: string } ? { [K in keyof Omit<T, "$case">]?: DeepPartial<T[K]> } & { $case: T["$case"] }
  : T extends {} ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & { [K in Exclude<keyof I, KeysOfUnion<P>>]: never };

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
