// @dcl/protocol/out-ts/decentraland/sdk/components/ui_input_result.gen.ts
/* eslint-disable */
import _m0 from "protobufjs/minimal";

export const protobufPackage = "decentraland.sdk.components";

export interface PBUiInputResult {
  value: string;
  /** default: false */
  isSubmit?: boolean | undefined;
}

function createBasePBUiInputResult(): PBUiInputResult {
  return { value: "", isSubmit: undefined };
}

export namespace PBUiInputResult {
  export function encode(message: PBUiInputResult, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.value !== "") {
      writer.uint32(10).string(message.value);
    }
    if (message.isSubmit !== undefined) {
      writer.uint32(16).bool(message.isSubmit);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): PBUiInputResult {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBasePBUiInputResult();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.value = reader.string();
          continue;
        case 2:
          if (tag !== 16) {
            break;
          }

          message.isSubmit = reader.bool();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): PBUiInputResult {
    return {
      value: isSet(object.value) ? String(object.value) : "",
      isSubmit: isSet(object.isSubmit) ? Boolean(object.isSubmit) : undefined,
    };
  }

  export function toJSON(message: PBUiInputResult): unknown {
    const obj: any = {};
    message.value !== undefined && (obj.value = message.value);
    message.isSubmit !== undefined && (obj.isSubmit = message.isSubmit);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<PBUiInputResult>, I>>(base?: I): PBUiInputResult {
    return PBUiInputResult.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<PBUiInputResult>, I>>(object: I): PBUiInputResult {
    const message = createBasePBUiInputResult();
    message.value = object.value ?? "";
    message.isSubmit = object.isSubmit ?? undefined;
    return message;
  }
}

type Builtin = Date | Function | Uint8Array | string | number | boolean | undefined;

export type DeepPartial<T> = T extends Builtin ? T
  : T extends Array<infer U> ? Array<DeepPartial<U>> : T extends ReadonlyArray<infer U> ? ReadonlyArray<DeepPartial<U>>
  : T extends { $case: string } ? { [K in keyof Omit<T, "$case">]?: DeepPartial<T[K]> } & { $case: T["$case"] }
  : T extends {} ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & { [K in Exclude<keyof I, KeysOfUnion<P>>]: never };

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
