// @dcl/protocol/out-js/decentraland/sdk/ecs6/engine_interface_ecs6.gen.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityAction = exports.EntityActionPayload = exports.InitMessagesFinishedBody = exports.ComponentUpdatedBody = exports.ComponentDisposedBody = exports.ComponentCreatedBody = exports.QueryBody_RayQuery = exports.QueryBody_Ray = exports.QueryBody = exports.SetEntityParentBody = exports.ComponentRemovedBody = exports.AttachEntityComponentBody = exports.UpdateEntityComponentBody = exports.RemoveEntityBody = exports.CreateEntityBody = exports.ComponentBodyPayload = exports.OpenNFTDialogBody = exports.OpenExternalUrlBody = exports.protobufPackage = void 0;
/* eslint-disable */
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const vectors_gen_1 = require("../../common/vectors.gen");
const components_ecs6_gen_1 = require("./components_ecs6.gen");
exports.protobufPackage = "decentraland.sdk.ecs6";
function createBaseOpenExternalUrlBody() {
    return { url: "" };
}
var OpenExternalUrlBody;
(function (OpenExternalUrlBody) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.url !== "") {
            writer.uint32(10).string(message.url);
        }
        return writer;
    }
    OpenExternalUrlBody.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseOpenExternalUrlBody();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.url = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    OpenExternalUrlBody.decode = decode;
    function fromJSON(object) {
        return { url: isSet(object.url) ? String(object.url) : "" };
    }
    OpenExternalUrlBody.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.url !== undefined && (obj.url = message.url);
        return obj;
    }
    OpenExternalUrlBody.toJSON = toJSON;
    function create(base) {
        return OpenExternalUrlBody.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    OpenExternalUrlBody.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseOpenExternalUrlBody();
        message.url = (_a = object.url) !== null && _a !== void 0 ? _a : "";
        return message;
    }
    OpenExternalUrlBody.fromPartial = fromPartial;
})(OpenExternalUrlBody || (exports.OpenExternalUrlBody = OpenExternalUrlBody = {}));
function createBaseOpenNFTDialogBody() {
    return { assetContractAddress: "", tokenId: "", comment: undefined };
}
var OpenNFTDialogBody;
(function (OpenNFTDialogBody) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.assetContractAddress !== "") {
            writer.uint32(10).string(message.assetContractAddress);
        }
        if (message.tokenId !== "") {
            writer.uint32(18).string(message.tokenId);
        }
        if (message.comment !== undefined) {
            writer.uint32(26).string(message.comment);
        }
        return writer;
    }
    OpenNFTDialogBody.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseOpenNFTDialogBody();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.assetContractAddress = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.tokenId = reader.string();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.comment = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    OpenNFTDialogBody.decode = decode;
    function fromJSON(object) {
        return {
            assetContractAddress: isSet(object.assetContractAddress) ? String(object.assetContractAddress) : "",
            tokenId: isSet(object.tokenId) ? String(object.tokenId) : "",
            comment: isSet(object.comment) ? String(object.comment) : undefined,
        };
    }
    OpenNFTDialogBody.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.assetContractAddress !== undefined && (obj.assetContractAddress = message.assetContractAddress);
        message.tokenId !== undefined && (obj.tokenId = message.tokenId);
        message.comment !== undefined && (obj.comment = message.comment);
        return obj;
    }
    OpenNFTDialogBody.toJSON = toJSON;
    function create(base) {
        return OpenNFTDialogBody.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    OpenNFTDialogBody.create = create;
    function fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseOpenNFTDialogBody();
        message.assetContractAddress = (_a = object.assetContractAddress) !== null && _a !== void 0 ? _a : "";
        message.tokenId = (_b = object.tokenId) !== null && _b !== void 0 ? _b : "";
        message.comment = (_c = object.comment) !== null && _c !== void 0 ? _c : undefined;
        return message;
    }
    OpenNFTDialogBody.fromPartial = fromPartial;
})(OpenNFTDialogBody || (exports.OpenNFTDialogBody = OpenNFTDialogBody = {}));
function createBaseComponentBodyPayload() {
    return { payload: undefined };
}
var ComponentBodyPayload;
(function (ComponentBodyPayload) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        var _a;
        switch ((_a = message.payload) === null || _a === void 0 ? void 0 : _a.$case) {
            case "avatarModifierArea":
                components_ecs6_gen_1.ECS6ComponentAvatarModifierArea.encode(message.payload.avatarModifierArea, writer.uint32(10).fork()).ldelim();
                break;
            case "transform":
                components_ecs6_gen_1.ECS6ComponentTransform.encode(message.payload.transform, writer.uint32(18).fork()).ldelim();
                break;
            case "attachToAvatar":
                components_ecs6_gen_1.ECS6ComponentAttachToAvatar.encode(message.payload.attachToAvatar, writer.uint32(26).fork()).ldelim();
                break;
            case "billboard":
                components_ecs6_gen_1.ECS6ComponentBillboard.encode(message.payload.billboard, writer.uint32(34).fork()).ldelim();
                break;
            case "boxShape":
                components_ecs6_gen_1.ECS6ComponentBoxShape.encode(message.payload.boxShape, writer.uint32(42).fork()).ldelim();
                break;
            case "sphereShape":
                components_ecs6_gen_1.ECS6ComponentSphereShape.encode(message.payload.sphereShape, writer.uint32(50).fork()).ldelim();
                break;
            case "circleShape":
                components_ecs6_gen_1.ECS6ComponentCircleShape.encode(message.payload.circleShape, writer.uint32(58).fork()).ldelim();
                break;
            case "planeShape":
                components_ecs6_gen_1.ECS6ComponentPlaneShape.encode(message.payload.planeShape, writer.uint32(66).fork()).ldelim();
                break;
            case "coneShape":
                components_ecs6_gen_1.ECS6ComponentConeShape.encode(message.payload.coneShape, writer.uint32(74).fork()).ldelim();
                break;
            case "cylinderShape":
                components_ecs6_gen_1.ECS6ComponentCylinderShape.encode(message.payload.cylinderShape, writer.uint32(82).fork()).ldelim();
                break;
            case "gltfShape":
                components_ecs6_gen_1.ECS6ComponentGltfShape.encode(message.payload.gltfShape, writer.uint32(90).fork()).ldelim();
                break;
            case "nftShape":
                components_ecs6_gen_1.ECS6ComponentNftShape.encode(message.payload.nftShape, writer.uint32(98).fork()).ldelim();
                break;
            case "texture":
                components_ecs6_gen_1.ECS6ComponentTexture.encode(message.payload.texture, writer.uint32(106).fork()).ldelim();
                break;
            case "animator":
                components_ecs6_gen_1.ECS6ComponentAnimator.encode(message.payload.animator, writer.uint32(114).fork()).ldelim();
                break;
            case "objShape":
                components_ecs6_gen_1.ECS6ComponentObjShape.encode(message.payload.objShape, writer.uint32(122).fork()).ldelim();
                break;
            case "font":
                components_ecs6_gen_1.ECS6ComponentFont.encode(message.payload.font, writer.uint32(130).fork()).ldelim();
                break;
            case "textShape":
                components_ecs6_gen_1.ECS6ComponentTextShape.encode(message.payload.textShape, writer.uint32(138).fork()).ldelim();
                break;
            case "material":
                components_ecs6_gen_1.ECS6ComponentMaterial.encode(message.payload.material, writer.uint32(146).fork()).ldelim();
                break;
            case "basicMaterial":
                components_ecs6_gen_1.ECS6ComponentBasicMaterial.encode(message.payload.basicMaterial, writer.uint32(154).fork()).ldelim();
                break;
            case "uuidCallback":
                components_ecs6_gen_1.ECS6ComponentUuidCallback.encode(message.payload.uuidCallback, writer.uint32(162).fork()).ldelim();
                break;
            case "smartItem":
                components_ecs6_gen_1.ECS6ComponentSmartItem.encode(message.payload.smartItem, writer.uint32(170).fork()).ldelim();
                break;
            case "videoClip":
                components_ecs6_gen_1.ECS6ComponentVideoClip.encode(message.payload.videoClip, writer.uint32(178).fork()).ldelim();
                break;
            case "videoTexture":
                components_ecs6_gen_1.ECS6ComponentVideoTexture.encode(message.payload.videoTexture, writer.uint32(186).fork()).ldelim();
                break;
            case "cameraModeArea":
                components_ecs6_gen_1.ECS6ComponentCameraModeArea.encode(message.payload.cameraModeArea, writer.uint32(194).fork()).ldelim();
                break;
            case "avatarTexture":
                components_ecs6_gen_1.ECS6ComponentAvatarTexture.encode(message.payload.avatarTexture, writer.uint32(202).fork()).ldelim();
                break;
            case "audioClip":
                components_ecs6_gen_1.ECS6ComponentAudioClip.encode(message.payload.audioClip, writer.uint32(210).fork()).ldelim();
                break;
            case "audioSource":
                components_ecs6_gen_1.ECS6ComponentAudioSource.encode(message.payload.audioSource, writer.uint32(218).fork()).ldelim();
                break;
            case "audioStream":
                components_ecs6_gen_1.ECS6ComponentAudioStream.encode(message.payload.audioStream, writer.uint32(226).fork()).ldelim();
                break;
            case "avatarShape":
                components_ecs6_gen_1.ECS6ComponentAvatarShape.encode(message.payload.avatarShape, writer.uint32(234).fork()).ldelim();
                break;
            case "gizmos":
                components_ecs6_gen_1.ECS6ComponentGizmos.encode(message.payload.gizmos, writer.uint32(242).fork()).ldelim();
                break;
            case "uiShape":
                components_ecs6_gen_1.ECS6ComponentUiShape.encode(message.payload.uiShape, writer.uint32(250).fork()).ldelim();
                break;
            case "uiContainerRect":
                components_ecs6_gen_1.ECS6ComponentUiContainerRect.encode(message.payload.uiContainerRect, writer.uint32(258).fork()).ldelim();
                break;
            case "uiContainerStack":
                components_ecs6_gen_1.ECS6ComponentUiContainerStack.encode(message.payload.uiContainerStack, writer.uint32(266).fork()).ldelim();
                break;
            case "uiButton":
                components_ecs6_gen_1.ECS6ComponentUiButton.encode(message.payload.uiButton, writer.uint32(274).fork()).ldelim();
                break;
            case "uiText":
                components_ecs6_gen_1.ECS6ComponentUiText.encode(message.payload.uiText, writer.uint32(282).fork()).ldelim();
                break;
            case "uiInputText":
                components_ecs6_gen_1.ECS6ComponentUiInputText.encode(message.payload.uiInputText, writer.uint32(290).fork()).ldelim();
                break;
            case "uiImage":
                components_ecs6_gen_1.ECS6ComponentUiImage.encode(message.payload.uiImage, writer.uint32(298).fork()).ldelim();
                break;
            case "uiScrollRect":
                components_ecs6_gen_1.ECS6ComponentUiScrollRect.encode(message.payload.uiScrollRect, writer.uint32(306).fork()).ldelim();
                break;
            case "uiWorldSpaceShape":
                components_ecs6_gen_1.ECS6ComponentUiWorldSpaceShape.encode(message.payload.uiWorldSpaceShape, writer.uint32(314).fork()).ldelim();
                break;
            case "uiScreenSpaceShape":
                components_ecs6_gen_1.ECS6ComponentUiScreenSpaceShape.encode(message.payload.uiScreenSpaceShape, writer.uint32(322).fork()).ldelim();
                break;
            case "uiFullScreenShape":
                components_ecs6_gen_1.ECS6ComponentUiFullScreenShape.encode(message.payload.uiFullScreenShape, writer.uint32(330).fork()).ldelim();
                break;
        }
        return writer;
    }
    ComponentBodyPayload.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseComponentBodyPayload();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.payload = {
                        $case: "avatarModifierArea",
                        avatarModifierArea: components_ecs6_gen_1.ECS6ComponentAvatarModifierArea.decode(reader, reader.uint32()),
                    };
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.payload = { $case: "transform", transform: components_ecs6_gen_1.ECS6ComponentTransform.decode(reader, reader.uint32()) };
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.payload = {
                        $case: "attachToAvatar",
                        attachToAvatar: components_ecs6_gen_1.ECS6ComponentAttachToAvatar.decode(reader, reader.uint32()),
                    };
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.payload = { $case: "billboard", billboard: components_ecs6_gen_1.ECS6ComponentBillboard.decode(reader, reader.uint32()) };
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.payload = { $case: "boxShape", boxShape: components_ecs6_gen_1.ECS6ComponentBoxShape.decode(reader, reader.uint32()) };
                    continue;
                case 6:
                    if (tag !== 50) {
                        break;
                    }
                    message.payload = {
                        $case: "sphereShape",
                        sphereShape: components_ecs6_gen_1.ECS6ComponentSphereShape.decode(reader, reader.uint32()),
                    };
                    continue;
                case 7:
                    if (tag !== 58) {
                        break;
                    }
                    message.payload = {
                        $case: "circleShape",
                        circleShape: components_ecs6_gen_1.ECS6ComponentCircleShape.decode(reader, reader.uint32()),
                    };
                    continue;
                case 8:
                    if (tag !== 66) {
                        break;
                    }
                    message.payload = {
                        $case: "planeShape",
                        planeShape: components_ecs6_gen_1.ECS6ComponentPlaneShape.decode(reader, reader.uint32()),
                    };
                    continue;
                case 9:
                    if (tag !== 74) {
                        break;
                    }
                    message.payload = { $case: "coneShape", coneShape: components_ecs6_gen_1.ECS6ComponentConeShape.decode(reader, reader.uint32()) };
                    continue;
                case 10:
                    if (tag !== 82) {
                        break;
                    }
                    message.payload = {
                        $case: "cylinderShape",
                        cylinderShape: components_ecs6_gen_1.ECS6ComponentCylinderShape.decode(reader, reader.uint32()),
                    };
                    continue;
                case 11:
                    if (tag !== 90) {
                        break;
                    }
                    message.payload = { $case: "gltfShape", gltfShape: components_ecs6_gen_1.ECS6ComponentGltfShape.decode(reader, reader.uint32()) };
                    continue;
                case 12:
                    if (tag !== 98) {
                        break;
                    }
                    message.payload = { $case: "nftShape", nftShape: components_ecs6_gen_1.ECS6ComponentNftShape.decode(reader, reader.uint32()) };
                    continue;
                case 13:
                    if (tag !== 106) {
                        break;
                    }
                    message.payload = { $case: "texture", texture: components_ecs6_gen_1.ECS6ComponentTexture.decode(reader, reader.uint32()) };
                    continue;
                case 14:
                    if (tag !== 114) {
                        break;
                    }
                    message.payload = { $case: "animator", animator: components_ecs6_gen_1.ECS6ComponentAnimator.decode(reader, reader.uint32()) };
                    continue;
                case 15:
                    if (tag !== 122) {
                        break;
                    }
                    message.payload = { $case: "objShape", objShape: components_ecs6_gen_1.ECS6ComponentObjShape.decode(reader, reader.uint32()) };
                    continue;
                case 16:
                    if (tag !== 130) {
                        break;
                    }
                    message.payload = { $case: "font", font: components_ecs6_gen_1.ECS6ComponentFont.decode(reader, reader.uint32()) };
                    continue;
                case 17:
                    if (tag !== 138) {
                        break;
                    }
                    message.payload = { $case: "textShape", textShape: components_ecs6_gen_1.ECS6ComponentTextShape.decode(reader, reader.uint32()) };
                    continue;
                case 18:
                    if (tag !== 146) {
                        break;
                    }
                    message.payload = { $case: "material", material: components_ecs6_gen_1.ECS6ComponentMaterial.decode(reader, reader.uint32()) };
                    continue;
                case 19:
                    if (tag !== 154) {
                        break;
                    }
                    message.payload = {
                        $case: "basicMaterial",
                        basicMaterial: components_ecs6_gen_1.ECS6ComponentBasicMaterial.decode(reader, reader.uint32()),
                    };
                    continue;
                case 20:
                    if (tag !== 162) {
                        break;
                    }
                    message.payload = {
                        $case: "uuidCallback",
                        uuidCallback: components_ecs6_gen_1.ECS6ComponentUuidCallback.decode(reader, reader.uint32()),
                    };
                    continue;
                case 21:
                    if (tag !== 170) {
                        break;
                    }
                    message.payload = { $case: "smartItem", smartItem: components_ecs6_gen_1.ECS6ComponentSmartItem.decode(reader, reader.uint32()) };
                    continue;
                case 22:
                    if (tag !== 178) {
                        break;
                    }
                    message.payload = { $case: "videoClip", videoClip: components_ecs6_gen_1.ECS6ComponentVideoClip.decode(reader, reader.uint32()) };
                    continue;
                case 23:
                    if (tag !== 186) {
                        break;
                    }
                    message.payload = {
                        $case: "videoTexture",
                        videoTexture: components_ecs6_gen_1.ECS6ComponentVideoTexture.decode(reader, reader.uint32()),
                    };
                    continue;
                case 24:
                    if (tag !== 194) {
                        break;
                    }
                    message.payload = {
                        $case: "cameraModeArea",
                        cameraModeArea: components_ecs6_gen_1.ECS6ComponentCameraModeArea.decode(reader, reader.uint32()),
                    };
                    continue;
                case 25:
                    if (tag !== 202) {
                        break;
                    }
                    message.payload = {
                        $case: "avatarTexture",
                        avatarTexture: components_ecs6_gen_1.ECS6ComponentAvatarTexture.decode(reader, reader.uint32()),
                    };
                    continue;
                case 26:
                    if (tag !== 210) {
                        break;
                    }
                    message.payload = { $case: "audioClip", audioClip: components_ecs6_gen_1.ECS6ComponentAudioClip.decode(reader, reader.uint32()) };
                    continue;
                case 27:
                    if (tag !== 218) {
                        break;
                    }
                    message.payload = {
                        $case: "audioSource",
                        audioSource: components_ecs6_gen_1.ECS6ComponentAudioSource.decode(reader, reader.uint32()),
                    };
                    continue;
                case 28:
                    if (tag !== 226) {
                        break;
                    }
                    message.payload = {
                        $case: "audioStream",
                        audioStream: components_ecs6_gen_1.ECS6ComponentAudioStream.decode(reader, reader.uint32()),
                    };
                    continue;
                case 29:
                    if (tag !== 234) {
                        break;
                    }
                    message.payload = {
                        $case: "avatarShape",
                        avatarShape: components_ecs6_gen_1.ECS6ComponentAvatarShape.decode(reader, reader.uint32()),
                    };
                    continue;
                case 30:
                    if (tag !== 242) {
                        break;
                    }
                    message.payload = { $case: "gizmos", gizmos: components_ecs6_gen_1.ECS6ComponentGizmos.decode(reader, reader.uint32()) };
                    continue;
                case 31:
                    if (tag !== 250) {
                        break;
                    }
                    message.payload = { $case: "uiShape", uiShape: components_ecs6_gen_1.ECS6ComponentUiShape.decode(reader, reader.uint32()) };
                    continue;
                case 32:
                    if (tag !== 258) {
                        break;
                    }
                    message.payload = {
                        $case: "uiContainerRect",
                        uiContainerRect: components_ecs6_gen_1.ECS6ComponentUiContainerRect.decode(reader, reader.uint32()),
                    };
                    continue;
                case 33:
                    if (tag !== 266) {
                        break;
                    }
                    message.payload = {
                        $case: "uiContainerStack",
                        uiContainerStack: components_ecs6_gen_1.ECS6ComponentUiContainerStack.decode(reader, reader.uint32()),
                    };
                    continue;
                case 34:
                    if (tag !== 274) {
                        break;
                    }
                    message.payload = { $case: "uiButton", uiButton: components_ecs6_gen_1.ECS6ComponentUiButton.decode(reader, reader.uint32()) };
                    continue;
                case 35:
                    if (tag !== 282) {
                        break;
                    }
                    message.payload = { $case: "uiText", uiText: components_ecs6_gen_1.ECS6ComponentUiText.decode(reader, reader.uint32()) };
                    continue;
                case 36:
                    if (tag !== 290) {
                        break;
                    }
                    message.payload = {
                        $case: "uiInputText",
                        uiInputText: components_ecs6_gen_1.ECS6ComponentUiInputText.decode(reader, reader.uint32()),
                    };
                    continue;
                case 37:
                    if (tag !== 298) {
                        break;
                    }
                    message.payload = { $case: "uiImage", uiImage: components_ecs6_gen_1.ECS6ComponentUiImage.decode(reader, reader.uint32()) };
                    continue;
                case 38:
                    if (tag !== 306) {
                        break;
                    }
                    message.payload = {
                        $case: "uiScrollRect",
                        uiScrollRect: components_ecs6_gen_1.ECS6ComponentUiScrollRect.decode(reader, reader.uint32()),
                    };
                    continue;
                case 39:
                    if (tag !== 314) {
                        break;
                    }
                    message.payload = {
                        $case: "uiWorldSpaceShape",
                        uiWorldSpaceShape: components_ecs6_gen_1.ECS6ComponentUiWorldSpaceShape.decode(reader, reader.uint32()),
                    };
                    continue;
                case 40:
                    if (tag !== 322) {
                        break;
                    }
                    message.payload = {
                        $case: "uiScreenSpaceShape",
                        uiScreenSpaceShape: components_ecs6_gen_1.ECS6ComponentUiScreenSpaceShape.decode(reader, reader.uint32()),
                    };
                    continue;
                case 41:
                    if (tag !== 330) {
                        break;
                    }
                    message.payload = {
                        $case: "uiFullScreenShape",
                        uiFullScreenShape: components_ecs6_gen_1.ECS6ComponentUiFullScreenShape.decode(reader, reader.uint32()),
                    };
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    ComponentBodyPayload.decode = decode;
    function fromJSON(object) {
        return {
            payload: isSet(object.avatarModifierArea)
                ? {
                    $case: "avatarModifierArea",
                    avatarModifierArea: components_ecs6_gen_1.ECS6ComponentAvatarModifierArea.fromJSON(object.avatarModifierArea),
                }
                : isSet(object.transform)
                    ? { $case: "transform", transform: components_ecs6_gen_1.ECS6ComponentTransform.fromJSON(object.transform) }
                    : isSet(object.attachToAvatar)
                        ? { $case: "attachToAvatar", attachToAvatar: components_ecs6_gen_1.ECS6ComponentAttachToAvatar.fromJSON(object.attachToAvatar) }
                        : isSet(object.billboard)
                            ? { $case: "billboard", billboard: components_ecs6_gen_1.ECS6ComponentBillboard.fromJSON(object.billboard) }
                            : isSet(object.boxShape)
                                ? { $case: "boxShape", boxShape: components_ecs6_gen_1.ECS6ComponentBoxShape.fromJSON(object.boxShape) }
                                : isSet(object.sphereShape)
                                    ? { $case: "sphereShape", sphereShape: components_ecs6_gen_1.ECS6ComponentSphereShape.fromJSON(object.sphereShape) }
                                    : isSet(object.circleShape)
                                        ? { $case: "circleShape", circleShape: components_ecs6_gen_1.ECS6ComponentCircleShape.fromJSON(object.circleShape) }
                                        : isSet(object.planeShape)
                                            ? { $case: "planeShape", planeShape: components_ecs6_gen_1.ECS6ComponentPlaneShape.fromJSON(object.planeShape) }
                                            : isSet(object.coneShape)
                                                ? { $case: "coneShape", coneShape: components_ecs6_gen_1.ECS6ComponentConeShape.fromJSON(object.coneShape) }
                                                : isSet(object.cylinderShape)
                                                    ? { $case: "cylinderShape", cylinderShape: components_ecs6_gen_1.ECS6ComponentCylinderShape.fromJSON(object.cylinderShape) }
                                                    : isSet(object.gltfShape)
                                                        ? { $case: "gltfShape", gltfShape: components_ecs6_gen_1.ECS6ComponentGltfShape.fromJSON(object.gltfShape) }
                                                        : isSet(object.nftShape)
                                                            ? { $case: "nftShape", nftShape: components_ecs6_gen_1.ECS6ComponentNftShape.fromJSON(object.nftShape) }
                                                            : isSet(object.texture)
                                                                ? { $case: "texture", texture: components_ecs6_gen_1.ECS6ComponentTexture.fromJSON(object.texture) }
                                                                : isSet(object.animator)
                                                                    ? { $case: "animator", animator: components_ecs6_gen_1.ECS6ComponentAnimator.fromJSON(object.animator) }
                                                                    : isSet(object.objShape)
                                                                        ? { $case: "objShape", objShape: components_ecs6_gen_1.ECS6ComponentObjShape.fromJSON(object.objShape) }
                                                                        : isSet(object.font)
                                                                            ? { $case: "font", font: components_ecs6_gen_1.ECS6ComponentFont.fromJSON(object.font) }
                                                                            : isSet(object.textShape)
                                                                                ? { $case: "textShape", textShape: components_ecs6_gen_1.ECS6ComponentTextShape.fromJSON(object.textShape) }
                                                                                : isSet(object.material)
                                                                                    ? { $case: "material", material: components_ecs6_gen_1.ECS6ComponentMaterial.fromJSON(object.material) }
                                                                                    : isSet(object.basicMaterial)
                                                                                        ? { $case: "basicMaterial", basicMaterial: components_ecs6_gen_1.ECS6ComponentBasicMaterial.fromJSON(object.basicMaterial) }
                                                                                        : isSet(object.uuidCallback)
                                                                                            ? { $case: "uuidCallback", uuidCallback: components_ecs6_gen_1.ECS6ComponentUuidCallback.fromJSON(object.uuidCallback) }
                                                                                            : isSet(object.smartItem)
                                                                                                ? { $case: "smartItem", smartItem: components_ecs6_gen_1.ECS6ComponentSmartItem.fromJSON(object.smartItem) }
                                                                                                : isSet(object.videoClip)
                                                                                                    ? { $case: "videoClip", videoClip: components_ecs6_gen_1.ECS6ComponentVideoClip.fromJSON(object.videoClip) }
                                                                                                    : isSet(object.videoTexture)
                                                                                                        ? { $case: "videoTexture", videoTexture: components_ecs6_gen_1.ECS6ComponentVideoTexture.fromJSON(object.videoTexture) }
                                                                                                        : isSet(object.cameraModeArea)
                                                                                                            ? { $case: "cameraModeArea", cameraModeArea: components_ecs6_gen_1.ECS6ComponentCameraModeArea.fromJSON(object.cameraModeArea) }
                                                                                                            : isSet(object.avatarTexture)
                                                                                                                ? { $case: "avatarTexture", avatarTexture: components_ecs6_gen_1.ECS6ComponentAvatarTexture.fromJSON(object.avatarTexture) }
                                                                                                                : isSet(object.audioClip)
                                                                                                                    ? { $case: "audioClip", audioClip: components_ecs6_gen_1.ECS6ComponentAudioClip.fromJSON(object.audioClip) }
                                                                                                                    : isSet(object.audioSource)
                                                                                                                        ? { $case: "audioSource", audioSource: components_ecs6_gen_1.ECS6ComponentAudioSource.fromJSON(object.audioSource) }
                                                                                                                        : isSet(object.audioStream)
                                                                                                                            ? { $case: "audioStream", audioStream: components_ecs6_gen_1.ECS6ComponentAudioStream.fromJSON(object.audioStream) }
                                                                                                                            : isSet(object.avatarShape)
                                                                                                                                ? { $case: "avatarShape", avatarShape: components_ecs6_gen_1.ECS6ComponentAvatarShape.fromJSON(object.avatarShape) }
                                                                                                                                : isSet(object.gizmos)
                                                                                                                                    ? { $case: "gizmos", gizmos: components_ecs6_gen_1.ECS6ComponentGizmos.fromJSON(object.gizmos) }
                                                                                                                                    : isSet(object.uiShape)
                                                                                                                                        ? { $case: "uiShape", uiShape: components_ecs6_gen_1.ECS6ComponentUiShape.fromJSON(object.uiShape) }
                                                                                                                                        : isSet(object.uiContainerRect)
                                                                                                                                            ? { $case: "uiContainerRect", uiContainerRect: components_ecs6_gen_1.ECS6ComponentUiContainerRect.fromJSON(object.uiContainerRect) }
                                                                                                                                            : isSet(object.uiContainerStack)
                                                                                                                                                ? {
                                                                                                                                                    $case: "uiContainerStack",
                                                                                                                                                    uiContainerStack: components_ecs6_gen_1.ECS6ComponentUiContainerStack.fromJSON(object.uiContainerStack),
                                                                                                                                                }
                                                                                                                                                : isSet(object.uiButton)
                                                                                                                                                    ? { $case: "uiButton", uiButton: components_ecs6_gen_1.ECS6ComponentUiButton.fromJSON(object.uiButton) }
                                                                                                                                                    : isSet(object.uiText)
                                                                                                                                                        ? { $case: "uiText", uiText: components_ecs6_gen_1.ECS6ComponentUiText.fromJSON(object.uiText) }
                                                                                                                                                        : isSet(object.uiInputText)
                                                                                                                                                            ? { $case: "uiInputText", uiInputText: components_ecs6_gen_1.ECS6ComponentUiInputText.fromJSON(object.uiInputText) }
                                                                                                                                                            : isSet(object.uiImage)
                                                                                                                                                                ? { $case: "uiImage", uiImage: components_ecs6_gen_1.ECS6ComponentUiImage.fromJSON(object.uiImage) }
                                                                                                                                                                : isSet(object.uiScrollRect)
                                                                                                                                                                    ? { $case: "uiScrollRect", uiScrollRect: components_ecs6_gen_1.ECS6ComponentUiScrollRect.fromJSON(object.uiScrollRect) }
                                                                                                                                                                    : isSet(object.uiWorldSpaceShape)
                                                                                                                                                                        ? {
                                                                                                                                                                            $case: "uiWorldSpaceShape",
                                                                                                                                                                            uiWorldSpaceShape: components_ecs6_gen_1.ECS6ComponentUiWorldSpaceShape.fromJSON(object.uiWorldSpaceShape),
                                                                                                                                                                        }
                                                                                                                                                                        : isSet(object.uiScreenSpaceShape)
                                                                                                                                                                            ? {
                                                                                                                                                                                $case: "uiScreenSpaceShape",
                                                                                                                                                                                uiScreenSpaceShape: components_ecs6_gen_1.ECS6ComponentUiScreenSpaceShape.fromJSON(object.uiScreenSpaceShape),
                                                                                                                                                                            }
                                                                                                                                                                            : isSet(object.uiFullScreenShape)
                                                                                                                                                                                ? {
                                                                                                                                                                                    $case: "uiFullScreenShape",
                                                                                                                                                                                    uiFullScreenShape: components_ecs6_gen_1.ECS6ComponentUiFullScreenShape.fromJSON(object.uiFullScreenShape),
                                                                                                                                                                                }
                                                                                                                                                                                : undefined,
        };
    }
    ComponentBodyPayload.fromJSON = fromJSON;
    function toJSON(message) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, _29, _30, _31, _32, _33, _34, _35, _36, _37, _38, _39, _40, _41, _42, _43, _44, _45, _46, _47, _48, _49, _50, _51, _52, _53, _54, _55, _56, _57, _58, _59, _60, _61, _62, _63, _64, _65, _66, _67, _68, _69, _70, _71, _72, _73, _74, _75, _76, _77, _78, _79, _80, _81, _82, _83, _84, _85, _86, _87, _88, _89, _90, _91, _92, _93, _94, _95, _96, _97, _98;
        const obj = {};
        ((_a = message.payload) === null || _a === void 0 ? void 0 : _a.$case) === "avatarModifierArea" &&
            (obj.avatarModifierArea = ((_b = message.payload) === null || _b === void 0 ? void 0 : _b.avatarModifierArea)
                ? components_ecs6_gen_1.ECS6ComponentAvatarModifierArea.toJSON((_c = message.payload) === null || _c === void 0 ? void 0 : _c.avatarModifierArea)
                : undefined);
        ((_d = message.payload) === null || _d === void 0 ? void 0 : _d.$case) === "transform" &&
            (obj.transform = ((_e = message.payload) === null || _e === void 0 ? void 0 : _e.transform)
                ? components_ecs6_gen_1.ECS6ComponentTransform.toJSON((_f = message.payload) === null || _f === void 0 ? void 0 : _f.transform)
                : undefined);
        ((_g = message.payload) === null || _g === void 0 ? void 0 : _g.$case) === "attachToAvatar" &&
            (obj.attachToAvatar = ((_h = message.payload) === null || _h === void 0 ? void 0 : _h.attachToAvatar)
                ? components_ecs6_gen_1.ECS6ComponentAttachToAvatar.toJSON((_j = message.payload) === null || _j === void 0 ? void 0 : _j.attachToAvatar)
                : undefined);
        ((_k = message.payload) === null || _k === void 0 ? void 0 : _k.$case) === "billboard" &&
            (obj.billboard = ((_l = message.payload) === null || _l === void 0 ? void 0 : _l.billboard)
                ? components_ecs6_gen_1.ECS6ComponentBillboard.toJSON((_m = message.payload) === null || _m === void 0 ? void 0 : _m.billboard)
                : undefined);
        ((_o = message.payload) === null || _o === void 0 ? void 0 : _o.$case) === "boxShape" &&
            (obj.boxShape = ((_p = message.payload) === null || _p === void 0 ? void 0 : _p.boxShape) ? components_ecs6_gen_1.ECS6ComponentBoxShape.toJSON((_q = message.payload) === null || _q === void 0 ? void 0 : _q.boxShape) : undefined);
        ((_r = message.payload) === null || _r === void 0 ? void 0 : _r.$case) === "sphereShape" &&
            (obj.sphereShape = ((_s = message.payload) === null || _s === void 0 ? void 0 : _s.sphereShape)
                ? components_ecs6_gen_1.ECS6ComponentSphereShape.toJSON((_t = message.payload) === null || _t === void 0 ? void 0 : _t.sphereShape)
                : undefined);
        ((_u = message.payload) === null || _u === void 0 ? void 0 : _u.$case) === "circleShape" &&
            (obj.circleShape = ((_v = message.payload) === null || _v === void 0 ? void 0 : _v.circleShape)
                ? components_ecs6_gen_1.ECS6ComponentCircleShape.toJSON((_w = message.payload) === null || _w === void 0 ? void 0 : _w.circleShape)
                : undefined);
        ((_x = message.payload) === null || _x === void 0 ? void 0 : _x.$case) === "planeShape" &&
            (obj.planeShape = ((_y = message.payload) === null || _y === void 0 ? void 0 : _y.planeShape)
                ? components_ecs6_gen_1.ECS6ComponentPlaneShape.toJSON((_z = message.payload) === null || _z === void 0 ? void 0 : _z.planeShape)
                : undefined);
        ((_0 = message.payload) === null || _0 === void 0 ? void 0 : _0.$case) === "coneShape" &&
            (obj.coneShape = ((_1 = message.payload) === null || _1 === void 0 ? void 0 : _1.coneShape)
                ? components_ecs6_gen_1.ECS6ComponentConeShape.toJSON((_2 = message.payload) === null || _2 === void 0 ? void 0 : _2.coneShape)
                : undefined);
        ((_3 = message.payload) === null || _3 === void 0 ? void 0 : _3.$case) === "cylinderShape" &&
            (obj.cylinderShape = ((_4 = message.payload) === null || _4 === void 0 ? void 0 : _4.cylinderShape)
                ? components_ecs6_gen_1.ECS6ComponentCylinderShape.toJSON((_5 = message.payload) === null || _5 === void 0 ? void 0 : _5.cylinderShape)
                : undefined);
        ((_6 = message.payload) === null || _6 === void 0 ? void 0 : _6.$case) === "gltfShape" &&
            (obj.gltfShape = ((_7 = message.payload) === null || _7 === void 0 ? void 0 : _7.gltfShape)
                ? components_ecs6_gen_1.ECS6ComponentGltfShape.toJSON((_8 = message.payload) === null || _8 === void 0 ? void 0 : _8.gltfShape)
                : undefined);
        ((_9 = message.payload) === null || _9 === void 0 ? void 0 : _9.$case) === "nftShape" &&
            (obj.nftShape = ((_10 = message.payload) === null || _10 === void 0 ? void 0 : _10.nftShape) ? components_ecs6_gen_1.ECS6ComponentNftShape.toJSON((_11 = message.payload) === null || _11 === void 0 ? void 0 : _11.nftShape) : undefined);
        ((_12 = message.payload) === null || _12 === void 0 ? void 0 : _12.$case) === "texture" &&
            (obj.texture = ((_13 = message.payload) === null || _13 === void 0 ? void 0 : _13.texture) ? components_ecs6_gen_1.ECS6ComponentTexture.toJSON((_14 = message.payload) === null || _14 === void 0 ? void 0 : _14.texture) : undefined);
        ((_15 = message.payload) === null || _15 === void 0 ? void 0 : _15.$case) === "animator" &&
            (obj.animator = ((_16 = message.payload) === null || _16 === void 0 ? void 0 : _16.animator) ? components_ecs6_gen_1.ECS6ComponentAnimator.toJSON((_17 = message.payload) === null || _17 === void 0 ? void 0 : _17.animator) : undefined);
        ((_18 = message.payload) === null || _18 === void 0 ? void 0 : _18.$case) === "objShape" &&
            (obj.objShape = ((_19 = message.payload) === null || _19 === void 0 ? void 0 : _19.objShape) ? components_ecs6_gen_1.ECS6ComponentObjShape.toJSON((_20 = message.payload) === null || _20 === void 0 ? void 0 : _20.objShape) : undefined);
        ((_21 = message.payload) === null || _21 === void 0 ? void 0 : _21.$case) === "font" &&
            (obj.font = ((_22 = message.payload) === null || _22 === void 0 ? void 0 : _22.font) ? components_ecs6_gen_1.ECS6ComponentFont.toJSON((_23 = message.payload) === null || _23 === void 0 ? void 0 : _23.font) : undefined);
        ((_24 = message.payload) === null || _24 === void 0 ? void 0 : _24.$case) === "textShape" &&
            (obj.textShape = ((_25 = message.payload) === null || _25 === void 0 ? void 0 : _25.textShape)
                ? components_ecs6_gen_1.ECS6ComponentTextShape.toJSON((_26 = message.payload) === null || _26 === void 0 ? void 0 : _26.textShape)
                : undefined);
        ((_27 = message.payload) === null || _27 === void 0 ? void 0 : _27.$case) === "material" &&
            (obj.material = ((_28 = message.payload) === null || _28 === void 0 ? void 0 : _28.material) ? components_ecs6_gen_1.ECS6ComponentMaterial.toJSON((_29 = message.payload) === null || _29 === void 0 ? void 0 : _29.material) : undefined);
        ((_30 = message.payload) === null || _30 === void 0 ? void 0 : _30.$case) === "basicMaterial" &&
            (obj.basicMaterial = ((_31 = message.payload) === null || _31 === void 0 ? void 0 : _31.basicMaterial)
                ? components_ecs6_gen_1.ECS6ComponentBasicMaterial.toJSON((_32 = message.payload) === null || _32 === void 0 ? void 0 : _32.basicMaterial)
                : undefined);
        ((_33 = message.payload) === null || _33 === void 0 ? void 0 : _33.$case) === "uuidCallback" &&
            (obj.uuidCallback = ((_34 = message.payload) === null || _34 === void 0 ? void 0 : _34.uuidCallback)
                ? components_ecs6_gen_1.ECS6ComponentUuidCallback.toJSON((_35 = message.payload) === null || _35 === void 0 ? void 0 : _35.uuidCallback)
                : undefined);
        ((_36 = message.payload) === null || _36 === void 0 ? void 0 : _36.$case) === "smartItem" &&
            (obj.smartItem = ((_37 = message.payload) === null || _37 === void 0 ? void 0 : _37.smartItem)
                ? components_ecs6_gen_1.ECS6ComponentSmartItem.toJSON((_38 = message.payload) === null || _38 === void 0 ? void 0 : _38.smartItem)
                : undefined);
        ((_39 = message.payload) === null || _39 === void 0 ? void 0 : _39.$case) === "videoClip" &&
            (obj.videoClip = ((_40 = message.payload) === null || _40 === void 0 ? void 0 : _40.videoClip)
                ? components_ecs6_gen_1.ECS6ComponentVideoClip.toJSON((_41 = message.payload) === null || _41 === void 0 ? void 0 : _41.videoClip)
                : undefined);
        ((_42 = message.payload) === null || _42 === void 0 ? void 0 : _42.$case) === "videoTexture" &&
            (obj.videoTexture = ((_43 = message.payload) === null || _43 === void 0 ? void 0 : _43.videoTexture)
                ? components_ecs6_gen_1.ECS6ComponentVideoTexture.toJSON((_44 = message.payload) === null || _44 === void 0 ? void 0 : _44.videoTexture)
                : undefined);
        ((_45 = message.payload) === null || _45 === void 0 ? void 0 : _45.$case) === "cameraModeArea" &&
            (obj.cameraModeArea = ((_46 = message.payload) === null || _46 === void 0 ? void 0 : _46.cameraModeArea)
                ? components_ecs6_gen_1.ECS6ComponentCameraModeArea.toJSON((_47 = message.payload) === null || _47 === void 0 ? void 0 : _47.cameraModeArea)
                : undefined);
        ((_48 = message.payload) === null || _48 === void 0 ? void 0 : _48.$case) === "avatarTexture" &&
            (obj.avatarTexture = ((_49 = message.payload) === null || _49 === void 0 ? void 0 : _49.avatarTexture)
                ? components_ecs6_gen_1.ECS6ComponentAvatarTexture.toJSON((_50 = message.payload) === null || _50 === void 0 ? void 0 : _50.avatarTexture)
                : undefined);
        ((_51 = message.payload) === null || _51 === void 0 ? void 0 : _51.$case) === "audioClip" &&
            (obj.audioClip = ((_52 = message.payload) === null || _52 === void 0 ? void 0 : _52.audioClip)
                ? components_ecs6_gen_1.ECS6ComponentAudioClip.toJSON((_53 = message.payload) === null || _53 === void 0 ? void 0 : _53.audioClip)
                : undefined);
        ((_54 = message.payload) === null || _54 === void 0 ? void 0 : _54.$case) === "audioSource" &&
            (obj.audioSource = ((_55 = message.payload) === null || _55 === void 0 ? void 0 : _55.audioSource)
                ? components_ecs6_gen_1.ECS6ComponentAudioSource.toJSON((_56 = message.payload) === null || _56 === void 0 ? void 0 : _56.audioSource)
                : undefined);
        ((_57 = message.payload) === null || _57 === void 0 ? void 0 : _57.$case) === "audioStream" &&
            (obj.audioStream = ((_58 = message.payload) === null || _58 === void 0 ? void 0 : _58.audioStream)
                ? components_ecs6_gen_1.ECS6ComponentAudioStream.toJSON((_59 = message.payload) === null || _59 === void 0 ? void 0 : _59.audioStream)
                : undefined);
        ((_60 = message.payload) === null || _60 === void 0 ? void 0 : _60.$case) === "avatarShape" &&
            (obj.avatarShape = ((_61 = message.payload) === null || _61 === void 0 ? void 0 : _61.avatarShape)
                ? components_ecs6_gen_1.ECS6ComponentAvatarShape.toJSON((_62 = message.payload) === null || _62 === void 0 ? void 0 : _62.avatarShape)
                : undefined);
        ((_63 = message.payload) === null || _63 === void 0 ? void 0 : _63.$case) === "gizmos" &&
            (obj.gizmos = ((_64 = message.payload) === null || _64 === void 0 ? void 0 : _64.gizmos) ? components_ecs6_gen_1.ECS6ComponentGizmos.toJSON((_65 = message.payload) === null || _65 === void 0 ? void 0 : _65.gizmos) : undefined);
        ((_66 = message.payload) === null || _66 === void 0 ? void 0 : _66.$case) === "uiShape" &&
            (obj.uiShape = ((_67 = message.payload) === null || _67 === void 0 ? void 0 : _67.uiShape) ? components_ecs6_gen_1.ECS6ComponentUiShape.toJSON((_68 = message.payload) === null || _68 === void 0 ? void 0 : _68.uiShape) : undefined);
        ((_69 = message.payload) === null || _69 === void 0 ? void 0 : _69.$case) === "uiContainerRect" &&
            (obj.uiContainerRect = ((_70 = message.payload) === null || _70 === void 0 ? void 0 : _70.uiContainerRect)
                ? components_ecs6_gen_1.ECS6ComponentUiContainerRect.toJSON((_71 = message.payload) === null || _71 === void 0 ? void 0 : _71.uiContainerRect)
                : undefined);
        ((_72 = message.payload) === null || _72 === void 0 ? void 0 : _72.$case) === "uiContainerStack" &&
            (obj.uiContainerStack = ((_73 = message.payload) === null || _73 === void 0 ? void 0 : _73.uiContainerStack)
                ? components_ecs6_gen_1.ECS6ComponentUiContainerStack.toJSON((_74 = message.payload) === null || _74 === void 0 ? void 0 : _74.uiContainerStack)
                : undefined);
        ((_75 = message.payload) === null || _75 === void 0 ? void 0 : _75.$case) === "uiButton" &&
            (obj.uiButton = ((_76 = message.payload) === null || _76 === void 0 ? void 0 : _76.uiButton) ? components_ecs6_gen_1.ECS6ComponentUiButton.toJSON((_77 = message.payload) === null || _77 === void 0 ? void 0 : _77.uiButton) : undefined);
        ((_78 = message.payload) === null || _78 === void 0 ? void 0 : _78.$case) === "uiText" &&
            (obj.uiText = ((_79 = message.payload) === null || _79 === void 0 ? void 0 : _79.uiText) ? components_ecs6_gen_1.ECS6ComponentUiText.toJSON((_80 = message.payload) === null || _80 === void 0 ? void 0 : _80.uiText) : undefined);
        ((_81 = message.payload) === null || _81 === void 0 ? void 0 : _81.$case) === "uiInputText" &&
            (obj.uiInputText = ((_82 = message.payload) === null || _82 === void 0 ? void 0 : _82.uiInputText)
                ? components_ecs6_gen_1.ECS6ComponentUiInputText.toJSON((_83 = message.payload) === null || _83 === void 0 ? void 0 : _83.uiInputText)
                : undefined);
        ((_84 = message.payload) === null || _84 === void 0 ? void 0 : _84.$case) === "uiImage" &&
            (obj.uiImage = ((_85 = message.payload) === null || _85 === void 0 ? void 0 : _85.uiImage) ? components_ecs6_gen_1.ECS6ComponentUiImage.toJSON((_86 = message.payload) === null || _86 === void 0 ? void 0 : _86.uiImage) : undefined);
        ((_87 = message.payload) === null || _87 === void 0 ? void 0 : _87.$case) === "uiScrollRect" &&
            (obj.uiScrollRect = ((_88 = message.payload) === null || _88 === void 0 ? void 0 : _88.uiScrollRect)
                ? components_ecs6_gen_1.ECS6ComponentUiScrollRect.toJSON((_89 = message.payload) === null || _89 === void 0 ? void 0 : _89.uiScrollRect)
                : undefined);
        ((_90 = message.payload) === null || _90 === void 0 ? void 0 : _90.$case) === "uiWorldSpaceShape" &&
            (obj.uiWorldSpaceShape = ((_91 = message.payload) === null || _91 === void 0 ? void 0 : _91.uiWorldSpaceShape)
                ? components_ecs6_gen_1.ECS6ComponentUiWorldSpaceShape.toJSON((_92 = message.payload) === null || _92 === void 0 ? void 0 : _92.uiWorldSpaceShape)
                : undefined);
        ((_93 = message.payload) === null || _93 === void 0 ? void 0 : _93.$case) === "uiScreenSpaceShape" &&
            (obj.uiScreenSpaceShape = ((_94 = message.payload) === null || _94 === void 0 ? void 0 : _94.uiScreenSpaceShape)
                ? components_ecs6_gen_1.ECS6ComponentUiScreenSpaceShape.toJSON((_95 = message.payload) === null || _95 === void 0 ? void 0 : _95.uiScreenSpaceShape)
                : undefined);
        ((_96 = message.payload) === null || _96 === void 0 ? void 0 : _96.$case) === "uiFullScreenShape" &&
            (obj.uiFullScreenShape = ((_97 = message.payload) === null || _97 === void 0 ? void 0 : _97.uiFullScreenShape)
                ? components_ecs6_gen_1.ECS6ComponentUiFullScreenShape.toJSON((_98 = message.payload) === null || _98 === void 0 ? void 0 : _98.uiFullScreenShape)
                : undefined);
        return obj;
    }
    ComponentBodyPayload.toJSON = toJSON;
    function create(base) {
        return ComponentBodyPayload.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    ComponentBodyPayload.create = create;
    function fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, _29, _30, _31, _32, _33, _34, _35, _36, _37, _38, _39, _40, _41, _42, _43, _44, _45, _46, _47, _48, _49, _50, _51, _52, _53, _54, _55, _56, _57, _58, _59, _60, _61, _62, _63, _64, _65, _66, _67, _68, _69, _70, _71, _72, _73, _74, _75, _76, _77, _78, _79, _80, _81, _82, _83, _84, _85, _86, _87, _88, _89, _90, _91, _92, _93, _94, _95, _96, _97, _98;
        const message = createBaseComponentBodyPayload();
        if (((_a = object.payload) === null || _a === void 0 ? void 0 : _a.$case) === "avatarModifierArea" &&
            ((_b = object.payload) === null || _b === void 0 ? void 0 : _b.avatarModifierArea) !== undefined &&
            ((_c = object.payload) === null || _c === void 0 ? void 0 : _c.avatarModifierArea) !== null) {
            message.payload = {
                $case: "avatarModifierArea",
                avatarModifierArea: components_ecs6_gen_1.ECS6ComponentAvatarModifierArea.fromPartial(object.payload.avatarModifierArea),
            };
        }
        if (((_d = object.payload) === null || _d === void 0 ? void 0 : _d.$case) === "transform" &&
            ((_e = object.payload) === null || _e === void 0 ? void 0 : _e.transform) !== undefined &&
            ((_f = object.payload) === null || _f === void 0 ? void 0 : _f.transform) !== null) {
            message.payload = { $case: "transform", transform: components_ecs6_gen_1.ECS6ComponentTransform.fromPartial(object.payload.transform) };
        }
        if (((_g = object.payload) === null || _g === void 0 ? void 0 : _g.$case) === "attachToAvatar" &&
            ((_h = object.payload) === null || _h === void 0 ? void 0 : _h.attachToAvatar) !== undefined &&
            ((_j = object.payload) === null || _j === void 0 ? void 0 : _j.attachToAvatar) !== null) {
            message.payload = {
                $case: "attachToAvatar",
                attachToAvatar: components_ecs6_gen_1.ECS6ComponentAttachToAvatar.fromPartial(object.payload.attachToAvatar),
            };
        }
        if (((_k = object.payload) === null || _k === void 0 ? void 0 : _k.$case) === "billboard" &&
            ((_l = object.payload) === null || _l === void 0 ? void 0 : _l.billboard) !== undefined &&
            ((_m = object.payload) === null || _m === void 0 ? void 0 : _m.billboard) !== null) {
            message.payload = { $case: "billboard", billboard: components_ecs6_gen_1.ECS6ComponentBillboard.fromPartial(object.payload.billboard) };
        }
        if (((_o = object.payload) === null || _o === void 0 ? void 0 : _o.$case) === "boxShape" &&
            ((_p = object.payload) === null || _p === void 0 ? void 0 : _p.boxShape) !== undefined &&
            ((_q = object.payload) === null || _q === void 0 ? void 0 : _q.boxShape) !== null) {
            message.payload = { $case: "boxShape", boxShape: components_ecs6_gen_1.ECS6ComponentBoxShape.fromPartial(object.payload.boxShape) };
        }
        if (((_r = object.payload) === null || _r === void 0 ? void 0 : _r.$case) === "sphereShape" &&
            ((_s = object.payload) === null || _s === void 0 ? void 0 : _s.sphereShape) !== undefined &&
            ((_t = object.payload) === null || _t === void 0 ? void 0 : _t.sphereShape) !== null) {
            message.payload = {
                $case: "sphereShape",
                sphereShape: components_ecs6_gen_1.ECS6ComponentSphereShape.fromPartial(object.payload.sphereShape),
            };
        }
        if (((_u = object.payload) === null || _u === void 0 ? void 0 : _u.$case) === "circleShape" &&
            ((_v = object.payload) === null || _v === void 0 ? void 0 : _v.circleShape) !== undefined &&
            ((_w = object.payload) === null || _w === void 0 ? void 0 : _w.circleShape) !== null) {
            message.payload = {
                $case: "circleShape",
                circleShape: components_ecs6_gen_1.ECS6ComponentCircleShape.fromPartial(object.payload.circleShape),
            };
        }
        if (((_x = object.payload) === null || _x === void 0 ? void 0 : _x.$case) === "planeShape" &&
            ((_y = object.payload) === null || _y === void 0 ? void 0 : _y.planeShape) !== undefined &&
            ((_z = object.payload) === null || _z === void 0 ? void 0 : _z.planeShape) !== null) {
            message.payload = {
                $case: "planeShape",
                planeShape: components_ecs6_gen_1.ECS6ComponentPlaneShape.fromPartial(object.payload.planeShape),
            };
        }
        if (((_0 = object.payload) === null || _0 === void 0 ? void 0 : _0.$case) === "coneShape" &&
            ((_1 = object.payload) === null || _1 === void 0 ? void 0 : _1.coneShape) !== undefined &&
            ((_2 = object.payload) === null || _2 === void 0 ? void 0 : _2.coneShape) !== null) {
            message.payload = { $case: "coneShape", coneShape: components_ecs6_gen_1.ECS6ComponentConeShape.fromPartial(object.payload.coneShape) };
        }
        if (((_3 = object.payload) === null || _3 === void 0 ? void 0 : _3.$case) === "cylinderShape" &&
            ((_4 = object.payload) === null || _4 === void 0 ? void 0 : _4.cylinderShape) !== undefined &&
            ((_5 = object.payload) === null || _5 === void 0 ? void 0 : _5.cylinderShape) !== null) {
            message.payload = {
                $case: "cylinderShape",
                cylinderShape: components_ecs6_gen_1.ECS6ComponentCylinderShape.fromPartial(object.payload.cylinderShape),
            };
        }
        if (((_6 = object.payload) === null || _6 === void 0 ? void 0 : _6.$case) === "gltfShape" &&
            ((_7 = object.payload) === null || _7 === void 0 ? void 0 : _7.gltfShape) !== undefined &&
            ((_8 = object.payload) === null || _8 === void 0 ? void 0 : _8.gltfShape) !== null) {
            message.payload = { $case: "gltfShape", gltfShape: components_ecs6_gen_1.ECS6ComponentGltfShape.fromPartial(object.payload.gltfShape) };
        }
        if (((_9 = object.payload) === null || _9 === void 0 ? void 0 : _9.$case) === "nftShape" &&
            ((_10 = object.payload) === null || _10 === void 0 ? void 0 : _10.nftShape) !== undefined &&
            ((_11 = object.payload) === null || _11 === void 0 ? void 0 : _11.nftShape) !== null) {
            message.payload = { $case: "nftShape", nftShape: components_ecs6_gen_1.ECS6ComponentNftShape.fromPartial(object.payload.nftShape) };
        }
        if (((_12 = object.payload) === null || _12 === void 0 ? void 0 : _12.$case) === "texture" && ((_13 = object.payload) === null || _13 === void 0 ? void 0 : _13.texture) !== undefined && ((_14 = object.payload) === null || _14 === void 0 ? void 0 : _14.texture) !== null) {
            message.payload = { $case: "texture", texture: components_ecs6_gen_1.ECS6ComponentTexture.fromPartial(object.payload.texture) };
        }
        if (((_15 = object.payload) === null || _15 === void 0 ? void 0 : _15.$case) === "animator" &&
            ((_16 = object.payload) === null || _16 === void 0 ? void 0 : _16.animator) !== undefined &&
            ((_17 = object.payload) === null || _17 === void 0 ? void 0 : _17.animator) !== null) {
            message.payload = { $case: "animator", animator: components_ecs6_gen_1.ECS6ComponentAnimator.fromPartial(object.payload.animator) };
        }
        if (((_18 = object.payload) === null || _18 === void 0 ? void 0 : _18.$case) === "objShape" &&
            ((_19 = object.payload) === null || _19 === void 0 ? void 0 : _19.objShape) !== undefined &&
            ((_20 = object.payload) === null || _20 === void 0 ? void 0 : _20.objShape) !== null) {
            message.payload = { $case: "objShape", objShape: components_ecs6_gen_1.ECS6ComponentObjShape.fromPartial(object.payload.objShape) };
        }
        if (((_21 = object.payload) === null || _21 === void 0 ? void 0 : _21.$case) === "font" && ((_22 = object.payload) === null || _22 === void 0 ? void 0 : _22.font) !== undefined && ((_23 = object.payload) === null || _23 === void 0 ? void 0 : _23.font) !== null) {
            message.payload = { $case: "font", font: components_ecs6_gen_1.ECS6ComponentFont.fromPartial(object.payload.font) };
        }
        if (((_24 = object.payload) === null || _24 === void 0 ? void 0 : _24.$case) === "textShape" &&
            ((_25 = object.payload) === null || _25 === void 0 ? void 0 : _25.textShape) !== undefined &&
            ((_26 = object.payload) === null || _26 === void 0 ? void 0 : _26.textShape) !== null) {
            message.payload = { $case: "textShape", textShape: components_ecs6_gen_1.ECS6ComponentTextShape.fromPartial(object.payload.textShape) };
        }
        if (((_27 = object.payload) === null || _27 === void 0 ? void 0 : _27.$case) === "material" &&
            ((_28 = object.payload) === null || _28 === void 0 ? void 0 : _28.material) !== undefined &&
            ((_29 = object.payload) === null || _29 === void 0 ? void 0 : _29.material) !== null) {
            message.payload = { $case: "material", material: components_ecs6_gen_1.ECS6ComponentMaterial.fromPartial(object.payload.material) };
        }
        if (((_30 = object.payload) === null || _30 === void 0 ? void 0 : _30.$case) === "basicMaterial" &&
            ((_31 = object.payload) === null || _31 === void 0 ? void 0 : _31.basicMaterial) !== undefined &&
            ((_32 = object.payload) === null || _32 === void 0 ? void 0 : _32.basicMaterial) !== null) {
            message.payload = {
                $case: "basicMaterial",
                basicMaterial: components_ecs6_gen_1.ECS6ComponentBasicMaterial.fromPartial(object.payload.basicMaterial),
            };
        }
        if (((_33 = object.payload) === null || _33 === void 0 ? void 0 : _33.$case) === "uuidCallback" &&
            ((_34 = object.payload) === null || _34 === void 0 ? void 0 : _34.uuidCallback) !== undefined &&
            ((_35 = object.payload) === null || _35 === void 0 ? void 0 : _35.uuidCallback) !== null) {
            message.payload = {
                $case: "uuidCallback",
                uuidCallback: components_ecs6_gen_1.ECS6ComponentUuidCallback.fromPartial(object.payload.uuidCallback),
            };
        }
        if (((_36 = object.payload) === null || _36 === void 0 ? void 0 : _36.$case) === "smartItem" &&
            ((_37 = object.payload) === null || _37 === void 0 ? void 0 : _37.smartItem) !== undefined &&
            ((_38 = object.payload) === null || _38 === void 0 ? void 0 : _38.smartItem) !== null) {
            message.payload = { $case: "smartItem", smartItem: components_ecs6_gen_1.ECS6ComponentSmartItem.fromPartial(object.payload.smartItem) };
        }
        if (((_39 = object.payload) === null || _39 === void 0 ? void 0 : _39.$case) === "videoClip" &&
            ((_40 = object.payload) === null || _40 === void 0 ? void 0 : _40.videoClip) !== undefined &&
            ((_41 = object.payload) === null || _41 === void 0 ? void 0 : _41.videoClip) !== null) {
            message.payload = { $case: "videoClip", videoClip: components_ecs6_gen_1.ECS6ComponentVideoClip.fromPartial(object.payload.videoClip) };
        }
        if (((_42 = object.payload) === null || _42 === void 0 ? void 0 : _42.$case) === "videoTexture" &&
            ((_43 = object.payload) === null || _43 === void 0 ? void 0 : _43.videoTexture) !== undefined &&
            ((_44 = object.payload) === null || _44 === void 0 ? void 0 : _44.videoTexture) !== null) {
            message.payload = {
                $case: "videoTexture",
                videoTexture: components_ecs6_gen_1.ECS6ComponentVideoTexture.fromPartial(object.payload.videoTexture),
            };
        }
        if (((_45 = object.payload) === null || _45 === void 0 ? void 0 : _45.$case) === "cameraModeArea" &&
            ((_46 = object.payload) === null || _46 === void 0 ? void 0 : _46.cameraModeArea) !== undefined &&
            ((_47 = object.payload) === null || _47 === void 0 ? void 0 : _47.cameraModeArea) !== null) {
            message.payload = {
                $case: "cameraModeArea",
                cameraModeArea: components_ecs6_gen_1.ECS6ComponentCameraModeArea.fromPartial(object.payload.cameraModeArea),
            };
        }
        if (((_48 = object.payload) === null || _48 === void 0 ? void 0 : _48.$case) === "avatarTexture" &&
            ((_49 = object.payload) === null || _49 === void 0 ? void 0 : _49.avatarTexture) !== undefined &&
            ((_50 = object.payload) === null || _50 === void 0 ? void 0 : _50.avatarTexture) !== null) {
            message.payload = {
                $case: "avatarTexture",
                avatarTexture: components_ecs6_gen_1.ECS6ComponentAvatarTexture.fromPartial(object.payload.avatarTexture),
            };
        }
        if (((_51 = object.payload) === null || _51 === void 0 ? void 0 : _51.$case) === "audioClip" &&
            ((_52 = object.payload) === null || _52 === void 0 ? void 0 : _52.audioClip) !== undefined &&
            ((_53 = object.payload) === null || _53 === void 0 ? void 0 : _53.audioClip) !== null) {
            message.payload = { $case: "audioClip", audioClip: components_ecs6_gen_1.ECS6ComponentAudioClip.fromPartial(object.payload.audioClip) };
        }
        if (((_54 = object.payload) === null || _54 === void 0 ? void 0 : _54.$case) === "audioSource" &&
            ((_55 = object.payload) === null || _55 === void 0 ? void 0 : _55.audioSource) !== undefined &&
            ((_56 = object.payload) === null || _56 === void 0 ? void 0 : _56.audioSource) !== null) {
            message.payload = {
                $case: "audioSource",
                audioSource: components_ecs6_gen_1.ECS6ComponentAudioSource.fromPartial(object.payload.audioSource),
            };
        }
        if (((_57 = object.payload) === null || _57 === void 0 ? void 0 : _57.$case) === "audioStream" &&
            ((_58 = object.payload) === null || _58 === void 0 ? void 0 : _58.audioStream) !== undefined &&
            ((_59 = object.payload) === null || _59 === void 0 ? void 0 : _59.audioStream) !== null) {
            message.payload = {
                $case: "audioStream",
                audioStream: components_ecs6_gen_1.ECS6ComponentAudioStream.fromPartial(object.payload.audioStream),
            };
        }
        if (((_60 = object.payload) === null || _60 === void 0 ? void 0 : _60.$case) === "avatarShape" &&
            ((_61 = object.payload) === null || _61 === void 0 ? void 0 : _61.avatarShape) !== undefined &&
            ((_62 = object.payload) === null || _62 === void 0 ? void 0 : _62.avatarShape) !== null) {
            message.payload = {
                $case: "avatarShape",
                avatarShape: components_ecs6_gen_1.ECS6ComponentAvatarShape.fromPartial(object.payload.avatarShape),
            };
        }
        if (((_63 = object.payload) === null || _63 === void 0 ? void 0 : _63.$case) === "gizmos" && ((_64 = object.payload) === null || _64 === void 0 ? void 0 : _64.gizmos) !== undefined && ((_65 = object.payload) === null || _65 === void 0 ? void 0 : _65.gizmos) !== null) {
            message.payload = { $case: "gizmos", gizmos: components_ecs6_gen_1.ECS6ComponentGizmos.fromPartial(object.payload.gizmos) };
        }
        if (((_66 = object.payload) === null || _66 === void 0 ? void 0 : _66.$case) === "uiShape" && ((_67 = object.payload) === null || _67 === void 0 ? void 0 : _67.uiShape) !== undefined && ((_68 = object.payload) === null || _68 === void 0 ? void 0 : _68.uiShape) !== null) {
            message.payload = { $case: "uiShape", uiShape: components_ecs6_gen_1.ECS6ComponentUiShape.fromPartial(object.payload.uiShape) };
        }
        if (((_69 = object.payload) === null || _69 === void 0 ? void 0 : _69.$case) === "uiContainerRect" &&
            ((_70 = object.payload) === null || _70 === void 0 ? void 0 : _70.uiContainerRect) !== undefined &&
            ((_71 = object.payload) === null || _71 === void 0 ? void 0 : _71.uiContainerRect) !== null) {
            message.payload = {
                $case: "uiContainerRect",
                uiContainerRect: components_ecs6_gen_1.ECS6ComponentUiContainerRect.fromPartial(object.payload.uiContainerRect),
            };
        }
        if (((_72 = object.payload) === null || _72 === void 0 ? void 0 : _72.$case) === "uiContainerStack" &&
            ((_73 = object.payload) === null || _73 === void 0 ? void 0 : _73.uiContainerStack) !== undefined &&
            ((_74 = object.payload) === null || _74 === void 0 ? void 0 : _74.uiContainerStack) !== null) {
            message.payload = {
                $case: "uiContainerStack",
                uiContainerStack: components_ecs6_gen_1.ECS6ComponentUiContainerStack.fromPartial(object.payload.uiContainerStack),
            };
        }
        if (((_75 = object.payload) === null || _75 === void 0 ? void 0 : _75.$case) === "uiButton" &&
            ((_76 = object.payload) === null || _76 === void 0 ? void 0 : _76.uiButton) !== undefined &&
            ((_77 = object.payload) === null || _77 === void 0 ? void 0 : _77.uiButton) !== null) {
            message.payload = { $case: "uiButton", uiButton: components_ecs6_gen_1.ECS6ComponentUiButton.fromPartial(object.payload.uiButton) };
        }
        if (((_78 = object.payload) === null || _78 === void 0 ? void 0 : _78.$case) === "uiText" && ((_79 = object.payload) === null || _79 === void 0 ? void 0 : _79.uiText) !== undefined && ((_80 = object.payload) === null || _80 === void 0 ? void 0 : _80.uiText) !== null) {
            message.payload = { $case: "uiText", uiText: components_ecs6_gen_1.ECS6ComponentUiText.fromPartial(object.payload.uiText) };
        }
        if (((_81 = object.payload) === null || _81 === void 0 ? void 0 : _81.$case) === "uiInputText" &&
            ((_82 = object.payload) === null || _82 === void 0 ? void 0 : _82.uiInputText) !== undefined &&
            ((_83 = object.payload) === null || _83 === void 0 ? void 0 : _83.uiInputText) !== null) {
            message.payload = {
                $case: "uiInputText",
                uiInputText: components_ecs6_gen_1.ECS6ComponentUiInputText.fromPartial(object.payload.uiInputText),
            };
        }
        if (((_84 = object.payload) === null || _84 === void 0 ? void 0 : _84.$case) === "uiImage" && ((_85 = object.payload) === null || _85 === void 0 ? void 0 : _85.uiImage) !== undefined && ((_86 = object.payload) === null || _86 === void 0 ? void 0 : _86.uiImage) !== null) {
            message.payload = { $case: "uiImage", uiImage: components_ecs6_gen_1.ECS6ComponentUiImage.fromPartial(object.payload.uiImage) };
        }
        if (((_87 = object.payload) === null || _87 === void 0 ? void 0 : _87.$case) === "uiScrollRect" &&
            ((_88 = object.payload) === null || _88 === void 0 ? void 0 : _88.uiScrollRect) !== undefined &&
            ((_89 = object.payload) === null || _89 === void 0 ? void 0 : _89.uiScrollRect) !== null) {
            message.payload = {
                $case: "uiScrollRect",
                uiScrollRect: components_ecs6_gen_1.ECS6ComponentUiScrollRect.fromPartial(object.payload.uiScrollRect),
            };
        }
        if (((_90 = object.payload) === null || _90 === void 0 ? void 0 : _90.$case) === "uiWorldSpaceShape" &&
            ((_91 = object.payload) === null || _91 === void 0 ? void 0 : _91.uiWorldSpaceShape) !== undefined &&
            ((_92 = object.payload) === null || _92 === void 0 ? void 0 : _92.uiWorldSpaceShape) !== null) {
            message.payload = {
                $case: "uiWorldSpaceShape",
                uiWorldSpaceShape: components_ecs6_gen_1.ECS6ComponentUiWorldSpaceShape.fromPartial(object.payload.uiWorldSpaceShape),
            };
        }
        if (((_93 = object.payload) === null || _93 === void 0 ? void 0 : _93.$case) === "uiScreenSpaceShape" &&
            ((_94 = object.payload) === null || _94 === void 0 ? void 0 : _94.uiScreenSpaceShape) !== undefined &&
            ((_95 = object.payload) === null || _95 === void 0 ? void 0 : _95.uiScreenSpaceShape) !== null) {
            message.payload = {
                $case: "uiScreenSpaceShape",
                uiScreenSpaceShape: components_ecs6_gen_1.ECS6ComponentUiScreenSpaceShape.fromPartial(object.payload.uiScreenSpaceShape),
            };
        }
        if (((_96 = object.payload) === null || _96 === void 0 ? void 0 : _96.$case) === "uiFullScreenShape" &&
            ((_97 = object.payload) === null || _97 === void 0 ? void 0 : _97.uiFullScreenShape) !== undefined &&
            ((_98 = object.payload) === null || _98 === void 0 ? void 0 : _98.uiFullScreenShape) !== null) {
            message.payload = {
                $case: "uiFullScreenShape",
                uiFullScreenShape: components_ecs6_gen_1.ECS6ComponentUiFullScreenShape.fromPartial(object.payload.uiFullScreenShape),
            };
        }
        return message;
    }
    ComponentBodyPayload.fromPartial = fromPartial;
})(ComponentBodyPayload || (exports.ComponentBodyPayload = ComponentBodyPayload = {}));
function createBaseCreateEntityBody() {
    return { id: "" };
}
var CreateEntityBody;
(function (CreateEntityBody) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.id !== "") {
            writer.uint32(10).string(message.id);
        }
        return writer;
    }
    CreateEntityBody.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCreateEntityBody();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.id = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    CreateEntityBody.decode = decode;
    function fromJSON(object) {
        return { id: isSet(object.id) ? String(object.id) : "" };
    }
    CreateEntityBody.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.id !== undefined && (obj.id = message.id);
        return obj;
    }
    CreateEntityBody.toJSON = toJSON;
    function create(base) {
        return CreateEntityBody.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    CreateEntityBody.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseCreateEntityBody();
        message.id = (_a = object.id) !== null && _a !== void 0 ? _a : "";
        return message;
    }
    CreateEntityBody.fromPartial = fromPartial;
})(CreateEntityBody || (exports.CreateEntityBody = CreateEntityBody = {}));
function createBaseRemoveEntityBody() {
    return { id: "" };
}
var RemoveEntityBody;
(function (RemoveEntityBody) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.id !== "") {
            writer.uint32(10).string(message.id);
        }
        return writer;
    }
    RemoveEntityBody.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRemoveEntityBody();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.id = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    RemoveEntityBody.decode = decode;
    function fromJSON(object) {
        return { id: isSet(object.id) ? String(object.id) : "" };
    }
    RemoveEntityBody.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.id !== undefined && (obj.id = message.id);
        return obj;
    }
    RemoveEntityBody.toJSON = toJSON;
    function create(base) {
        return RemoveEntityBody.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    RemoveEntityBody.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseRemoveEntityBody();
        message.id = (_a = object.id) !== null && _a !== void 0 ? _a : "";
        return message;
    }
    RemoveEntityBody.fromPartial = fromPartial;
})(RemoveEntityBody || (exports.RemoveEntityBody = RemoveEntityBody = {}));
function createBaseUpdateEntityComponentBody() {
    return { entityId: "", classId: 0, name: "", componentData: undefined };
}
var UpdateEntityComponentBody;
(function (UpdateEntityComponentBody) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.entityId !== "") {
            writer.uint32(10).string(message.entityId);
        }
        if (message.classId !== 0) {
            writer.uint32(16).int32(message.classId);
        }
        if (message.name !== "") {
            writer.uint32(26).string(message.name);
        }
        if (message.componentData !== undefined) {
            ComponentBodyPayload.encode(message.componentData, writer.uint32(34).fork()).ldelim();
        }
        return writer;
    }
    UpdateEntityComponentBody.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseUpdateEntityComponentBody();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.entityId = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.classId = reader.int32();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.componentData = ComponentBodyPayload.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    UpdateEntityComponentBody.decode = decode;
    function fromJSON(object) {
        return {
            entityId: isSet(object.entityId) ? String(object.entityId) : "",
            classId: isSet(object.classId) ? Number(object.classId) : 0,
            name: isSet(object.name) ? String(object.name) : "",
            componentData: isSet(object.componentData) ? ComponentBodyPayload.fromJSON(object.componentData) : undefined,
        };
    }
    UpdateEntityComponentBody.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.entityId !== undefined && (obj.entityId = message.entityId);
        message.classId !== undefined && (obj.classId = Math.round(message.classId));
        message.name !== undefined && (obj.name = message.name);
        message.componentData !== undefined &&
            (obj.componentData = message.componentData ? ComponentBodyPayload.toJSON(message.componentData) : undefined);
        return obj;
    }
    UpdateEntityComponentBody.toJSON = toJSON;
    function create(base) {
        return UpdateEntityComponentBody.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    UpdateEntityComponentBody.create = create;
    function fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseUpdateEntityComponentBody();
        message.entityId = (_a = object.entityId) !== null && _a !== void 0 ? _a : "";
        message.classId = (_b = object.classId) !== null && _b !== void 0 ? _b : 0;
        message.name = (_c = object.name) !== null && _c !== void 0 ? _c : "";
        message.componentData = (object.componentData !== undefined && object.componentData !== null)
            ? ComponentBodyPayload.fromPartial(object.componentData)
            : undefined;
        return message;
    }
    UpdateEntityComponentBody.fromPartial = fromPartial;
})(UpdateEntityComponentBody || (exports.UpdateEntityComponentBody = UpdateEntityComponentBody = {}));
function createBaseAttachEntityComponentBody() {
    return { entityId: "", name: "", id: "" };
}
var AttachEntityComponentBody;
(function (AttachEntityComponentBody) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.entityId !== "") {
            writer.uint32(10).string(message.entityId);
        }
        if (message.name !== "") {
            writer.uint32(18).string(message.name);
        }
        if (message.id !== "") {
            writer.uint32(26).string(message.id);
        }
        return writer;
    }
    AttachEntityComponentBody.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAttachEntityComponentBody();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.entityId = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.id = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    AttachEntityComponentBody.decode = decode;
    function fromJSON(object) {
        return {
            entityId: isSet(object.entityId) ? String(object.entityId) : "",
            name: isSet(object.name) ? String(object.name) : "",
            id: isSet(object.id) ? String(object.id) : "",
        };
    }
    AttachEntityComponentBody.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.entityId !== undefined && (obj.entityId = message.entityId);
        message.name !== undefined && (obj.name = message.name);
        message.id !== undefined && (obj.id = message.id);
        return obj;
    }
    AttachEntityComponentBody.toJSON = toJSON;
    function create(base) {
        return AttachEntityComponentBody.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    AttachEntityComponentBody.create = create;
    function fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseAttachEntityComponentBody();
        message.entityId = (_a = object.entityId) !== null && _a !== void 0 ? _a : "";
        message.name = (_b = object.name) !== null && _b !== void 0 ? _b : "";
        message.id = (_c = object.id) !== null && _c !== void 0 ? _c : "";
        return message;
    }
    AttachEntityComponentBody.fromPartial = fromPartial;
})(AttachEntityComponentBody || (exports.AttachEntityComponentBody = AttachEntityComponentBody = {}));
function createBaseComponentRemovedBody() {
    return { entityId: "", name: "" };
}
var ComponentRemovedBody;
(function (ComponentRemovedBody) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.entityId !== "") {
            writer.uint32(10).string(message.entityId);
        }
        if (message.name !== "") {
            writer.uint32(18).string(message.name);
        }
        return writer;
    }
    ComponentRemovedBody.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseComponentRemovedBody();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.entityId = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    ComponentRemovedBody.decode = decode;
    function fromJSON(object) {
        return {
            entityId: isSet(object.entityId) ? String(object.entityId) : "",
            name: isSet(object.name) ? String(object.name) : "",
        };
    }
    ComponentRemovedBody.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.entityId !== undefined && (obj.entityId = message.entityId);
        message.name !== undefined && (obj.name = message.name);
        return obj;
    }
    ComponentRemovedBody.toJSON = toJSON;
    function create(base) {
        return ComponentRemovedBody.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    ComponentRemovedBody.create = create;
    function fromPartial(object) {
        var _a, _b;
        const message = createBaseComponentRemovedBody();
        message.entityId = (_a = object.entityId) !== null && _a !== void 0 ? _a : "";
        message.name = (_b = object.name) !== null && _b !== void 0 ? _b : "";
        return message;
    }
    ComponentRemovedBody.fromPartial = fromPartial;
})(ComponentRemovedBody || (exports.ComponentRemovedBody = ComponentRemovedBody = {}));
function createBaseSetEntityParentBody() {
    return { entityId: "", parentId: "" };
}
var SetEntityParentBody;
(function (SetEntityParentBody) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.entityId !== "") {
            writer.uint32(10).string(message.entityId);
        }
        if (message.parentId !== "") {
            writer.uint32(18).string(message.parentId);
        }
        return writer;
    }
    SetEntityParentBody.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSetEntityParentBody();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.entityId = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.parentId = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    SetEntityParentBody.decode = decode;
    function fromJSON(object) {
        return {
            entityId: isSet(object.entityId) ? String(object.entityId) : "",
            parentId: isSet(object.parentId) ? String(object.parentId) : "",
        };
    }
    SetEntityParentBody.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.entityId !== undefined && (obj.entityId = message.entityId);
        message.parentId !== undefined && (obj.parentId = message.parentId);
        return obj;
    }
    SetEntityParentBody.toJSON = toJSON;
    function create(base) {
        return SetEntityParentBody.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    SetEntityParentBody.create = create;
    function fromPartial(object) {
        var _a, _b;
        const message = createBaseSetEntityParentBody();
        message.entityId = (_a = object.entityId) !== null && _a !== void 0 ? _a : "";
        message.parentId = (_b = object.parentId) !== null && _b !== void 0 ? _b : "";
        return message;
    }
    SetEntityParentBody.fromPartial = fromPartial;
})(SetEntityParentBody || (exports.SetEntityParentBody = SetEntityParentBody = {}));
function createBaseQueryBody() {
    return { queryId: "", payload: undefined };
}
var QueryBody;
(function (QueryBody) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.queryId !== "") {
            writer.uint32(10).string(message.queryId);
        }
        if (message.payload !== undefined) {
            QueryBody_RayQuery.encode(message.payload, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    }
    QueryBody.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryBody();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.queryId = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.payload = QueryBody_RayQuery.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    QueryBody.decode = decode;
    function fromJSON(object) {
        return {
            queryId: isSet(object.queryId) ? String(object.queryId) : "",
            payload: isSet(object.payload) ? QueryBody_RayQuery.fromJSON(object.payload) : undefined,
        };
    }
    QueryBody.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.queryId !== undefined && (obj.queryId = message.queryId);
        message.payload !== undefined &&
            (obj.payload = message.payload ? QueryBody_RayQuery.toJSON(message.payload) : undefined);
        return obj;
    }
    QueryBody.toJSON = toJSON;
    function create(base) {
        return QueryBody.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    QueryBody.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseQueryBody();
        message.queryId = (_a = object.queryId) !== null && _a !== void 0 ? _a : "";
        message.payload = (object.payload !== undefined && object.payload !== null)
            ? QueryBody_RayQuery.fromPartial(object.payload)
            : undefined;
        return message;
    }
    QueryBody.fromPartial = fromPartial;
})(QueryBody || (exports.QueryBody = QueryBody = {}));
function createBaseQueryBody_Ray() {
    return { origin: undefined, direction: undefined, distance: 0 };
}
var QueryBody_Ray;
(function (QueryBody_Ray) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.origin !== undefined) {
            vectors_gen_1.Vector3.encode(message.origin, writer.uint32(10).fork()).ldelim();
        }
        if (message.direction !== undefined) {
            vectors_gen_1.Vector3.encode(message.direction, writer.uint32(18).fork()).ldelim();
        }
        if (message.distance !== 0) {
            writer.uint32(29).float(message.distance);
        }
        return writer;
    }
    QueryBody_Ray.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryBody_Ray();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.origin = vectors_gen_1.Vector3.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.direction = vectors_gen_1.Vector3.decode(reader, reader.uint32());
                    continue;
                case 3:
                    if (tag !== 29) {
                        break;
                    }
                    message.distance = reader.float();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    QueryBody_Ray.decode = decode;
    function fromJSON(object) {
        return {
            origin: isSet(object.origin) ? vectors_gen_1.Vector3.fromJSON(object.origin) : undefined,
            direction: isSet(object.direction) ? vectors_gen_1.Vector3.fromJSON(object.direction) : undefined,
            distance: isSet(object.distance) ? Number(object.distance) : 0,
        };
    }
    QueryBody_Ray.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.origin !== undefined && (obj.origin = message.origin ? vectors_gen_1.Vector3.toJSON(message.origin) : undefined);
        message.direction !== undefined &&
            (obj.direction = message.direction ? vectors_gen_1.Vector3.toJSON(message.direction) : undefined);
        message.distance !== undefined && (obj.distance = message.distance);
        return obj;
    }
    QueryBody_Ray.toJSON = toJSON;
    function create(base) {
        return QueryBody_Ray.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    QueryBody_Ray.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseQueryBody_Ray();
        message.origin = (object.origin !== undefined && object.origin !== null)
            ? vectors_gen_1.Vector3.fromPartial(object.origin)
            : undefined;
        message.direction = (object.direction !== undefined && object.direction !== null)
            ? vectors_gen_1.Vector3.fromPartial(object.direction)
            : undefined;
        message.distance = (_a = object.distance) !== null && _a !== void 0 ? _a : 0;
        return message;
    }
    QueryBody_Ray.fromPartial = fromPartial;
})(QueryBody_Ray || (exports.QueryBody_Ray = QueryBody_Ray = {}));
function createBaseQueryBody_RayQuery() {
    return { queryId: "", queryType: "", ray: undefined };
}
var QueryBody_RayQuery;
(function (QueryBody_RayQuery) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.queryId !== "") {
            writer.uint32(10).string(message.queryId);
        }
        if (message.queryType !== "") {
            writer.uint32(18).string(message.queryType);
        }
        if (message.ray !== undefined) {
            QueryBody_Ray.encode(message.ray, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    }
    QueryBody_RayQuery.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryBody_RayQuery();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.queryId = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.queryType = reader.string();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.ray = QueryBody_Ray.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    QueryBody_RayQuery.decode = decode;
    function fromJSON(object) {
        return {
            queryId: isSet(object.queryId) ? String(object.queryId) : "",
            queryType: isSet(object.queryType) ? String(object.queryType) : "",
            ray: isSet(object.ray) ? QueryBody_Ray.fromJSON(object.ray) : undefined,
        };
    }
    QueryBody_RayQuery.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.queryId !== undefined && (obj.queryId = message.queryId);
        message.queryType !== undefined && (obj.queryType = message.queryType);
        message.ray !== undefined && (obj.ray = message.ray ? QueryBody_Ray.toJSON(message.ray) : undefined);
        return obj;
    }
    QueryBody_RayQuery.toJSON = toJSON;
    function create(base) {
        return QueryBody_RayQuery.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    QueryBody_RayQuery.create = create;
    function fromPartial(object) {
        var _a, _b;
        const message = createBaseQueryBody_RayQuery();
        message.queryId = (_a = object.queryId) !== null && _a !== void 0 ? _a : "";
        message.queryType = (_b = object.queryType) !== null && _b !== void 0 ? _b : "";
        message.ray = (object.ray !== undefined && object.ray !== null) ? QueryBody_Ray.fromPartial(object.ray) : undefined;
        return message;
    }
    QueryBody_RayQuery.fromPartial = fromPartial;
})(QueryBody_RayQuery || (exports.QueryBody_RayQuery = QueryBody_RayQuery = {}));
function createBaseComponentCreatedBody() {
    return { id: "", classId: 0, name: "" };
}
var ComponentCreatedBody;
(function (ComponentCreatedBody) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.id !== "") {
            writer.uint32(10).string(message.id);
        }
        if (message.classId !== 0) {
            writer.uint32(16).int32(message.classId);
        }
        if (message.name !== "") {
            writer.uint32(26).string(message.name);
        }
        return writer;
    }
    ComponentCreatedBody.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseComponentCreatedBody();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.id = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.classId = reader.int32();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    ComponentCreatedBody.decode = decode;
    function fromJSON(object) {
        return {
            id: isSet(object.id) ? String(object.id) : "",
            classId: isSet(object.classId) ? Number(object.classId) : 0,
            name: isSet(object.name) ? String(object.name) : "",
        };
    }
    ComponentCreatedBody.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.id !== undefined && (obj.id = message.id);
        message.classId !== undefined && (obj.classId = Math.round(message.classId));
        message.name !== undefined && (obj.name = message.name);
        return obj;
    }
    ComponentCreatedBody.toJSON = toJSON;
    function create(base) {
        return ComponentCreatedBody.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    ComponentCreatedBody.create = create;
    function fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseComponentCreatedBody();
        message.id = (_a = object.id) !== null && _a !== void 0 ? _a : "";
        message.classId = (_b = object.classId) !== null && _b !== void 0 ? _b : 0;
        message.name = (_c = object.name) !== null && _c !== void 0 ? _c : "";
        return message;
    }
    ComponentCreatedBody.fromPartial = fromPartial;
})(ComponentCreatedBody || (exports.ComponentCreatedBody = ComponentCreatedBody = {}));
function createBaseComponentDisposedBody() {
    return { id: "" };
}
var ComponentDisposedBody;
(function (ComponentDisposedBody) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.id !== "") {
            writer.uint32(10).string(message.id);
        }
        return writer;
    }
    ComponentDisposedBody.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseComponentDisposedBody();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.id = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    ComponentDisposedBody.decode = decode;
    function fromJSON(object) {
        return { id: isSet(object.id) ? String(object.id) : "" };
    }
    ComponentDisposedBody.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.id !== undefined && (obj.id = message.id);
        return obj;
    }
    ComponentDisposedBody.toJSON = toJSON;
    function create(base) {
        return ComponentDisposedBody.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    ComponentDisposedBody.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseComponentDisposedBody();
        message.id = (_a = object.id) !== null && _a !== void 0 ? _a : "";
        return message;
    }
    ComponentDisposedBody.fromPartial = fromPartial;
})(ComponentDisposedBody || (exports.ComponentDisposedBody = ComponentDisposedBody = {}));
function createBaseComponentUpdatedBody() {
    return { id: "", componentData: undefined };
}
var ComponentUpdatedBody;
(function (ComponentUpdatedBody) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.id !== "") {
            writer.uint32(10).string(message.id);
        }
        if (message.componentData !== undefined) {
            ComponentBodyPayload.encode(message.componentData, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    }
    ComponentUpdatedBody.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseComponentUpdatedBody();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.id = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.componentData = ComponentBodyPayload.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    ComponentUpdatedBody.decode = decode;
    function fromJSON(object) {
        return {
            id: isSet(object.id) ? String(object.id) : "",
            componentData: isSet(object.componentData) ? ComponentBodyPayload.fromJSON(object.componentData) : undefined,
        };
    }
    ComponentUpdatedBody.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.id !== undefined && (obj.id = message.id);
        message.componentData !== undefined &&
            (obj.componentData = message.componentData ? ComponentBodyPayload.toJSON(message.componentData) : undefined);
        return obj;
    }
    ComponentUpdatedBody.toJSON = toJSON;
    function create(base) {
        return ComponentUpdatedBody.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    ComponentUpdatedBody.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseComponentUpdatedBody();
        message.id = (_a = object.id) !== null && _a !== void 0 ? _a : "";
        message.componentData = (object.componentData !== undefined && object.componentData !== null)
            ? ComponentBodyPayload.fromPartial(object.componentData)
            : undefined;
        return message;
    }
    ComponentUpdatedBody.fromPartial = fromPartial;
})(ComponentUpdatedBody || (exports.ComponentUpdatedBody = ComponentUpdatedBody = {}));
function createBaseInitMessagesFinishedBody() {
    return {};
}
var InitMessagesFinishedBody;
(function (InitMessagesFinishedBody) {
    function encode(_, writer = minimal_1.default.Writer.create()) {
        return writer;
    }
    InitMessagesFinishedBody.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInitMessagesFinishedBody();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    InitMessagesFinishedBody.decode = decode;
    function fromJSON(_) {
        return {};
    }
    InitMessagesFinishedBody.fromJSON = fromJSON;
    function toJSON(_) {
        const obj = {};
        return obj;
    }
    InitMessagesFinishedBody.toJSON = toJSON;
    function create(base) {
        return InitMessagesFinishedBody.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    InitMessagesFinishedBody.create = create;
    function fromPartial(_) {
        const message = createBaseInitMessagesFinishedBody();
        return message;
    }
    InitMessagesFinishedBody.fromPartial = fromPartial;
})(InitMessagesFinishedBody || (exports.InitMessagesFinishedBody = InitMessagesFinishedBody = {}));
function createBaseEntityActionPayload() {
    return { payload: undefined };
}
var EntityActionPayload;
(function (EntityActionPayload) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        var _a;
        switch ((_a = message.payload) === null || _a === void 0 ? void 0 : _a.$case) {
            case "openExternalUrl":
                OpenExternalUrlBody.encode(message.payload.openExternalUrl, writer.uint32(10).fork()).ldelim();
                break;
            case "openNftDialog":
                OpenNFTDialogBody.encode(message.payload.openNftDialog, writer.uint32(18).fork()).ldelim();
                break;
            case "createEntity":
                CreateEntityBody.encode(message.payload.createEntity, writer.uint32(26).fork()).ldelim();
                break;
            case "removeEntity":
                RemoveEntityBody.encode(message.payload.removeEntity, writer.uint32(34).fork()).ldelim();
                break;
            case "updateEntityComponent":
                UpdateEntityComponentBody.encode(message.payload.updateEntityComponent, writer.uint32(42).fork()).ldelim();
                break;
            case "attachEntityComponent":
                AttachEntityComponentBody.encode(message.payload.attachEntityComponent, writer.uint32(50).fork()).ldelim();
                break;
            case "componentRemoved":
                ComponentRemovedBody.encode(message.payload.componentRemoved, writer.uint32(58).fork()).ldelim();
                break;
            case "setEntityParent":
                SetEntityParentBody.encode(message.payload.setEntityParent, writer.uint32(66).fork()).ldelim();
                break;
            case "query":
                QueryBody.encode(message.payload.query, writer.uint32(74).fork()).ldelim();
                break;
            case "componentCreated":
                ComponentCreatedBody.encode(message.payload.componentCreated, writer.uint32(82).fork()).ldelim();
                break;
            case "componentDisposed":
                ComponentDisposedBody.encode(message.payload.componentDisposed, writer.uint32(90).fork()).ldelim();
                break;
            case "componentUpdated":
                ComponentUpdatedBody.encode(message.payload.componentUpdated, writer.uint32(98).fork()).ldelim();
                break;
            case "initMessagesFinished":
                InitMessagesFinishedBody.encode(message.payload.initMessagesFinished, writer.uint32(106).fork()).ldelim();
                break;
        }
        return writer;
    }
    EntityActionPayload.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseEntityActionPayload();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.payload = {
                        $case: "openExternalUrl",
                        openExternalUrl: OpenExternalUrlBody.decode(reader, reader.uint32()),
                    };
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.payload = {
                        $case: "openNftDialog",
                        openNftDialog: OpenNFTDialogBody.decode(reader, reader.uint32()),
                    };
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.payload = { $case: "createEntity", createEntity: CreateEntityBody.decode(reader, reader.uint32()) };
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.payload = { $case: "removeEntity", removeEntity: RemoveEntityBody.decode(reader, reader.uint32()) };
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.payload = {
                        $case: "updateEntityComponent",
                        updateEntityComponent: UpdateEntityComponentBody.decode(reader, reader.uint32()),
                    };
                    continue;
                case 6:
                    if (tag !== 50) {
                        break;
                    }
                    message.payload = {
                        $case: "attachEntityComponent",
                        attachEntityComponent: AttachEntityComponentBody.decode(reader, reader.uint32()),
                    };
                    continue;
                case 7:
                    if (tag !== 58) {
                        break;
                    }
                    message.payload = {
                        $case: "componentRemoved",
                        componentRemoved: ComponentRemovedBody.decode(reader, reader.uint32()),
                    };
                    continue;
                case 8:
                    if (tag !== 66) {
                        break;
                    }
                    message.payload = {
                        $case: "setEntityParent",
                        setEntityParent: SetEntityParentBody.decode(reader, reader.uint32()),
                    };
                    continue;
                case 9:
                    if (tag !== 74) {
                        break;
                    }
                    message.payload = { $case: "query", query: QueryBody.decode(reader, reader.uint32()) };
                    continue;
                case 10:
                    if (tag !== 82) {
                        break;
                    }
                    message.payload = {
                        $case: "componentCreated",
                        componentCreated: ComponentCreatedBody.decode(reader, reader.uint32()),
                    };
                    continue;
                case 11:
                    if (tag !== 90) {
                        break;
                    }
                    message.payload = {
                        $case: "componentDisposed",
                        componentDisposed: ComponentDisposedBody.decode(reader, reader.uint32()),
                    };
                    continue;
                case 12:
                    if (tag !== 98) {
                        break;
                    }
                    message.payload = {
                        $case: "componentUpdated",
                        componentUpdated: ComponentUpdatedBody.decode(reader, reader.uint32()),
                    };
                    continue;
                case 13:
                    if (tag !== 106) {
                        break;
                    }
                    message.payload = {
                        $case: "initMessagesFinished",
                        initMessagesFinished: InitMessagesFinishedBody.decode(reader, reader.uint32()),
                    };
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    EntityActionPayload.decode = decode;
    function fromJSON(object) {
        return {
            payload: isSet(object.openExternalUrl)
                ? { $case: "openExternalUrl", openExternalUrl: OpenExternalUrlBody.fromJSON(object.openExternalUrl) }
                : isSet(object.openNftDialog)
                    ? { $case: "openNftDialog", openNftDialog: OpenNFTDialogBody.fromJSON(object.openNftDialog) }
                    : isSet(object.createEntity)
                        ? { $case: "createEntity", createEntity: CreateEntityBody.fromJSON(object.createEntity) }
                        : isSet(object.removeEntity)
                            ? { $case: "removeEntity", removeEntity: RemoveEntityBody.fromJSON(object.removeEntity) }
                            : isSet(object.updateEntityComponent)
                                ? {
                                    $case: "updateEntityComponent",
                                    updateEntityComponent: UpdateEntityComponentBody.fromJSON(object.updateEntityComponent),
                                }
                                : isSet(object.attachEntityComponent)
                                    ? {
                                        $case: "attachEntityComponent",
                                        attachEntityComponent: AttachEntityComponentBody.fromJSON(object.attachEntityComponent),
                                    }
                                    : isSet(object.componentRemoved)
                                        ? { $case: "componentRemoved", componentRemoved: ComponentRemovedBody.fromJSON(object.componentRemoved) }
                                        : isSet(object.setEntityParent)
                                            ? { $case: "setEntityParent", setEntityParent: SetEntityParentBody.fromJSON(object.setEntityParent) }
                                            : isSet(object.query)
                                                ? { $case: "query", query: QueryBody.fromJSON(object.query) }
                                                : isSet(object.componentCreated)
                                                    ? { $case: "componentCreated", componentCreated: ComponentCreatedBody.fromJSON(object.componentCreated) }
                                                    : isSet(object.componentDisposed)
                                                        ? { $case: "componentDisposed", componentDisposed: ComponentDisposedBody.fromJSON(object.componentDisposed) }
                                                        : isSet(object.componentUpdated)
                                                            ? { $case: "componentUpdated", componentUpdated: ComponentUpdatedBody.fromJSON(object.componentUpdated) }
                                                            : isSet(object.initMessagesFinished)
                                                                ? {
                                                                    $case: "initMessagesFinished",
                                                                    initMessagesFinished: InitMessagesFinishedBody.fromJSON(object.initMessagesFinished),
                                                                }
                                                                : undefined,
        };
    }
    EntityActionPayload.fromJSON = fromJSON;
    function toJSON(message) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14;
        const obj = {};
        ((_a = message.payload) === null || _a === void 0 ? void 0 : _a.$case) === "openExternalUrl" &&
            (obj.openExternalUrl = ((_b = message.payload) === null || _b === void 0 ? void 0 : _b.openExternalUrl)
                ? OpenExternalUrlBody.toJSON((_c = message.payload) === null || _c === void 0 ? void 0 : _c.openExternalUrl)
                : undefined);
        ((_d = message.payload) === null || _d === void 0 ? void 0 : _d.$case) === "openNftDialog" &&
            (obj.openNftDialog = ((_e = message.payload) === null || _e === void 0 ? void 0 : _e.openNftDialog)
                ? OpenNFTDialogBody.toJSON((_f = message.payload) === null || _f === void 0 ? void 0 : _f.openNftDialog)
                : undefined);
        ((_g = message.payload) === null || _g === void 0 ? void 0 : _g.$case) === "createEntity" &&
            (obj.createEntity = ((_h = message.payload) === null || _h === void 0 ? void 0 : _h.createEntity)
                ? CreateEntityBody.toJSON((_j = message.payload) === null || _j === void 0 ? void 0 : _j.createEntity)
                : undefined);
        ((_k = message.payload) === null || _k === void 0 ? void 0 : _k.$case) === "removeEntity" &&
            (obj.removeEntity = ((_l = message.payload) === null || _l === void 0 ? void 0 : _l.removeEntity)
                ? RemoveEntityBody.toJSON((_m = message.payload) === null || _m === void 0 ? void 0 : _m.removeEntity)
                : undefined);
        ((_o = message.payload) === null || _o === void 0 ? void 0 : _o.$case) === "updateEntityComponent" &&
            (obj.updateEntityComponent = ((_p = message.payload) === null || _p === void 0 ? void 0 : _p.updateEntityComponent)
                ? UpdateEntityComponentBody.toJSON((_q = message.payload) === null || _q === void 0 ? void 0 : _q.updateEntityComponent)
                : undefined);
        ((_r = message.payload) === null || _r === void 0 ? void 0 : _r.$case) === "attachEntityComponent" &&
            (obj.attachEntityComponent = ((_s = message.payload) === null || _s === void 0 ? void 0 : _s.attachEntityComponent)
                ? AttachEntityComponentBody.toJSON((_t = message.payload) === null || _t === void 0 ? void 0 : _t.attachEntityComponent)
                : undefined);
        ((_u = message.payload) === null || _u === void 0 ? void 0 : _u.$case) === "componentRemoved" &&
            (obj.componentRemoved = ((_v = message.payload) === null || _v === void 0 ? void 0 : _v.componentRemoved)
                ? ComponentRemovedBody.toJSON((_w = message.payload) === null || _w === void 0 ? void 0 : _w.componentRemoved)
                : undefined);
        ((_x = message.payload) === null || _x === void 0 ? void 0 : _x.$case) === "setEntityParent" &&
            (obj.setEntityParent = ((_y = message.payload) === null || _y === void 0 ? void 0 : _y.setEntityParent)
                ? SetEntityParentBody.toJSON((_z = message.payload) === null || _z === void 0 ? void 0 : _z.setEntityParent)
                : undefined);
        ((_0 = message.payload) === null || _0 === void 0 ? void 0 : _0.$case) === "query" &&
            (obj.query = ((_1 = message.payload) === null || _1 === void 0 ? void 0 : _1.query) ? QueryBody.toJSON((_2 = message.payload) === null || _2 === void 0 ? void 0 : _2.query) : undefined);
        ((_3 = message.payload) === null || _3 === void 0 ? void 0 : _3.$case) === "componentCreated" &&
            (obj.componentCreated = ((_4 = message.payload) === null || _4 === void 0 ? void 0 : _4.componentCreated)
                ? ComponentCreatedBody.toJSON((_5 = message.payload) === null || _5 === void 0 ? void 0 : _5.componentCreated)
                : undefined);
        ((_6 = message.payload) === null || _6 === void 0 ? void 0 : _6.$case) === "componentDisposed" &&
            (obj.componentDisposed = ((_7 = message.payload) === null || _7 === void 0 ? void 0 : _7.componentDisposed)
                ? ComponentDisposedBody.toJSON((_8 = message.payload) === null || _8 === void 0 ? void 0 : _8.componentDisposed)
                : undefined);
        ((_9 = message.payload) === null || _9 === void 0 ? void 0 : _9.$case) === "componentUpdated" &&
            (obj.componentUpdated = ((_10 = message.payload) === null || _10 === void 0 ? void 0 : _10.componentUpdated)
                ? ComponentUpdatedBody.toJSON((_11 = message.payload) === null || _11 === void 0 ? void 0 : _11.componentUpdated)
                : undefined);
        ((_12 = message.payload) === null || _12 === void 0 ? void 0 : _12.$case) === "initMessagesFinished" &&
            (obj.initMessagesFinished = ((_13 = message.payload) === null || _13 === void 0 ? void 0 : _13.initMessagesFinished)
                ? InitMessagesFinishedBody.toJSON((_14 = message.payload) === null || _14 === void 0 ? void 0 : _14.initMessagesFinished)
                : undefined);
        return obj;
    }
    EntityActionPayload.toJSON = toJSON;
    function create(base) {
        return EntityActionPayload.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    EntityActionPayload.create = create;
    function fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14;
        const message = createBaseEntityActionPayload();
        if (((_a = object.payload) === null || _a === void 0 ? void 0 : _a.$case) === "openExternalUrl" &&
            ((_b = object.payload) === null || _b === void 0 ? void 0 : _b.openExternalUrl) !== undefined &&
            ((_c = object.payload) === null || _c === void 0 ? void 0 : _c.openExternalUrl) !== null) {
            message.payload = {
                $case: "openExternalUrl",
                openExternalUrl: OpenExternalUrlBody.fromPartial(object.payload.openExternalUrl),
            };
        }
        if (((_d = object.payload) === null || _d === void 0 ? void 0 : _d.$case) === "openNftDialog" &&
            ((_e = object.payload) === null || _e === void 0 ? void 0 : _e.openNftDialog) !== undefined &&
            ((_f = object.payload) === null || _f === void 0 ? void 0 : _f.openNftDialog) !== null) {
            message.payload = {
                $case: "openNftDialog",
                openNftDialog: OpenNFTDialogBody.fromPartial(object.payload.openNftDialog),
            };
        }
        if (((_g = object.payload) === null || _g === void 0 ? void 0 : _g.$case) === "createEntity" &&
            ((_h = object.payload) === null || _h === void 0 ? void 0 : _h.createEntity) !== undefined &&
            ((_j = object.payload) === null || _j === void 0 ? void 0 : _j.createEntity) !== null) {
            message.payload = {
                $case: "createEntity",
                createEntity: CreateEntityBody.fromPartial(object.payload.createEntity),
            };
        }
        if (((_k = object.payload) === null || _k === void 0 ? void 0 : _k.$case) === "removeEntity" &&
            ((_l = object.payload) === null || _l === void 0 ? void 0 : _l.removeEntity) !== undefined &&
            ((_m = object.payload) === null || _m === void 0 ? void 0 : _m.removeEntity) !== null) {
            message.payload = {
                $case: "removeEntity",
                removeEntity: RemoveEntityBody.fromPartial(object.payload.removeEntity),
            };
        }
        if (((_o = object.payload) === null || _o === void 0 ? void 0 : _o.$case) === "updateEntityComponent" &&
            ((_p = object.payload) === null || _p === void 0 ? void 0 : _p.updateEntityComponent) !== undefined &&
            ((_q = object.payload) === null || _q === void 0 ? void 0 : _q.updateEntityComponent) !== null) {
            message.payload = {
                $case: "updateEntityComponent",
                updateEntityComponent: UpdateEntityComponentBody.fromPartial(object.payload.updateEntityComponent),
            };
        }
        if (((_r = object.payload) === null || _r === void 0 ? void 0 : _r.$case) === "attachEntityComponent" &&
            ((_s = object.payload) === null || _s === void 0 ? void 0 : _s.attachEntityComponent) !== undefined &&
            ((_t = object.payload) === null || _t === void 0 ? void 0 : _t.attachEntityComponent) !== null) {
            message.payload = {
                $case: "attachEntityComponent",
                attachEntityComponent: AttachEntityComponentBody.fromPartial(object.payload.attachEntityComponent),
            };
        }
        if (((_u = object.payload) === null || _u === void 0 ? void 0 : _u.$case) === "componentRemoved" &&
            ((_v = object.payload) === null || _v === void 0 ? void 0 : _v.componentRemoved) !== undefined &&
            ((_w = object.payload) === null || _w === void 0 ? void 0 : _w.componentRemoved) !== null) {
            message.payload = {
                $case: "componentRemoved",
                componentRemoved: ComponentRemovedBody.fromPartial(object.payload.componentRemoved),
            };
        }
        if (((_x = object.payload) === null || _x === void 0 ? void 0 : _x.$case) === "setEntityParent" &&
            ((_y = object.payload) === null || _y === void 0 ? void 0 : _y.setEntityParent) !== undefined &&
            ((_z = object.payload) === null || _z === void 0 ? void 0 : _z.setEntityParent) !== null) {
            message.payload = {
                $case: "setEntityParent",
                setEntityParent: SetEntityParentBody.fromPartial(object.payload.setEntityParent),
            };
        }
        if (((_0 = object.payload) === null || _0 === void 0 ? void 0 : _0.$case) === "query" && ((_1 = object.payload) === null || _1 === void 0 ? void 0 : _1.query) !== undefined && ((_2 = object.payload) === null || _2 === void 0 ? void 0 : _2.query) !== null) {
            message.payload = { $case: "query", query: QueryBody.fromPartial(object.payload.query) };
        }
        if (((_3 = object.payload) === null || _3 === void 0 ? void 0 : _3.$case) === "componentCreated" &&
            ((_4 = object.payload) === null || _4 === void 0 ? void 0 : _4.componentCreated) !== undefined &&
            ((_5 = object.payload) === null || _5 === void 0 ? void 0 : _5.componentCreated) !== null) {
            message.payload = {
                $case: "componentCreated",
                componentCreated: ComponentCreatedBody.fromPartial(object.payload.componentCreated),
            };
        }
        if (((_6 = object.payload) === null || _6 === void 0 ? void 0 : _6.$case) === "componentDisposed" &&
            ((_7 = object.payload) === null || _7 === void 0 ? void 0 : _7.componentDisposed) !== undefined &&
            ((_8 = object.payload) === null || _8 === void 0 ? void 0 : _8.componentDisposed) !== null) {
            message.payload = {
                $case: "componentDisposed",
                componentDisposed: ComponentDisposedBody.fromPartial(object.payload.componentDisposed),
            };
        }
        if (((_9 = object.payload) === null || _9 === void 0 ? void 0 : _9.$case) === "componentUpdated" &&
            ((_10 = object.payload) === null || _10 === void 0 ? void 0 : _10.componentUpdated) !== undefined &&
            ((_11 = object.payload) === null || _11 === void 0 ? void 0 : _11.componentUpdated) !== null) {
            message.payload = {
                $case: "componentUpdated",
                componentUpdated: ComponentUpdatedBody.fromPartial(object.payload.componentUpdated),
            };
        }
        if (((_12 = object.payload) === null || _12 === void 0 ? void 0 : _12.$case) === "initMessagesFinished" &&
            ((_13 = object.payload) === null || _13 === void 0 ? void 0 : _13.initMessagesFinished) !== undefined &&
            ((_14 = object.payload) === null || _14 === void 0 ? void 0 : _14.initMessagesFinished) !== null) {
            message.payload = {
                $case: "initMessagesFinished",
                initMessagesFinished: InitMessagesFinishedBody.fromPartial(object.payload.initMessagesFinished),
            };
        }
        return message;
    }
    EntityActionPayload.fromPartial = fromPartial;
})(EntityActionPayload || (exports.EntityActionPayload = EntityActionPayload = {}));
function createBaseEntityAction() {
    return { tag: undefined, payload: undefined };
}
var EntityAction;
(function (EntityAction) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.tag !== undefined) {
            writer.uint32(18).string(message.tag);
        }
        if (message.payload !== undefined) {
            EntityActionPayload.encode(message.payload, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    }
    EntityAction.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseEntityAction();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.tag = reader.string();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.payload = EntityActionPayload.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    EntityAction.decode = decode;
    function fromJSON(object) {
        return {
            tag: isSet(object.tag) ? String(object.tag) : undefined,
            payload: isSet(object.payload) ? EntityActionPayload.fromJSON(object.payload) : undefined,
        };
    }
    EntityAction.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.tag !== undefined && (obj.tag = message.tag);
        message.payload !== undefined &&
            (obj.payload = message.payload ? EntityActionPayload.toJSON(message.payload) : undefined);
        return obj;
    }
    EntityAction.toJSON = toJSON;
    function create(base) {
        return EntityAction.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    EntityAction.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseEntityAction();
        message.tag = (_a = object.tag) !== null && _a !== void 0 ? _a : undefined;
        message.payload = (object.payload !== undefined && object.payload !== null)
            ? EntityActionPayload.fromPartial(object.payload)
            : undefined;
        return message;
    }
    EntityAction.fromPartial = fromPartial;
})(EntityAction || (exports.EntityAction = EntityAction = {}));
function isSet(value) {
    return value !== null && value !== undefined;
}
