// @dcl/ecs/dist/index.js
// The order of the following imports matters. Please do not auto-sort
export * from './engine';
export * from './schemas';
export * from './runtime/initialization';
export * from './runtime/types';
export * from './runtime/helpers';
export { cyclicParentingChecker } from './systems/cyclicParentingChecker';
export * from './systems/events';
export * from './systems/raycast';
export * from './systems/videoEvents';
export * from './systems/async-task';
export * from './systems/tween';
export * from './engine/entity';
export * from './components/types';
// @internal
import * as components from './components';
// @internal
export { components };
import { engine } from './runtime/initialization';
// export components for global engine
export const Transform = /* @__PURE__*/ components.Transform(engine);
export const Animator = /* @__PURE__*/ components.Animator(engine);
export const AudioSource = /* @__PURE__*/ components.AudioSource(engine);
export const AudioStream = /* @__PURE__*/ components.AudioStream(engine);
export const Material = /* @__PURE__*/ components.Material(engine);
export const MeshRenderer = /* @__PURE__*/ components.MeshRenderer(engine);
export const MeshCollider = /* @__PURE__*/ components.MeshCollider(engine);
export const Name = components.Name(engine);
export const Tween = /* @__PURE__*/ components.Tween(engine);
export const VirtualCamera = /* @__PURE__*/ components.VirtualCamera(engine);
export const InputModifier = /* @__PURE__*/ components.InputModifier(engine);
/**
 * @alpha
 * This is going to be used for sync components through a server.
 * Can be modified in the future since it's still in research
 */
export const SyncComponents = /* @__PURE__*/ components.SyncComponents(engine);
/**
 * @alpha
 * Tag a entity to be syncronized through comms
 */
export const NetworkEntity = /* @__PURE__*/ components.NetworkEntity(engine);
/**
 * @alpha
 * Tag a entity to be syncronized through comms
 */
export const NetworkParent = /* @__PURE__*/ components.NetworkParent(engine);
// export components for global engine
export * from './components/generated/global.gen';
export * from './components/generated/types.gen';
export * from './serialization/crdt';
export * from './composite';
