// dcl-npc-toolkit/dist/config.js
import { Color4 } from "@dcl/sdk/math";
export let extraFemaleShoes = [
    "urn:decentraland:off-chain:base-avatars:bun_shoes",
    "urn:decentraland:off-chain:base-avatars:pink_blue_socks",
    "urn:decentraland:off-chain:base-avatars:SchoolShoes",
    "urn:decentraland:off-chain:base-avatars:ruby_blue_loafer"
];
export let extraFemalePants = [
    "urn:decentraland:off-chain:base-avatars:f_stripe_white_pants",
    "urn:decentraland:off-chain:base-avatars:f_red_modern_pants",
    "urn:decentraland:off-chain:base-avatars:f_school_skirt",
    "urn:decentraland:off-chain:base-avatars:f_capris",
    "urn:decentraland:off-chain:base-avatars:f_red_comfy_pants"
];
export let extraFemaleTop = [
    "urn:decentraland:off-chain:base-avatars:baggy_pullover",
    "urn:decentraland:off-chain:base-avatars:light_green_shirt",
    "urn:decentraland:off-chain:base-avatars:school_shirt",
    "urn:decentraland:off-chain:base-avatars:f_blue_jacket",
    "urn:decentraland:matic:collections-v2:0xb8326d69371668bb8403463e4f71c10b408535c3:1"
];
export let extraMaleShoes = [
    "urn:decentraland:off-chain:base-avatars:comfy_sport_sandals",
    "urn:decentraland:off-chain:base-avatars:crocs",
    "urn:decentraland:off-chain:base-avatars:m_mountainshoes.glb",
    "urn:decentraland:off-chain:base-avatars:citycomfortableshoes",
    "urn:decentraland:off-chain:base-avatars:classic_shoes",
    "urn:decentraland:off-chain:base-avatars:sneakers",
    "urn:decentraland:off-chain:base-avatars:moccasin"
];
export let extraMalePants = [
    "urn:decentraland:off-chain:base-avatars:soccer_pants",
    "urn:decentraland:off-chain:base-avatars:swim_short",
    "urn:decentraland:off-chain:base-avatars:kilt",
    "urn:decentraland:off-chain:base-avatars:hip_hop_joggers",
    "urn:decentraland:off-chain:base-avatars:brown_pants_02",
    "urn:decentraland:off-chain:base-avatars:basketball_shorts",
    "urn:decentraland:off-chain:base-avatars:brown_pants_02",
    "urn:decentraland:off-chain:base-avatars:puffer_jacket_hoodie"
];
export let extraMaleTop = [
    "urn:decentraland:off-chain:base-avatars:yellow_tshirt",
    "urn:decentraland:off-chain:base-avatars:m_sweater",
    "urn:decentraland:off-chain:base-avatars:turtle_neck_sweater",
    "urn:decentraland:off-chain:base-avatars:puffer_jacket_hoodie",
    "urn:decentraland:off-chain:base-avatars:elegant_sweater",
    "urn:decentraland:off-chain:base-avatars:pride_tshirt",
    "urn:decentraland:off-chain:base-avatars:elegant_sweater",
    "urn:decentraland:matic:collections-v2:0xb8326d69371668bb8403463e4f71c10b408535c3:1"
];
export let extraskin = [
    Color4.create(.24, .13, .04),
    Color4.create(.6, .37, .22),
    Color4.create(.48, .28, .09),
    Color4.create(.98, .82, .51),
];
export let extrahair = [
    Color4.create(.6, .37, .22),
    Color4.create(.3, .52, .4),
    Color4.create(.55, .13, .08),
    Color4.create(.11, .11, .11),
    Color4.create(.48, .28, .09),
    Color4.create(.98, .82, .51),
    Color4.create(.55, .13, .08),
];
export let extraFemaleHairstyle = [
    "urn:decentraland:off-chain:base-avatars:hair_anime_01",
    "urn:decentraland:off-chain:base-avatars:pony_tail",
    "urn:decentraland:off-chain:base-avatars:standard_hair",
    "urn:decentraland:off-chain:base-avatars:shoulder_hair",
    "urn:decentraland:off-chain:base-avatars:two_tails"
];
export let extraMaleHairstyle = [
    "urn:decentraland:off-chain:base-avatars:standard_hair",
    "urn:decentraland:off-chain:base-avatars:hair_oldie",
    "urn:decentraland:off-chain:base-avatars:keanu_hair",
    "urn:decentraland:off-chain:base-avatars:tall_front_01",
    "urn:decentraland:off-chain:base-avatars:00_bald"
];
export let extraeyes = [
    Color4.create(.22, .49, .69),
    Color4.create(.6, .37, .22),
    Color4.create(.3, .52, .4),
    Color4.create(.55, .13, .08),
    Color4.create(.11, .11, .11),
    Color4.create(.48, .28, .09),
    Color4.create(.98, .82, .51),
];
