// dcl-npc-toolkit/dist/types.js
export var ButtonStyles;
(function (ButtonStyles) {
    ButtonStyles["E"] = "e";
    ButtonStyles["F"] = "f";
    ButtonStyles["DARK"] = "dark";
    ButtonStyles["RED"] = "red";
    ButtonStyles["ROUNDBLACK"] = "roundblack";
    ButtonStyles["ROUNDWHITE"] = "roundwhite";
    ButtonStyles["ROUNDSILVER"] = "roundsilver";
    ButtonStyles["ROUNDGOLD"] = "roundgold";
    ButtonStyles["SQUAREBLACK"] = "squareblack";
    ButtonStyles["SQUAREWHITE"] = "squarewhite";
    ButtonStyles["SQUARESILVER"] = "squaresilver";
    ButtonStyles["SQUAREGOLD"] = "squaregold";
    ButtonStyles["WHITE"] = "white";
})(ButtonStyles || (ButtonStyles = {}));
export var NPCState;
(function (NPCState) {
    NPCState["STANDING"] = "standing";
    NPCState["TALKING"] = "talking";
    NPCState["FOLLOWPATH"] = "followPath";
})(NPCState || (NPCState = {}));
export var NPCBodyType;
(function (NPCBodyType) {
    NPCBodyType["MALE"] = "BaseMale";
    NPCBodyType["FEMALE"] = "BaseFemale";
})(NPCBodyType || (NPCBodyType = {}));
export var NPCType;
(function (NPCType) {
    NPCType["BLANK"] = "blank";
    NPCType["CUSTOM"] = "custom";
    NPCType["AVATAR"] = "avatar";
})(NPCType || (NPCType = {}));
export var NPCPathType;
(function (NPCPathType) {
    NPCPathType["SMOOTH_PATH"] = "smooth";
    NPCPathType["RIGID_PATH"] = "rigid";
})(NPCPathType || (NPCPathType = {}));
