// dcl-npc-toolkit/dist/ui.js
import ReactEcs, { UiEntity } from '@dcl/sdk/react-ecs';
import { buttonClick, displayButton, displayDialog, displayFirstButtonContainer, displayImage, displayPortrait, displaySecondButtonContainer, displaySkipable, getButtonText, getbuttonSize, getFontSize, getButtonFontSize, getImage, getImageAtlasMapping, getLeftClickTheme, getPortrait, getSkipableTheme, getText, getTextColor, getTextPosition, getTheme, handleDialogClick, imageHeight, imageWidth, portraitHeight, portraitWidth, positionImageX, positionImageY, positionPortaitX, positionPortaitY, realHeight, realWidth, skipDialogs, getWindowHeight } from './dialog';
import { sourcesComponentsCoordinates } from './uiResources';
import { activeNPC } from './npc';
export let lightTheme = 'https://decentraland.org/images/ui/light-atlas-v3.png';
export let darkTheme = 'https://decentraland.org/images/ui/dark-atlas-v3.png';
export let bubblesTexture = 'https://decentraland.org/images/ui/dialog-bubbles.png';
export let section = {
    ...sourcesComponentsCoordinates.backgrounds.NPCDialog,
    atlasHeight: sourcesComponentsCoordinates.atlasHeight,
    atlasWidth: sourcesComponentsCoordinates.atlasWidth
};
export let skipButtonSection = {
    ...sourcesComponentsCoordinates.buttons.F,
    atlasHeight: sourcesComponentsCoordinates.atlasHeight,
    atlasWidth: sourcesComponentsCoordinates.atlasWidth
};
export let skipButtonSectionBlack = {
    ...sourcesComponentsCoordinates.buttons.FBlack,
    atlasHeight: sourcesComponentsCoordinates.atlasHeight,
    atlasWidth: sourcesComponentsCoordinates.atlasWidth
};
export let leftClickSection = {
    ...sourcesComponentsCoordinates.icons.ClickWhite,
    atlasHeight: sourcesComponentsCoordinates.atlasHeight,
    atlasWidth: sourcesComponentsCoordinates.atlasWidth
};
export let leftClickSectionbBlack = {
    ...sourcesComponentsCoordinates.icons.ClickDark,
    atlasHeight: sourcesComponentsCoordinates.atlasHeight,
    atlasWidth: sourcesComponentsCoordinates.atlasWidth
};
export let primaryButtonSection = {
    ...sourcesComponentsCoordinates.buttons.E,
    atlasHeight: sourcesComponentsCoordinates.atlasHeight,
    atlasWidth: sourcesComponentsCoordinates.atlasWidth
};
export let secondaryButtonSection = {
    ...sourcesComponentsCoordinates.buttons.F,
    atlasHeight: sourcesComponentsCoordinates.atlasHeight,
    atlasWidth: sourcesComponentsCoordinates.atlasWidth
};
export let darkButtonSection = {
    ...sourcesComponentsCoordinates.buttons.buttonF,
    atlasHeight: sourcesComponentsCoordinates.atlasHeight,
    atlasWidth: sourcesComponentsCoordinates.atlasWidth
};
export let darkButtonCorner = {
    ...sourcesComponentsCoordinates.buttons.buttonFCorner,
    atlasHeight: sourcesComponentsCoordinates.atlasHeight,
    atlasWidth: sourcesComponentsCoordinates.atlasWidth
};
export let darkButtonEdge = {
    ...sourcesComponentsCoordinates.buttons.buttonFEdge,
    atlasHeight: sourcesComponentsCoordinates.atlasHeight,
    atlasWidth: sourcesComponentsCoordinates.atlasWidth
};
export let redButtonSection = {
    ...sourcesComponentsCoordinates.buttons.buttonE,
    atlasHeight: sourcesComponentsCoordinates.atlasHeight,
    atlasWidth: sourcesComponentsCoordinates.atlasWidth
};
export let redButtonCorner = {
    ...sourcesComponentsCoordinates.buttons.buttonECorner,
    atlasHeight: sourcesComponentsCoordinates.atlasHeight,
    atlasWidth: sourcesComponentsCoordinates.atlasWidth
};
export let redButtonEdge = {
    ...sourcesComponentsCoordinates.buttons.buttonEEdge,
    atlasHeight: sourcesComponentsCoordinates.atlasHeight,
    atlasWidth: sourcesComponentsCoordinates.atlasWidth
};
let modalScale = 1;
let modelFontSizeScale = 1;
let modalTextWrapScale = 1;
export function setupNPCUiScaling(inModalScale, inFontSize, inModalTextWrapScale) {
    modalScale = inModalScale;
    modelFontSizeScale = inFontSize;
    modalTextWrapScale = inModalTextWrapScale;
    console.log('NPC-TOOLKIT', 'Scale UI:', modalScale, 'TextFontSize:', modelFontSizeScale, 'TextWrapScaling:', modalTextWrapScale);
}
function getScaledSize(size) {
    return size * modalScale;
}
function getScaledFontSize(size) {
    return size * modelFontSizeScale;
}
function getScaledTextWrap(size) {
    return size * modalTextWrapScale;
}
function getScaledButtonWidth(button) {
    return typeof (getbuttonSize(button)) === 'number' ? getScaledSize(getbuttonSize(button)) : 'auto';
}
export const NpcUtilsUi = () => {
    return (ReactEcs.createElement(UiEntity, {
        uiTransform: {
            display: displayDialog() ? 'flex' : 'none',
            flexDirection: 'column',
            alignItems: 'center',
            justifyContent: 'center',
            positionType: 'absolute',
            position: { bottom: '10%', left: '25vw' },
            padding: { top: 40, bottom: 40 },
            width: '40vw',
            height: '14vw'
        }
    },
        ReactEcs.createElement(UiEntity, {
            uiTransform: {
                positionType: 'absolute',
                position: { top: 0, left: 0 },
                width: '100%',
                height: '100%'
            }, uiBackground: {
                textureMode: 'stretch',
                texture: {
                    src: 'images/Dialogue.png'
                }
            }, onMouseDown: () => {
                handleDialogClick();
            }
        }),
        ReactEcs.createElement(UiEntity, {
            uiTransform: {
                display: displayPortrait() ? 'flex' : 'none',
                width: '35%',
                height: '70%',
                positionType: 'absolute',
                position: {
                    bottom: '15%',
                    right: '70%'
                }
            }, uiBackground: {
                textureMode: 'stretch',
                texture: {
                    src: getPortrait()
                }
            }
        }),
        ReactEcs.createElement(UiEntity, {
            uiTransform: {
                display: displayImage() ? 'flex' : 'none',
                width: getScaledSize(imageWidth()),
                height: getScaledSize(imageHeight()),
                positionType: 'absolute',
                position: { bottom: positionImageY(), right: positionImageX() }
            }, uiBackground: {
                textureMode: 'stretch',
                texture: {
                    src: getImage()
                }
            }
        }),
        ReactEcs.createElement(UiEntity, {
            uiTransform: {
                display: displaySkipable() ? 'flex' : 'none',
                width: getScaledSize(15),
                height: getScaledSize(15),
                positionType: 'absolute',
                position: { bottom: '7%', left: '25%' }
            }, uiBackground: {
                textureMode: 'stretch',
                texture: {
                    src: getTheme()
                },
                uvs: getImageAtlasMapping(getSkipableTheme())
            }, onMouseDown: () => {
                skipDialogs(activeNPC);
            }
        },
            ReactEcs.createElement(UiEntity, {
                uiTransform: {
                    display: 'flex',
                    positionType: 'absolute',
                    position: { left: '215%' }
                }, uiText: {
                    value: 'Skip',
                    color: getTextColor(),
                    fontSize: getScaledFontSize(12)
                }
            })),
        ReactEcs.createElement(UiEntity, {
            uiTransform: {
                display: 'flex',
                width: getScaledSize(24),
                height: getScaledSize(36),
                positionType: 'absolute',
                position: { bottom: '5%', right: '2%' }
            }
        }),
        ReactEcs.createElement(UiEntity, {
            uiTransform: {
                height: 'auto',
                width: 'auto',
                alignSelf: 'flex-start',
                alignItems: 'flex-start',
                justifyContent: 'flex-start',
                flexDirection: 'row',
                position: getTextPosition(),
                display: 'flex',
                flexGrow: 1,
                minHeight: 100
            }, uiText: {
                value: getText(),
                color: getTextColor(),
                fontSize: '1vw',
                textAlign: 'middle-left'
            }
        }),
        ReactEcs.createElement(UiEntity, {
            uiTransform: {
                width: getScaledButtonWidth(450),
                alignItems: 'center',
                flexDirection: 'row',
                justifyContent: 'center',
                display: displayFirstButtonContainer() ? 'flex' : 'none',
            }
        },
            ReactEcs.createElement(UiEntity, {
                onMouseDown: () => {
                    buttonClick(0);
                }
            },
                ReactEcs.createElement(UiEntity, {
                    uiTransform: {
                        height: 'auto',
                        width: getScaledSize(12)
                    }, uiBackground: {
                        textureMode: 'stretch',
                        texture: {
                            src: getTheme()
                        },
                        uvs: getImageAtlasMapping(darkButtonCorner)
                    }
                }),
                ReactEcs.createElement(UiEntity, {
                    uiTransform: {
                        width: getScaledButtonWidth(0),
                        maxWidth: getScaledSize(300),
                        height: getScaledSize(45),
                        flexDirection: 'row',
                        alignItems: 'center',
                        justifyContent: 'flex-start',
                        alignContent: 'flex-start',
                        display: displayButton(1) ? 'flex' : 'none',
                    }, uiBackground: {
                        textureMode: 'stretch',
                        texture: {
                            src: getTheme()
                        },
                        uvs: getImageAtlasMapping(darkButtonSection)
                    }
                },
                    ReactEcs.createElement(UiEntity, {
                        uiTransform: {
                            width: getScaledSize(25),
                            height: getScaledSize(25),
                            margin: { right: getScaledSize(5) },
                            positionType: 'absolute'
                        }, uiBackground: {
                            textureMode: 'stretch',
                            texture: {
                                src: getTheme()
                            },
                            uvs: getImageAtlasMapping(secondaryButtonSection)
                        }
                    }),
                    ReactEcs.createElement(UiEntity, {
                        uiTransform: {
                            width: 'auto',
                            overflow: 'hidden',
                            maxWidth: getScaledSize(217),
                            padding: { right: getScaledSize(5) },
                            margin: { left: getScaledSize(30) }
                        }, uiText: { value: getButtonText(0), fontSize: getScaledFontSize(getButtonFontSize(0)), textAlign: 'middle-left' }
                    })),
                ReactEcs.createElement(UiEntity, {
                    uiTransform: {
                        height: 'auto',
                        width: getScaledSize(12),
                        margin: { right: '5%' },
                    }, uiBackground: {
                        textureMode: 'stretch',
                        texture: {
                            src: getTheme()
                        },
                        uvs: getImageAtlasMapping(darkButtonEdge)
                    }
                })),
            ReactEcs.createElement(UiEntity, {
                onMouseDown: () => {
                    buttonClick(1);
                }
            },
                ReactEcs.createElement(UiEntity, {
                    uiTransform: {
                        height: 'auto',
                        width: getScaledSize(12)
                    }, uiBackground: {
                        textureMode: 'stretch',
                        texture: {
                            src: getTheme()
                        },
                        uvs: getImageAtlasMapping(redButtonCorner)
                    }
                }),
                ReactEcs.createElement(UiEntity, {
                    uiTransform: {
                        width: getScaledButtonWidth(1),
                        maxWidth: getScaledSize(300),
                        height: getScaledSize(45),
                        flexDirection: 'row',
                        alignItems: 'center',
                        justifyContent: 'flex-start',
                        alignContent: 'flex-start',
                        display: displayButton(2) ? 'flex' : 'none',
                    }, uiBackground: {
                        textureMode: 'stretch',
                        texture: {
                            src: getTheme()
                        },
                        uvs: getImageAtlasMapping(redButtonSection)
                    }
                },
                    ReactEcs.createElement(UiEntity, {
                        uiTransform: {
                            width: getScaledSize(25),
                            height: getScaledSize(25),
                            margin: { right: getScaledSize(5) },
                            positionType: 'absolute',
                        }, uiBackground: {
                            textureMode: 'stretch',
                            texture: {
                                src: getTheme()
                            },
                            uvs: getImageAtlasMapping(primaryButtonSection)
                        }
                    }),
                    ReactEcs.createElement(UiEntity, {
                        uiTransform: {
                            width: 'auto',
                            maxWidth: getScaledSize(217),
                            overflow: 'hidden',
                            padding: { right: getScaledSize(5) },
                            margin: { left: getScaledSize(30) }
                        }, uiText: { value: getButtonText(1), fontSize: getScaledFontSize(getButtonFontSize(1)), textAlign: 'middle-left' }
                    })),
                ReactEcs.createElement(UiEntity, {
                    uiTransform: {
                        height: 'auto',
                        width: getScaledSize(12)
                    }, uiBackground: {
                        textureMode: 'stretch',
                        texture: {
                            src: getTheme()
                        },
                        uvs: getImageAtlasMapping(redButtonEdge)
                    }
                }))),
        ReactEcs.createElement(UiEntity, {
            uiTransform: {
                width: getScaledButtonWidth(450),
                alignItems: 'center',
                flexDirection: 'row',
                justifyContent: 'center',
                margin: { top: getScaledSize(20) },
                display: displaySecondButtonContainer() ? 'flex' : 'none',
            }
        },
            ReactEcs.createElement(UiEntity, {
                onMouseDown: () => {
                    buttonClick(3);
                }
            },
                ReactEcs.createElement(UiEntity, {
                    uiTransform: {
                        height: 'auto',
                        width: getScaledSize(12),
                    }, uiBackground: {
                        textureMode: 'stretch',
                        texture: {
                            src: getTheme()
                        },
                        uvs: getImageAtlasMapping(darkButtonCorner)
                    }
                }),
                ReactEcs.createElement(UiEntity, {
                    uiTransform: {
                        width: getScaledButtonWidth(2),
                        maxWidth: getScaledSize(300),
                        overflow: 'hidden',
                        height: getScaledSize(45),
                        alignItems: 'center',
                        justifyContent: 'center',
                        alignContent: 'flex-start',
                        display: displayButton(3) ? 'flex' : 'none'
                    }, uiBackground: {
                        textureMode: 'stretch',
                        texture: {
                            src: getTheme()
                        },
                        uvs: getImageAtlasMapping(darkButtonSection)
                    }
                },
                    ReactEcs.createElement(UiEntity, {
                        uiTransform: {
                            width: 'auto',
                            maxWidth: getScaledSize(252),
                            overflow: 'hidden',
                        }, uiText: { value: getButtonText(2), fontSize: getScaledFontSize(getButtonFontSize(2)), textAlign: 'middle-left' }
                    })),
                ReactEcs.createElement(UiEntity, {
                    uiTransform: {
                        height: 'auto',
                        width: getScaledSize(12),
                        margin: { right: '5%' },
                    }, uiBackground: {
                        textureMode: 'stretch',
                        texture: {
                            src: getTheme()
                        },
                        uvs: getImageAtlasMapping(darkButtonEdge)
                    }
                })),
            ReactEcs.createElement(UiEntity, {
                onMouseDown: () => {
                    buttonClick(4);
                }
            },
                ReactEcs.createElement(UiEntity, {
                    uiTransform: {
                        height: 'auto',
                        width: getScaledSize(12),
                    }, uiBackground: {
                        textureMode: 'stretch',
                        texture: {
                            src: getTheme()
                        },
                        uvs: getImageAtlasMapping(darkButtonCorner)
                    }
                }),
                ReactEcs.createElement(UiEntity, {
                    uiTransform: {
                        width: getScaledButtonWidth(3),
                        maxWidth: getScaledSize(300),
                        overflow: 'hidden',
                        height: getScaledSize(45),
                        alignItems: 'center',
                        justifyContent: 'center',
                        alignContent: 'flex-start',
                        display: displayButton(4) ? 'flex' : 'none',
                    }, uiBackground: {
                        textureMode: 'stretch',
                        texture: {
                            src: getTheme()
                        },
                        uvs: getImageAtlasMapping(darkButtonSection)
                    }
                },
                    ReactEcs.createElement(UiEntity, {
                        uiTransform: {
                            width: 'auto',
                            maxWidth: getScaledSize(252),
                            overflow: 'hidden',
                        }, uiText: { value: getButtonText(3), fontSize: getScaledFontSize(getButtonFontSize(3)), textAlign: 'middle-left' }
                    })),
                ReactEcs.createElement(UiEntity, {
                    uiTransform: {
                        height: 'auto',
                        width: getScaledSize(12),
                    }, uiBackground: {
                        textureMode: 'stretch',
                        texture: {
                            src: getTheme()
                        },
                        uvs: getImageAtlasMapping(darkButtonEdge)
                    }
                })))));
};
