// @dcl/sdk-commands/dist/commands/deploy/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.main = exports.help = exports.args = void 0;
const path_1 = require("path");
const schemas_1 = require("@dcl/schemas");
const dcl_catalyst_client_1 = require("dcl-catalyst-client");
const fp_future_1 = __importDefault(require("fp-future"));
const scene_validations_1 = require("../../logic/scene-validations");
const args_1 = require("../../logic/args");
const error_1 = require("../../logic/error");
const beautiful_logs_1 = require("../../logic/beautiful-logs");
const project_files_1 = require("../../logic/project-files");
const utils_1 = require("./utils");
const build_1 = require("../build");
const workspace_validations_1 = require("../../logic/workspace-validations");
const analytics_features_1 = require("./analytics-features");
exports.args = (0, args_1.declareArgs)({
    '--dir': String,
    '--help': Boolean,
    '-h': '--help',
    '--target': String,
    '-t': '--target',
    '--target-content': String,
    '-tc': '--target-content',
    '--skip-validations': Boolean,
    '--skip-version-checks': Boolean,
    '--skip-build': Boolean,
    '--https': Boolean,
    '--force-upload': Boolean,
    '--yes': Boolean,
    '--no-browser': Boolean,
    '-b': '--no-browser',
    '--port': Number,
    '-p': '--port'
});
function help(options) {
    options.components.logger.log(`
  Usage: 'sdk-commands build [options]'
    Options:
      -h, --help                Displays complete help
      -p, --port        [port]  Select a custom port for the development server
      -t, --target              Specifies the address and port for the target catalyst server. Defaults to peer.decentraland.org
      -t, --target-content      Specifies the address and port for the target content server. Example: 'peer.decentraland.org/content'. Can't be set together with --target
      -b, --no-browser          Do not open a new browser window
      --skip-version-checks     Skip the ECS and CLI version checks, avoid the warning message and launch anyway
      --skip-build              Skip build before deploy
      --skip-validations        Skip permissions verifications on the client side when deploying content

    Example:
    - Deploy your scene:
      $ sdk-commands deploy
    - Deploy your scene to a specific content server:
      $ sdk-commands deploy --target my-favorite-catalyst-server.org:2323
`);
}
exports.help = help;
async function main(options) {
    const projectRoot = (0, path_1.resolve)(process.cwd(), options.args['--dir'] || '.');
    const workspace = await (0, workspace_validations_1.getValidWorkspace)(options.components, projectRoot);
    const project = workspace.projects[0];
    const openBrowser = !options.args['--no-browser'];
    const skipBuild = options.args['--skip-build'];
    const linkerPort = options.args['--port'];
    if (workspace.projects.length !== 1) {
        throw new error_1.CliError('Workspace is not supported for deploy command.');
    }
    if (project.kind !== 'scene') {
        throw new error_1.CliError('You can only deploy scenes.');
    }
    if (options.args['--target'] && options.args['--target-content']) {
        throw new error_1.CliError(`You can't set both the 'target' and 'target-content' arguments.`);
    }
    const sceneJson = await (0, scene_validations_1.getValidSceneJson)(options.components, projectRoot, { log: true });
    const coords = (0, scene_validations_1.getBaseCoords)(sceneJson);
    const isWorld = (0, utils_1.sceneHasWorldCfg)(sceneJson);
    const trackProps = {
        projectHash: await (0, project_files_1.b64HashingFunction)(projectRoot),
        coords,
        isWorld
    };
    const packageJson = await (0, project_files_1.getPackageJson)(options.components, projectRoot);
    const dependencies = Array.from(new Set([...Object.keys(packageJson.dependencies || {}), ...Object.keys(packageJson.devDependencies || {})]));
    options.components.analytics.track('Scene deploy started', trackProps);
    if (!skipBuild) {
        await (0, build_1.buildScene)({ ...options, args: { '--dir': projectRoot, '--watch': false, '--production': true, _: [] } }, project);
    }
    // Obtain list of files to deploy
    const files = await (0, scene_validations_1.getFiles)(options.components, projectRoot);
    (0, scene_validations_1.validateFilesSizes)(files);
    const contentFiles = new Map(files.map((file) => [file.path, file.content]));
    const trackFeatures = await (0, analytics_features_1.analyticsFeatures)(options.components, sceneJson.main);
    const { entityId, files: entityFiles } = await dcl_catalyst_client_1.DeploymentBuilder.buildEntity({
        type: schemas_1.EntityType.SCENE,
        pointers: sceneJson.scene.parcels,
        files: contentFiles,
        metadata: sceneJson
    });
    // Signing message
    const messageToSign = entityId;
    // Start the linker dapp and wait for the user to sign in (linker response).
    // And then close the program
    const awaitResponse = (0, fp_future_1.default)();
    const { program } = await (0, utils_1.getAddressAndSignature)(options.components, awaitResponse, messageToSign, sceneJson, files, !!options.args['--skip-validations'] || !!options.args['--target'] || !!options.args['--target-content'], {
        openBrowser,
        linkerPort,
        isHttps: !!options.args['--https']
    }, deployEntity);
    try {
        // Keep the CLI live till the user signs the payload
        await awaitResponse;
    }
    finally {
        void program?.stop();
    }
    async function deployEntity(linkerResponse) {
        const { authChain, chainId } = linkerResponse;
        // Uploading data
        const { client, url } = await (0, utils_1.getCatalyst)(chainId, options.args['--target'], options.args['--target-content']);
        (0, beautiful_logs_1.printProgressInfo)(options.components.logger, `Uploading data to: ${url}...`);
        const deployData = { entityId, files: entityFiles, authChain };
        const position = sceneJson.scene.base;
        const network = chainId === schemas_1.ChainId.ETHEREUM_SEPOLIA ? 'sepolia' : 'mainnet';
        const worldRealm = isWorld ? `&realm=${sceneJson.worldConfiguration?.name}` : '';
        const domain = chainId === schemas_1.ChainId.ETHEREUM_MAINNET ? 'https://play.decentraland.org' : 'https://play.decentraland.zone';
        const sceneUrl = `${domain}/?NETWORK=${network}&position=${position}${worldRealm}`;
        try {
            options.components.logger.info(`Address: ${linkerResponse.address}`);
            options.components.logger.info(`AuthChain: ${linkerResponse.authChain}`);
            options.components.logger.info(`Network: ${(0, schemas_1.getChainName)(linkerResponse.chainId)}`);
            const response = (await client.deploy(deployData, {
                timeout: 600000
            }));
            if (response.message) {
                (0, beautiful_logs_1.printProgressInfo)(options.components.logger, response.message);
            }
            (0, beautiful_logs_1.printSuccess)(options.components.logger, 'Content uploaded successfully', sceneUrl);
            options.components.analytics.track('Scene deploy success', {
                ...trackProps,
                sceneId: entityId,
                targetContentServer: url,
                worldName: sceneJson.worldConfiguration?.name,
                isPortableExperience: !!sceneJson.isPortableExperience,
                dependencies,
                serverlessMultiplayer: trackFeatures.serverlessMultiplayer
            });
            if (!isWorld) {
                options.components.logger.info('You can close the terminal now');
            }
        }
        catch (e) {
            options.components.logger.error('Could not upload content:');
            options.components.logger.error(e.message);
            options.components.analytics.track('Scene deploy failure', { ...trackProps, error: e.message ?? '' });
            throw e;
        }
    }
}
exports.main = main;
